{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2fa317db",
   "metadata": {},
   "source": [
    "## Setup and Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "245bf740",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working directory: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "from pathlib import Path\n",
    "import polars as pl\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "from rolypoly.utils.bio.trees import TaxonomyTree\n",
    "from rolypoly.utils.logging.loggit import setup_logging\n",
    "\n",
    "# Configuration\n",
    "threads = 6\n",
    "data_dir = \"/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\"\n",
    "log_file = \"notebooks/Exprimental/rrna_genome_mapping.log\"\n",
    "\n",
    "logger = setup_logging(log_file)\n",
    "print(f\"Working directory: {data_dir}\")\n",
    "\n",
    "# Directory structure\n",
    "rrna_dir = os.path.join(data_dir, \"contam\", \"rrna\")\n",
    "taxonomy_dir = os.path.join(data_dir, \"taxdump\")\n",
    "os.makedirs(rrna_dir, exist_ok=True)\n",
    "os.makedirs(taxonomy_dir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6abd421f",
   "metadata": {},
   "source": [
    "## Load Input Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f94aef8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading NCBI taxonomy...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7da4357490004ec89b722e6596f79450",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Reading nodes: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 2,708,826 taxonomy nodes\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9443d7aba64a4f448ccda8ab160f7d2e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Reading names: 0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 2,708,826 scientific names\n"
     ]
    }
   ],
   "source": [
    "# Load NCBI taxonomy data\n",
    "print(\"Loading NCBI taxonomy...\")\n",
    "nodes_path = os.path.join(taxonomy_dir, \"nodes.dmp\")\n",
    "names_path = os.path.join(taxonomy_dir, \"names.dmp\")\n",
    "\n",
    "# Parse nodes.dmp\n",
    "records = []\n",
    "with open(nodes_path, \"r\") as f:\n",
    "    for line in tqdm(f, desc=\"Reading nodes\"):\n",
    "        fields = line.rstrip(\"\\n\").rstrip(\"\\t|\").split(\"\\t|\\t\")\n",
    "        records.append({\n",
    "            \"tax_id\": int(fields[0]),\n",
    "            \"parent_tax_id\": int(fields[1]),\n",
    "            \"rank\": fields[2]\n",
    "        })\n",
    "nodes_df = pl.from_records(records)\n",
    "print(f\"Loaded {nodes_df.height:,} taxonomy nodes\")\n",
    "\n",
    "# Parse names.dmp (scientific names only)\n",
    "records = []\n",
    "with open(names_path, \"r\") as f:\n",
    "    for line in tqdm(f, desc=\"Reading names\"):\n",
    "        fields = line.rstrip(\"\\n\").rstrip(\"\\t|\").split(\"\\t|\\t\")\n",
    "        if fields[3] == \"scientific name\":\n",
    "            records.append({\n",
    "                \"tax_id\": int(fields[0]),\n",
    "                \"scientific_name\": fields[1]\n",
    "            })\n",
    "sci_name_df = pl.from_records(records)\n",
    "print(f\"Loaded {sci_name_df.height:,} scientific names\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc43130a",
   "metadata": {},
   "source": [
    "## Load NCBI + SILVA rRNA sequences\n",
    "reminder -\n",
    "`taxid@Source@type@seq_md5_hash`\n",
    "Using @ as separator as it's not common in fasta headers, and just going to assume the original header doesn't contain @, and that mappers who drop everything after the first white space will just ignore the really long headers\n",
    "```fasta\n",
    ">2888294@NCBI@ITS_RefSeq_Fungi@55c18f2cfc8490f5d5cb9b34162a44f5\n",
    "atgtgatgaga...\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ddb7552a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loading rRNA sequences...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (371_978, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>header</th><th>sequence</th><th>taxid</th><th>source</th><th>type</th><th>seq_hash</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;2888294@NCBI@ITS_RefSeq_Fungi@…</td><td>&quot;CAAGGTTTCCGTAGGTGAACCTGCGGAAGG…</td><td>&quot;2888294&quot;</td><td>&quot;NCBI&quot;</td><td>&quot;ITS_RefSeq_Fungi&quot;</td><td>&quot;55c18f2cfc8490f5d5cb9b34162a44…</td></tr><tr><td>&quot;46767@NCBI@ITS_eukaryote_seque…</td><td>&quot;GAGTTTCTAGTCTGTAAAAGGAATAGAGAT…</td><td>&quot;46767&quot;</td><td>&quot;NCBI&quot;</td><td>&quot;ITS_eukaryote_sequences&quot;</td><td>&quot;1a46fd8eddab80f42a516695114fae…</td></tr><tr><td>&quot;114883@NCBI@16S_ribosomal_RNA@…</td><td>&quot;CTGGCTGTGTGCCTAATACATGCATGTCGA…</td><td>&quot;114883&quot;</td><td>&quot;NCBI&quot;</td><td>&quot;16S_ribosomal_RNA&quot;</td><td>&quot;11454c6190b13c7d374db466485ae0…</td></tr><tr><td>&quot;44386@NCBI@ITS_eukaryote_seque…</td><td>&quot;GTAGGTGAACCTGCGGAAGGATCATTACCG…</td><td>&quot;44386&quot;</td><td>&quot;NCBI&quot;</td><td>&quot;ITS_eukaryote_sequences&quot;</td><td>&quot;b86fa7d27eeae7bfc4cda72eb6d41d…</td></tr><tr><td>&quot;7107@NCBI@ITS_eukaryote_sequen…</td><td>&quot;CCCCATAAACGAGGAATTCCTAGTAAGCGC…</td><td>&quot;7107&quot;</td><td>&quot;NCBI&quot;</td><td>&quot;ITS_eukaryote_sequences&quot;</td><td>&quot;568fc18e099967dea2d9f9070a248d…</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;28031@silva@SSU@0663d00aed8d6d…</td><td>&quot;CCCUAAUACAUGCAAGUCGAGCGAACAGAG…</td><td>&quot;28031&quot;</td><td>&quot;silva&quot;</td><td>&quot;SSU&quot;</td><td>&quot;0663d00aed8d6d56790b7717932896…</td></tr><tr><td>&quot;214473@silva@SSU@aceef2efde177…</td><td>&quot;GGGGGGGGGGGGGGCUAUACAUGCAAGUCG…</td><td>&quot;214473&quot;</td><td>&quot;silva&quot;</td><td>&quot;SSU&quot;</td><td>&quot;aceef2efde177d51b49b7e5dca236d…</td></tr><tr><td>&quot;1729607@silva@SSU@2e04e968777a…</td><td>&quot;GACAGAUCCCUUCGGGGAGACGUGGUGUGG…</td><td>&quot;1729607&quot;</td><td>&quot;silva&quot;</td><td>&quot;SSU&quot;</td><td>&quot;2e04e968777a4ca756a38de76b650b…</td></tr><tr><td>&quot;1399115@silva@SSU@853c6dae780c…</td><td>&quot;UUCAUGGAGAGUUUGAUCCUGGCUCAGGAC…</td><td>&quot;1399115&quot;</td><td>&quot;silva&quot;</td><td>&quot;SSU&quot;</td><td>&quot;853c6dae780c287eec1ce2361f3653…</td></tr><tr><td>&quot;45450@silva@LSU@d7d2d8fed30752…</td><td>&quot;CUGAAUUUAAGCAUAUUAAUGAGGGGAGGA…</td><td>&quot;45450&quot;</td><td>&quot;silva&quot;</td><td>&quot;LSU&quot;</td><td>&quot;d7d2d8fed307529345a5667235ed07…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (371_978, 6)\n",
       "┌───────────────────┬───────────────────┬─────────┬────────┬───────────────────┬───────────────────┐\n",
       "│ header            ┆ sequence          ┆ taxid   ┆ source ┆ type              ┆ seq_hash          │\n",
       "│ ---               ┆ ---               ┆ ---     ┆ ---    ┆ ---               ┆ ---               │\n",
       "│ str               ┆ str               ┆ str     ┆ str    ┆ str               ┆ str               │\n",
       "╞═══════════════════╪═══════════════════╪═════════╪════════╪═══════════════════╪═══════════════════╡\n",
       "│ 2888294@NCBI@ITS_ ┆ CAAGGTTTCCGTAGGTG ┆ 2888294 ┆ NCBI   ┆ ITS_RefSeq_Fungi  ┆ 55c18f2cfc8490f5d │\n",
       "│ RefSeq_Fungi@…    ┆ AACCTGCGGAAGG…    ┆         ┆        ┆                   ┆ 5cb9b34162a44…    │\n",
       "│ 46767@NCBI@ITS_eu ┆ GAGTTTCTAGTCTGTAA ┆ 46767   ┆ NCBI   ┆ ITS_eukaryote_seq ┆ 1a46fd8eddab80f42 │\n",
       "│ karyote_seque…    ┆ AAGGAATAGAGAT…    ┆         ┆        ┆ uences            ┆ a516695114fae…    │\n",
       "│ 114883@NCBI@16S_r ┆ CTGGCTGTGTGCCTAAT ┆ 114883  ┆ NCBI   ┆ 16S_ribosomal_RNA ┆ 11454c6190b13c7d3 │\n",
       "│ ibosomal_RNA@…    ┆ ACATGCATGTCGA…    ┆         ┆        ┆                   ┆ 74db466485ae0…    │\n",
       "│ 44386@NCBI@ITS_eu ┆ GTAGGTGAACCTGCGGA ┆ 44386   ┆ NCBI   ┆ ITS_eukaryote_seq ┆ b86fa7d27eeae7bfc │\n",
       "│ karyote_seque…    ┆ AGGATCATTACCG…    ┆         ┆        ┆ uences            ┆ 4cda72eb6d41d…    │\n",
       "│ 7107@NCBI@ITS_euk ┆ CCCCATAAACGAGGAAT ┆ 7107    ┆ NCBI   ┆ ITS_eukaryote_seq ┆ 568fc18e099967dea │\n",
       "│ aryote_sequen…    ┆ TCCTAGTAAGCGC…    ┆         ┆        ┆ uences            ┆ 2d9f9070a248d…    │\n",
       "│ …                 ┆ …                 ┆ …       ┆ …      ┆ …                 ┆ …                 │\n",
       "│ 28031@silva@SSU@0 ┆ CCCUAAUACAUGCAAGU ┆ 28031   ┆ silva  ┆ SSU               ┆ 0663d00aed8d6d567 │\n",
       "│ 663d00aed8d6d…    ┆ CGAGCGAACAGAG…    ┆         ┆        ┆                   ┆ 90b7717932896…    │\n",
       "│ 214473@silva@SSU@ ┆ GGGGGGGGGGGGGGCUA ┆ 214473  ┆ silva  ┆ SSU               ┆ aceef2efde177d51b │\n",
       "│ aceef2efde177…    ┆ UACAUGCAAGUCG…    ┆         ┆        ┆                   ┆ 49b7e5dca236d…    │\n",
       "│ 1729607@silva@SSU ┆ GACAGAUCCCUUCGGGG ┆ 1729607 ┆ silva  ┆ SSU               ┆ 2e04e968777a4ca75 │\n",
       "│ @2e04e968777a…    ┆ AGACGUGGUGUGG…    ┆         ┆        ┆                   ┆ 6a38de76b650b…    │\n",
       "│ 1399115@silva@SSU ┆ UUCAUGGAGAGUUUGAU ┆ 1399115 ┆ silva  ┆ SSU               ┆ 853c6dae780c287ee │\n",
       "│ @853c6dae780c…    ┆ CCUGGCUCAGGAC…    ┆         ┆        ┆                   ┆ c1ce2361f3653…    │\n",
       "│ 45450@silva@LSU@d ┆ CUGAAUUUAAGCAUAUU ┆ 45450   ┆ silva  ┆ LSU               ┆ d7d2d8fed30752934 │\n",
       "│ 7d2d8fed30752…    ┆ AAUGAGGGGAGGA…    ┆         ┆        ┆                   ┆ 5a5667235ed07…    │\n",
       "└───────────────────┴───────────────────┴─────────┴────────┴───────────────────┴───────────────────┘"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load rRNA sequences\n",
    "from rolypoly.utils.bio.polars_fastx import from_fastx_eager\n",
    "\n",
    "print(\"\\nLoading rRNA sequences...\")\n",
    "rrna_fas1 = Path(rrna_dir) / \"ncbi_rRNA_all_sequences_masked_entropy.fasta\" \n",
    "rrna_fas2 = Path(rrna_dir) / \"silva_rRNA_all_sequences_masked_entropy.fasta\"  \n",
    "rrna_df = pl.concat(\n",
    "    [\n",
    "        from_fastx_eager(rrna_fas1),\n",
    "        from_fastx_eager(rrna_fas2)\n",
    "    ]\n",
    ")\n",
    "rrna_df = rrna_df.with_columns(\n",
    "    pl.col(\"header\").str.split(\"@\").list.to_struct(fields=[\"taxid\", \"source\", \"type\", \"seq_hash\"]).alias(\"header_struct\")\n",
    ").unnest(\"header_struct\")\n",
    "rrna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cd6d8eca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 371,978 rRNA sequences\n",
      "Unique tax_ids in rRNA data: 172,219\n"
     ]
    }
   ],
   "source": [
    "rrna_df = rrna_df.with_columns(\n",
    "    pl.col(\"taxid\").cast(pl.Int64)\n",
    ")\n",
    "\n",
    "print(f\"Loaded {rrna_df.height:,} rRNA sequences\")\n",
    "print(f\"Unique tax_ids in rRNA data: {rrna_df['taxid'].n_unique():,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "15953926",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Loading NCBI assembly data...\n",
      "Selected 244,185 best assemblies\n"
     ]
    }
   ],
   "source": [
    "# Load GenBank assembly summary\n",
    "print(\"\\nLoading NCBI assembly data...\")\n",
    "ncbi_summary = pl.read_parquet(\n",
    "    os.path.join(rrna_dir, \"ncbi_assembly_summary.parquet\")\n",
    ")\n",
    "\n",
    "# Select best assembly per tax_id\n",
    "mini_ncbi = ncbi_summary.sort(\n",
    "    by=[\n",
    "        pl.col(\"refseq_category\").reverse(),\n",
    "        pl.col(\"protein_coding_gene_count\").cast(pl.Int64).reverse(),\n",
    "        pl.col(\"non_coding_gene_count\").cast(pl.Int64).reverse(),\n",
    "        pl.col(\"seq_rel_date\").reverse(),\n",
    "        pl.col(\"genome_size\").reverse(),\n",
    "    ]\n",
    ").unique(subset=[\"ncbi_taxonid\"], keep=\"first\")\n",
    "\n",
    "print(f\"Selected {mini_ncbi.height:,} best assemblies\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decdd159",
   "metadata": {},
   "source": [
    "## Build Taxonomy Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6da6ec4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Building taxonomy tree...\n",
      "Built tree with 2708826 nodes, 2457509 leaf nodes\n",
      "Set data availability for 179968 tax_ids\n",
      "Stored metadata for 179968 tax_ids\n",
      "\n",
      "Propagating data to ancestor nodes...\n",
      "\n",
      "Propagating data to ancestor nodes using method: any\n",
      "Propagation complete:\n",
      "  Initial nodes with data: 179,968\n",
      "  Nodes updated: 33,191\n",
      "  Final nodes with data: 213,159\n",
      "Updated 33,191 ancestor nodes with descendant data\n",
      "TAXONOMY TREE STATISTICS\n",
      "\n",
      "Tree Structure:\n",
      "  Total nodes:          2,708,826\n",
      "  Leaf nodes:           2,457,509\n",
      "  Internal nodes:       251,317\n",
      "\n",
      "Data Availability:\n",
      "  Nodes with data:      213,159\n",
      "  Leaves with data:     174,662\n",
      "  Leaves without data:  2,282,847\n",
      "  Internal nodes w/data:38,497\n",
      "\n",
      "Coverage:\n",
      "  All nodes:            7.87%\n",
      "  Leaf nodes only:      7.11%\n",
      "\n",
      "Cache:\n",
      "  Cached queries:       0\n",
      "\n",
      "Priority columns: protein_coding_gene_count, genome_size\n",
      "\n",
      "Top ranks by count:\n",
      "  species             : 2,223,790 total (104,684 with data,   4.7%)\n",
      "  no rank             :  253,736 total (67,306 with data,  26.5%)\n",
      "  genus               :  113,698 total ( 9,609 with data,   8.5%)\n",
      "  strain              :   46,928 total (24,841 with data,  52.9%)\n",
      "  subspecies          :   30,788 total (   515 with data,   1.7%)\n",
      "  family              :   10,817 total ( 2,339 with data,  21.6%)\n",
      "  varietas            :   10,355 total (    29 with data,   0.3%)\n",
      "  subfamily           :    3,349 total (   410 with data,  12.2%)\n",
      "  tribe               :    2,423 total (   143 with data,   5.9%)\n",
      "  order               :    2,069 total (   878 with data,  42.4%)\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "# Prepare nodes with names\n",
    "print(\"\\nBuilding taxonomy tree...\")\n",
    "nodes_with_names = nodes_df.join(\n",
    "    sci_name_df,\n",
    "    on=\"tax_id\",\n",
    "    how=\"left\"\n",
    ").select([\"tax_id\", \"parent_tax_id\", \"rank\", \"scientific_name\"])\n",
    "\n",
    "# Prepare data availability with metadata\n",
    "data_available = mini_ncbi.filter(\n",
    "    pl.col(\"total_gene_count\") > 0\n",
    ").select([\n",
    "    pl.col(\"ncbi_taxonid\").alias(\"tax_id\"),\n",
    "    \"ftp_path\",\n",
    "    \"protein_coding_gene_count\",\n",
    "    \"total_gene_count\",\n",
    "    \"genome_size\",\n",
    "    \"assembly_level\",\n",
    "    \"organism_name\"\n",
    "])\n",
    "\n",
    "# Create tree\n",
    "tax_tree = TaxonomyTree(\n",
    "    nodes_with_names,\n",
    "    data_availability_df=data_available,\n",
    "    priority_columns=['protein_coding_gene_count', 'genome_size']\n",
    ")\n",
    "\n",
    "# Propagate data to ancestors\n",
    "print(\"\\nPropagating data to ancestor nodes...\")\n",
    "nodes_updated = tax_tree.propagate_data_to_ancestors(aggregation_method='any')\n",
    "print(f\"Updated {nodes_updated:,} ancestor nodes with descendant data\")\n",
    "\n",
    "# Print statistics\n",
    "tax_tree.print_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936401c6",
   "metadata": {},
   "source": [
    "## Map rRNA Sequences to Genome References (Genus-Constrained)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "20ea9b19",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Mapping 172,219 rRNA tax_ids to genome references...\n",
      "Searching 172,219 tax_ids with max_rank='genus'\n",
      "\n",
      "Results:\n",
      "  Total queried:         172,219\n",
      "  Self has data:         26,937 (15.6%)\n",
      "  Ancestor found:        1,058 (0.6%)\n",
      "  Relative found:        50,358 (29.2%)\n",
      "  No match:              93,866 (54.5%)\n",
      "  Limited by genus rank: 134,832 (78.3%)\n",
      "\n",
      "Overall coverage: 78,353 / 172,219 (45.5%)\n"
     ]
    }
   ],
   "source": [
    "# Get all rRNA tax_ids\n",
    "all_rrna_tax_ids = rrna_df[\"taxid\"].unique().to_list()\n",
    "print(f\"\\nMapping {len(all_rrna_tax_ids):,} rRNA tax_ids to genome references...\")\n",
    "\n",
    "# Run simplified search with max_rank constraint only\n",
    "results, stats = tax_tree.find_best_reference_batch(\n",
    "    tax_ids=all_rrna_tax_ids,\n",
    "    max_rank='genus',\n",
    "    return_stats=True\n",
    ")\n",
    "\n",
    "# Print statistics\n",
    "print(\"\\nResults:\")\n",
    "print(f\"  Total queried:         {stats['total_queried']:,}\")\n",
    "print(f\"  Self has data:         {stats['self_found']:,} ({stats['self_found']/stats['total_queried']*100:.1f}%)\")\n",
    "print(f\"  Ancestor found:        {stats['ancestor_found']:,} ({stats['ancestor_found']/stats['total_queried']*100:.1f}%)\")\n",
    "print(f\"  Relative found:        {stats['relative_found']:,} ({stats['relative_found']/stats['total_queried']*100:.1f}%)\")\n",
    "print(f\"  No match:              {stats['no_match_found']:,} ({stats['no_match_found']/stats['total_queried']*100:.1f}%)\")\n",
    "print(f\"  Limited by genus rank: {stats['max_rank_limited']:,} ({stats['max_rank_limited']/stats['total_queried']*100:.1f}%)\")\n",
    "\n",
    "total_with_ref = stats['self_found'] + stats['ancestor_found'] + stats['relative_found']\n",
    "print(f\"\\nOverall coverage: {total_with_ref:,} / {stats['total_queried']:,} ({total_with_ref/stats['total_queried']*100:.1f}%)\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7435ac56",
   "metadata": {},
   "source": [
    "## Extract and Save Mapping Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6daf26d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Building mapping table...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created mapping table with 172,219 entries\n"
     ]
    }
   ],
   "source": [
    "# Build mapping table - even simpler now!\n",
    "print(\"\\nBuilding mapping table...\")\n",
    "mapping_df = tax_tree.results_to_dataframe(results)\n",
    "print(f\"Created mapping table with {mapping_df.height:,} entries\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b9bf6328",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (172_219, 12)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>query_tax_id</th><th>query_name</th><th>query_rank</th><th>reference_tax_id</th><th>reference_name</th><th>reference_rank</th><th>relationship</th><th>max_rank_reached</th><th>ftp_path</th><th>protein_coding_genes</th><th>genome_size</th><th>assembly_level</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>str</td><td>i64</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>0</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>false</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>7</td><td>&quot;Azorhizobium caulinodans&quot;</td><td>&quot;species&quot;</td><td>980631</td><td>&quot;Azorhizobium oxalatiphilum&quot;</td><td>&quot;species&quot;</td><td>&quot;relative&quot;</td><td>true</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>5742</td><td>6396426</td><td>&quot;Scaffold&quot;</td></tr><tr><td>9</td><td>&quot;Buchnera aphidicola&quot;</td><td>&quot;species&quot;</td><td>9</td><td>&quot;Buchnera aphidicola&quot;</td><td>&quot;species&quot;</td><td>&quot;self&quot;</td><td>false</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>1050</td><td>657623</td><td>&quot;Contig&quot;</td></tr><tr><td>14</td><td>&quot;Dictyoglomus thermophilum&quot;</td><td>&quot;species&quot;</td><td>14</td><td>&quot;Dictyoglomus thermophilum&quot;</td><td>&quot;species&quot;</td><td>&quot;self&quot;</td><td>false</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>1778</td><td>1963348</td><td>&quot;Contig&quot;</td></tr><tr><td>17</td><td>&quot;Methylophilus methylotrophus&quot;</td><td>&quot;species&quot;</td><td>17</td><td>&quot;Methylophilus methylotrophus&quot;</td><td>&quot;species&quot;</td><td>&quot;self&quot;</td><td>false</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>1125</td><td>1188656</td><td>&quot;Contig&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3463822</td><td>&quot;Acrohelicosporium aquaticum&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>3463823</td><td>&quot;Acrohelicosporium guizhouense&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>3463824</td><td>&quot;Acrohelicosporium viridisporum&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>3463825</td><td>&quot;Helicoma astrictum&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>3464214</td><td>&quot;Phaeoclavulina aurantilaeta&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (172_219, 12)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ query_tax ┆ query_nam ┆ query_ran ┆ reference ┆ … ┆ ftp_path  ┆ protein_c ┆ genome_si ┆ assembly │\n",
       "│ _id       ┆ e         ┆ k         ┆ _tax_id   ┆   ┆ ---       ┆ oding_gen ┆ ze        ┆ _level   │\n",
       "│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ str       ┆ es        ┆ ---       ┆ ---      │\n",
       "│ i64       ┆ str       ┆ str       ┆ i64       ┆   ┆           ┆ ---       ┆ i64       ┆ str      │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ i64       ┆           ┆          │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│ 0         ┆ null      ┆ null      ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│ 7         ┆ Azorhizob ┆ species   ┆ 980631    ┆ … ┆ https://f ┆ 5742      ┆ 6396426   ┆ Scaffold │\n",
       "│           ┆ ium cauli ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆          │\n",
       "│           ┆ nodans    ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "│ 9         ┆ Buchnera  ┆ species   ┆ 9         ┆ … ┆ https://f ┆ 1050      ┆ 657623    ┆ Contig   │\n",
       "│           ┆ aphidicol ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆          │\n",
       "│           ┆ a         ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "│ 14        ┆ Dictyoglo ┆ species   ┆ 14        ┆ … ┆ https://f ┆ 1778      ┆ 1963348   ┆ Contig   │\n",
       "│           ┆ mus therm ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆          │\n",
       "│           ┆ ophilum   ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "│ 17        ┆ Methyloph ┆ species   ┆ 17        ┆ … ┆ https://f ┆ 1125      ┆ 1188656   ┆ Contig   │\n",
       "│           ┆ ilus meth ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆          │\n",
       "│           ┆ ylotrophu ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆ s         ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │\n",
       "│ 3463822   ┆ Acrohelic ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ osporium  ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ aquaticum ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 3463823   ┆ Acrohelic ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ osporium  ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ guizhouen ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ se        ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 3463824   ┆ Acrohelic ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ osporium  ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ viridispo ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ rum       ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 3463825   ┆ Helicoma  ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ astrictum ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 3464214   ┆ Phaeoclav ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ ulina aur ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ antilaeta ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3d7bf331",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Saved mapping table to: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/rrna_to_genome_mapping.parquet\n",
      "\n",
      "Sample mappings:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (10, 12)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>query_tax_id</th><th>query_name</th><th>query_rank</th><th>reference_tax_id</th><th>reference_name</th><th>reference_rank</th><th>relationship</th><th>max_rank_reached</th><th>ftp_path</th><th>protein_coding_genes</th><th>genome_size</th><th>assembly_level</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>bool</td><td>str</td><td>i64</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>424539</td><td>&quot;Ruttnera lamellosa&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>1739746</td><td>&quot;Arthrobacter sp. IN209&quot;</td><td>&quot;species&quot;</td><td>1588023</td><td>&quot;Arthrobacter sp. Hiyo8&quot;</td><td>&quot;species&quot;</td><td>&quot;relative&quot;</td><td>true</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>7038</td><td>5026722</td><td>&quot;Complete Genome&quot;</td></tr><tr><td>284853</td><td>&quot;mucus bacterium 31&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>false</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>317713</td><td>&quot;Pittosporum venulosum&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>80292</td><td>&quot;Bursera coyucensis&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>409508</td><td>&quot;Spiraea decumbens&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>948534</td><td>&quot;Schistidium tenuinerve&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>246516</td><td>&quot;Cordia oncocalyx&quot;</td><td>&quot;species&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>true</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>543519</td><td>&quot;Myxobolus leptobarbi&quot;</td><td>&quot;species&quot;</td><td>59785</td><td>&quot;Myxobolus squamalis&quot;</td><td>&quot;species&quot;</td><td>&quot;relative&quot;</td><td>true</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>5712</td><td>43671844</td><td>&quot;Contig&quot;</td></tr><tr><td>862512</td><td>&quot;Corynebacterium accolens ATCC …</td><td>&quot;strain&quot;</td><td>862512</td><td>&quot;Corynebacterium accolens ATCC …</td><td>&quot;strain&quot;</td><td>&quot;self&quot;</td><td>false</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>2360</td><td>2465636</td><td>&quot;Scaffold&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (10, 12)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ query_tax ┆ query_nam ┆ query_ran ┆ reference ┆ … ┆ ftp_path  ┆ protein_c ┆ genome_si ┆ assembly │\n",
       "│ _id       ┆ e         ┆ k         ┆ _tax_id   ┆   ┆ ---       ┆ oding_gen ┆ ze        ┆ _level   │\n",
       "│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ str       ┆ es        ┆ ---       ┆ ---      │\n",
       "│ i64       ┆ str       ┆ str       ┆ i64       ┆   ┆           ┆ ---       ┆ i64       ┆ str      │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ i64       ┆           ┆          │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│ 424539    ┆ Ruttnera  ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ lamellosa ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 1739746   ┆ Arthrobac ┆ species   ┆ 1588023   ┆ … ┆ https://f ┆ 7038      ┆ 5026722   ┆ Complete │\n",
       "│           ┆ ter sp.   ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆ Genome   │\n",
       "│           ┆ IN209     ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "│ 284853    ┆ mucus     ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ bacterium ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ 31        ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 317713    ┆ Pittospor ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ um        ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ venulosum ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 80292     ┆ Bursera   ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ coyucensi ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ s         ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 409508    ┆ Spiraea   ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ decumbens ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 948534    ┆ Schistidi ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ um tenuin ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│           ┆ erve      ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 246516    ┆ Cordia    ┆ species   ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│           ┆ oncocalyx ┆           ┆           ┆   ┆           ┆           ┆           ┆          │\n",
       "│ 543519    ┆ Myxobolus ┆ species   ┆ 59785     ┆ … ┆ https://f ┆ 5712      ┆ 43671844  ┆ Contig   │\n",
       "│           ┆ leptobarb ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆          │\n",
       "│           ┆ i         ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "│ 862512    ┆ Corynebac ┆ strain    ┆ 862512    ┆ … ┆ https://f ┆ 2360      ┆ 2465636   ┆ Scaffold │\n",
       "│           ┆ terium    ┆           ┆           ┆   ┆ tp.ncbi.n ┆           ┆           ┆          │\n",
       "│           ┆ accolens  ┆           ┆           ┆   ┆ lm.nih.go ┆           ┆           ┆          │\n",
       "│           ┆ ATCC …    ┆           ┆           ┆   ┆ v/g…      ┆           ┆           ┆          │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Save results\n",
    "output_path = os.path.join(rrna_dir, \"rrna_to_genome_mapping.parquet\")\n",
    "mapping_df.write_parquet(output_path)\n",
    "print(f\"\\nSaved mapping table to: {output_path}\")\n",
    "\n",
    "# Also save as TSV for easy viewing\n",
    "mapping_df.write_csv(\n",
    "    os.path.join(rrna_dir, \"rrna_to_genome_mapping.tsv\"),\n",
    "    separator=\"\\t\"\n",
    ")\n",
    "\n",
    "# Show sample\n",
    "print(\"\\nSample mappings:\")\n",
    "mapping_df.sample(min(10, mapping_df.height))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9328dc4a",
   "metadata": {},
   "source": [
    "## Create Additional Lookup Tables for Easy Access"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "05cb90c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Creating simplified lookup table...\n",
      "Saved simplified lookup to: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/taxid_to_ftp_lookup.parquet\n",
      "\n",
      "Lookup table statistics:\n",
      "  Total entries: 172,219\n",
      "  With FTP paths: 78,353 (45.5%)\n",
      "\n",
      "By relationship type:\n",
      "shape: (4, 2)\n",
      "┌──────────────┬───────┐\n",
      "│ relationship ┆ count │\n",
      "│ ---          ┆ ---   │\n",
      "│ str          ┆ u32   │\n",
      "╞══════════════╪═══════╡\n",
      "│ null         ┆ 93866 │\n",
      "│ relative     ┆ 50358 │\n",
      "│ self         ┆ 26937 │\n",
      "│ ancestor     ┆ 1058  │\n",
      "└──────────────┴───────┘\n"
     ]
    }
   ],
   "source": [
    "# Create a simplified taxid -> ftp_path lookup table\n",
    "# This is now trivial since our new API returns clean results\n",
    "print(\"\\nCreating simplified lookup table...\")\n",
    "\n",
    "lookup_df = mapping_df.select([\n",
    "    'query_tax_id',\n",
    "    'query_name',\n",
    "    'query_rank',\n",
    "    'reference_tax_id',\n",
    "    'ftp_path',\n",
    "    'relationship'\n",
    "]).rename({\n",
    "    'query_tax_id': 'tax_id',\n",
    "    'query_name': 'name',\n",
    "    'query_rank': 'rank'\n",
    "})\n",
    "\n",
    "# Save the simplified lookup\n",
    "lookup_path = os.path.join(rrna_dir, \"taxid_to_ftp_lookup.parquet\")\n",
    "lookup_df.write_parquet(lookup_path)\n",
    "print(f\"Saved simplified lookup to: {lookup_path}\")\n",
    "\n",
    "# Also save as TSV\n",
    "lookup_df.write_csv(\n",
    "    os.path.join(rrna_dir, \"taxid_to_ftp_lookup.tsv\"),\n",
    "    separator=\"\\t\"\n",
    ")\n",
    "\n",
    "# Show statistics\n",
    "with_ftp = lookup_df.filter(pl.col(\"ftp_path\").is_not_null()).height\n",
    "print(\"\\nLookup table statistics:\")\n",
    "print(f\"  Total entries: {lookup_df.height:,}\")\n",
    "print(f\"  With FTP paths: {with_ftp:,} ({with_ftp/lookup_df.height*100:.1f}%)\")\n",
    "print(\"\\nBy relationship type:\")\n",
    "print(lookup_df.group_by(\"relationship\").agg(pl.len().alias(\"count\")).sort(\"count\", descending=True))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75ad8a46",
   "metadata": {},
   "source": [
    "## Summary Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cae40a78",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FINAL SUMMARY\n",
      "======================================================================\n",
      "\n",
      "Input data:\n",
      "  rRNA sequences:           371,978\n",
      "  Unique rRNA tax_ids:      172,219\n",
      "  Taxonomy nodes:           2,708,826\n",
      "  GenBank assemblies:       244,185\n",
      "\n",
      "Tree after propagation:\n",
      "  Total nodes with data:    213,159\n",
      "  Leaf nodes with data:     174,662\n",
      "  Internal nodes with data: 38,497\n",
      "\n",
      "Mapping results (genus-constrained):\n",
      "  no_match       :   93,866 ( 54.5%)\n",
      "  relative       :   50,358 ( 29.2%)\n",
      "  self           :   26,937 ( 15.6%)\n",
      "  ancestor       :    1,058 (  0.6%)\n",
      "\n",
      "Total with genome reference: 78,353 / 172,219 (45.5%)\n",
      "Cases limited by genus rank: 134,832 (78.3%)\n",
      "======================================================================\n"
     ]
    }
   ],
   "source": [
    "print(\"=\" * 70)\n",
    "print(\"FINAL SUMMARY\")\n",
    "print(\"=\" * 70)\n",
    "\n",
    "print(\"\\nInput data:\")\n",
    "print(f\"  rRNA sequences:           {rrna_df.height:,}\")\n",
    "print(f\"  Unique rRNA tax_ids:      {len(all_rrna_tax_ids):,}\")\n",
    "print(f\"  Taxonomy nodes:           {nodes_df.height:,}\")\n",
    "print(f\"  GenBank assemblies:       {mini_ncbi.height:,}\")\n",
    "\n",
    "print(\"\\nTree after propagation:\")\n",
    "print(f\"  Total nodes with data:    {len(tax_tree.data_available):,}\")\n",
    "print(f\"  Leaf nodes with data:     {len(tax_tree.leaf_nodes & tax_tree.data_available):,}\")\n",
    "print(f\"  Internal nodes with data: {len(tax_tree.data_available - tax_tree.leaf_nodes):,}\")\n",
    "\n",
    "print(\"\\nMapping results (genus-constrained):\")\n",
    "by_relationship = mapping_df.group_by(\"relationship\").agg(pl.len().alias(\"count\")).sort(\"count\", descending=True)\n",
    "for row in by_relationship.iter_rows(named=True):\n",
    "    rel = row['relationship'] or 'no_match'\n",
    "    count = row['count']\n",
    "    pct = count / len(all_rrna_tax_ids) * 100\n",
    "    print(f\"  {rel:15s}: {count:8,} ({pct:5.1f}%)\")\n",
    "\n",
    "total_mapped = mapping_df.filter(pl.col(\"relationship\").is_not_null()).height\n",
    "print(f\"\\nTotal with genome reference: {total_mapped:,} / {len(all_rrna_tax_ids):,} ({total_mapped/len(all_rrna_tax_ids)*100:.1f}%)\")\n",
    "\n",
    "# Show rank-limited cases\n",
    "rank_limited = mapping_df.filter(pl.col(\"max_rank_reached\") == True).height\n",
    "print(f\"Cases limited by genus rank: {rank_limited:,} ({rank_limited/len(all_rrna_tax_ids)*100:.1f}%)\")\n",
    "print(\"=\" * 70)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "646d16b7",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
