{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "222afdb2",
   "metadata": {},
   "source": [
    "# testing tree APIs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57818007",
   "metadata": {},
   "source": [
    "Loading libraries and defining paths to sets already created/downloaded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a12c02e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting data preparation to : /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import os\n",
    "from collections import defaultdict, deque\n",
    "\n",
    "import polars as pl\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "\n",
    "from rolypoly.utils.bio.trees import TaxonomyTree\n",
    "\n",
    "\n",
    "\n",
    "### DEBUG ARGS (for manually building, not entering via CLI):\n",
    "threads = 6\n",
    "log_file = \"notebooks/Exprimental/trrna.log\"\n",
    "data_dir = \"/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3cdd589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (25_099, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>organism_name</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>null</td><td>117964</td></tr><tr><td>&quot;Giardia duodenalis&quot;</td><td>127</td></tr><tr><td>&quot;Globisporangium heterothallicu…</td><td>93</td></tr><tr><td>&quot;Babesia bovis&quot;</td><td>65</td></tr><tr><td>&quot;Globisporangium sylvaticum&quot;</td><td>61</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;Andreaea rupestris&quot;</td><td>1</td></tr><tr><td>&quot;Tetraphis pellucida&quot;</td><td>1</td></tr><tr><td>&quot;Anncaliia algerae&quot;</td><td>1</td></tr><tr><td>&quot;Dictyocoela muelleri&quot;</td><td>1</td></tr><tr><td>&quot;Vairimorpha necatrix&quot;</td><td>1</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (25_099, 2)\n",
       "┌─────────────────────────────────┬────────┐\n",
       "│ organism_name                   ┆ count  │\n",
       "│ ---                             ┆ ---    │\n",
       "│ str                             ┆ u32    │\n",
       "╞═════════════════════════════════╪════════╡\n",
       "│ null                            ┆ 117964 │\n",
       "│ Giardia duodenalis              ┆ 127    │\n",
       "│ Globisporangium heterothallicu… ┆ 93     │\n",
       "│ Babesia bovis                   ┆ 65     │\n",
       "│ Globisporangium sylvaticum      ┆ 61     │\n",
       "│ …                               ┆ …      │\n",
       "│ Andreaea rupestris              ┆ 1      │\n",
       "│ Tetraphis pellucida             ┆ 1      │\n",
       "│ Anncaliia algerae               ┆ 1      │\n",
       "│ Dictyocoela muelleri            ┆ 1      │\n",
       "│ Vairimorpha necatrix            ┆ 1      │\n",
       "└─────────────────────────────────┴────────┘"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrna_df = rrna_df = pl.scan_csv(\n",
    "    os.path.join(data_dir, \"/contam/rrna/*.tab\"),\n",
    "    separator=\"@\",\n",
    "    has_header=False,\n",
    "    null_values = [\"N/A\"],\n",
    "    new_columns=[\"taxid\", \"header\",\"name\",\"sequence\"],\n",
    "    include_file_paths=\"type\"\n",
    ").collect()\n",
    "mini_genebank = pl.read_parquet(\"/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/genbank_assembly_summary.parquet\")\n",
    "temp_df = rrna_df.join(\n",
    "    mini_genebank.select(\n",
    "        [\"ncbi_taxonid\", \"organism_name\"]\n",
    "    ).unique(),  # seq_length\n",
    "    left_on=[\"taxid\"],\n",
    "    right_on=[\"ncbi_taxonid\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "temp_df[\"organism_name\"].value_counts(sort=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe17993f",
   "metadata": {},
   "source": [
    "## NCBI taxdump"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b445183",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">02</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">19:16:59</span> - Downloading <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz</span> to                                  <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">taxdump.tar.gz</span>                                                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m02\u001b[0m \u001b[1;92m19:16:59\u001b[0m - Downloading \u001b[4;94mhttps://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz\u001b[0m to                                  \u001b]8;id=203324;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=200555;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/\u001b[0m\u001b[95mtaxdump.tar.gz\u001b[0m                                                  \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">02</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">19:17:06</span> - Successfully downloaded to <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">taxdump.tar.gz</span> <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m02\u001b[0m \u001b[1;92m19:17:06\u001b[0m - Successfully downloaded to \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/\u001b[0m\u001b[95mtaxdump.tar.gz\u001b[0m \u001b]8;id=382276;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=849077;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">02</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">19:17:06</span> - Extracting <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">taxdump.tar.gz</span> to              <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">255</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">taxdump</span>                                                         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m02\u001b[0m \u001b[1;92m19:17:06\u001b[0m - Extracting \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/\u001b[0m\u001b[95mtaxdump.tar.gz\u001b[0m to              \u001b]8;id=127845;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=832875;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\u001b\\\u001b[2m255\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/\u001b[0m\u001b[95mtaxdump\u001b[0m                                                         \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">02</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">19:17:09</span> - Successfully extracted to <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">taxdump</span>         <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">288</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m02\u001b[0m \u001b[1;92m19:17:09\u001b[0m - Successfully extracted to \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/\u001b[0m\u001b[95mtaxdump\u001b[0m         \u001b]8;id=591732;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=279401;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\u001b\\\u001b[2m288\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "PosixPath('/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/taxdump')"
      ]
     },
     "execution_count": 216,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# *.dmp files are bcp-like dump from GenBank taxonomy database.\n",
    "\n",
    "# General information.\n",
    "# Field terminator is \"\\t|\\t\"\n",
    "# Row terminator is \"\\t|\\n\"\n",
    "\n",
    "# nodes.dmp file consists of taxonomy nodes. The description for each node includes the following\n",
    "# fields:\n",
    "# \ttax_id\t\t\t\t\t-- node id in GenBank taxonomy database\n",
    "#  \tparent tax_id\t\t\t\t-- parent node id in GenBank taxonomy database\n",
    "#  \trank\t\t\t\t\t-- rank of this node (domain, kingdom, ...) \n",
    "#  \tembl code\t\t\t\t-- locus-name prefix; not unique\n",
    "#  \tdivision id\t\t\t\t-- see division.dmp file\n",
    "#  \tinherited div flag  (1 or 0)\t\t-- 1 if node inherits division from parent\n",
    "#  \tgenetic code id\t\t\t\t-- see gencode.dmp file\n",
    "#  \tinherited GC  flag  (1 or 0)\t\t-- 1 if node inherits genetic code from parent\n",
    "#  \tmitochondrial genetic code id\t\t-- see gencode.dmp file\n",
    "#  \tinherited MGC flag  (1 or 0)\t\t-- 1 if node inherits mitochondrial gencode from parent\n",
    "#  \tGenBank hidden flag (1 or 0)            -- 1 if name is suppressed in GenBank entry lineage\n",
    "#  \thidden subtree root flag (1 or 0)       -- 1 if this subtree has no sequence data yet\n",
    "#  \tcomments\t\t\t\t-- free-text comments and citations\n",
    "\n",
    "# Taxonomy names file (names.dmp):\n",
    "# \ttax_id\t\t\t\t\t-- the id of node associated with this name\n",
    "# \tname_txt\t\t\t\t-- name itself\n",
    "# \tunique name\t\t\t\t-- the unique variant of this name if name not unique\n",
    "# \tname class\t\t\t\t-- (synonym, common name, ...)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3d0963b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "79e1565234fa4446a17b11d7c6d87d7b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2708826 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (2_708_826, 12)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>tax_id</th><th>parent_tax_id</th><th>rank</th><th>embl_code</th><th>division_id</th><th>inherited_div_flag</th><th>genetic_code_id</th><th>inherited_GC_flag</th><th>mitochondrial_genetic_code_id</th><th>inherited_MGC_flag</th><th>GenBank_hidden_flag</th><th>hidden_subtree_root_flag</th></tr><tr><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>1</td><td>1</td><td>&quot;no rank&quot;</td><td>&quot;&quot;</td><td>&quot;8&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>2</td><td>131567</td><td>&quot;domain&quot;</td><td>&quot;&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;11&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>6</td><td>335928</td><td>&quot;genus&quot;</td><td>&quot;&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;11&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>7</td><td>6</td><td>&quot;species&quot;</td><td>&quot;AC&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;11&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td></tr><tr><td>9</td><td>32199</td><td>&quot;species&quot;</td><td>&quot;BA&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;11&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3475891</td><td>47833</td><td>&quot;no rank&quot;</td><td>&quot;&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;5&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>3476007</td><td>479661</td><td>&quot;varietas&quot;</td><td>&quot;RC&quot;</td><td>&quot;4&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>3476068</td><td>2653920</td><td>&quot;genus&quot;</td><td>&quot;&quot;</td><td>&quot;4&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>3476159</td><td>140471</td><td>&quot;family&quot;</td><td>&quot;&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;4&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr><tr><td>3476160</td><td>3476159</td><td>&quot;genus&quot;</td><td>&quot;&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;1&quot;</td><td>&quot;4&quot;</td><td>&quot;1&quot;</td><td>&quot;0&quot;</td><td>&quot;0&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (2_708_826, 12)\n",
       "┌─────────┬────────────┬──────────┬───────────┬───┬────────────┬───────────┬───────────┬───────────┐\n",
       "│ tax_id  ┆ parent_tax ┆ rank     ┆ embl_code ┆ … ┆ mitochondr ┆ inherited ┆ GenBank_h ┆ hidden_su │\n",
       "│ ---     ┆ _id        ┆ ---      ┆ ---       ┆   ┆ ial_geneti ┆ _MGC_flag ┆ idden_fla ┆ btree_roo │\n",
       "│ i64     ┆ ---        ┆ str      ┆ str       ┆   ┆ c_code_id  ┆ ---       ┆ g         ┆ t_flag    │\n",
       "│         ┆ i64        ┆          ┆           ┆   ┆ ---        ┆ str       ┆ ---       ┆ ---       │\n",
       "│         ┆            ┆          ┆           ┆   ┆ str        ┆           ┆ str       ┆ str       │\n",
       "╞═════════╪════════════╪══════════╪═══════════╪═══╪════════════╪═══════════╪═══════════╪═══════════╡\n",
       "│ 1       ┆ 1          ┆ no rank  ┆           ┆ … ┆ 0          ┆ 0         ┆ 0         ┆ 0         │\n",
       "│ 2       ┆ 131567     ┆ domain   ┆           ┆ … ┆ 0          ┆ 0         ┆ 0         ┆ 0         │\n",
       "│ 6       ┆ 335928     ┆ genus    ┆           ┆ … ┆ 0          ┆ 1         ┆ 0         ┆ 0         │\n",
       "│ 7       ┆ 6          ┆ species  ┆ AC        ┆ … ┆ 0          ┆ 1         ┆ 1         ┆ 0         │\n",
       "│ 9       ┆ 32199      ┆ species  ┆ BA        ┆ … ┆ 0          ┆ 1         ┆ 1         ┆ 0         │\n",
       "│ …       ┆ …          ┆ …        ┆ …         ┆ … ┆ …          ┆ …         ┆ …         ┆ …         │\n",
       "│ 3475891 ┆ 47833      ┆ no rank  ┆           ┆ … ┆ 5          ┆ 1         ┆ 0         ┆ 0         │\n",
       "│ 3476007 ┆ 479661     ┆ varietas ┆ RC        ┆ … ┆ 1          ┆ 1         ┆ 0         ┆ 0         │\n",
       "│ 3476068 ┆ 2653920    ┆ genus    ┆           ┆ … ┆ 1          ┆ 1         ┆ 0         ┆ 0         │\n",
       "│ 3476159 ┆ 140471     ┆ family   ┆           ┆ … ┆ 4          ┆ 1         ┆ 0         ┆ 0         │\n",
       "│ 3476160 ┆ 3476159    ┆ genus    ┆           ┆ … ┆ 4          ┆ 1         ┆ 0         ┆ 0         │\n",
       "└─────────┴────────────┴──────────┴───────────┴───┴────────────┴───────────┴───────────┴───────────┘"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy_dir = os.path.join(data_dir, \"taxdump\")\n",
    "nodes_path = os.path.join(taxonomy_dir, \"nodes.dmp\")\n",
    "names_path = os.path.join(taxonomy_dir, \"names.dmp\")\n",
    "# can not read with polars cause of irregular separator, so using line by line read...\n",
    "\n",
    "records = []\n",
    "with open(nodes_path, \"r\") as f:\n",
    "    nodes_lines = f.readlines()\n",
    "    for line in tqdm(nodes_lines):\n",
    "        line = line.rstrip(\"\\n\").rstrip(\"\\t|\")\n",
    "        fields = line.split(\"\\t|\\t\")\n",
    "        record = {\n",
    "            \"tax_id\": fields[0],\n",
    "            \"parent_tax_id\": fields[1],\n",
    "            \"rank\": fields[2],\n",
    "            \"embl_code\": fields[3],\n",
    "            \"division_id\": fields[4],\n",
    "            \"inherited_div_flag\": fields[5],\n",
    "            \"genetic_code_id\": fields[6],\n",
    "            \"inherited_GC_flag\": fields[7],\n",
    "            \"mitochondrial_genetic_code_id\": fields[8],\n",
    "            \"inherited_MGC_flag\": fields[9],\n",
    "            \"GenBank_hidden_flag\": fields[10],\n",
    "            \"hidden_subtree_root_flag\": fields[11],\n",
    "            # \"comments\": fields[12]\n",
    "        }\n",
    "        records.append(record)\n",
    "nodes_df = pl.from_records(records).with_columns(\n",
    "    pl.col(\"tax_id\").cast(pl.Int64),\n",
    "    pl.col(\"parent_tax_id\").cast(pl.Int64),\n",
    "    \n",
    ")\n",
    "nodes_df\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3dfc8a9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ae3ea78900564bea9d89378026ab141f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4495871 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (4_495_871, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>tax_id</th><th>name_txt</th><th>unique_name</th><th>name_class</th></tr><tr><td>i64</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>1</td><td>&quot;all&quot;</td><td>&quot;&quot;</td><td>&quot;synonym&quot;</td></tr><tr><td>1</td><td>&quot;root&quot;</td><td>&quot;&quot;</td><td>&quot;scientific name&quot;</td></tr><tr><td>2</td><td>&quot;Bacteria&quot;</td><td>&quot;Bacteria &lt;bacteria&gt;&quot;</td><td>&quot;scientific name&quot;</td></tr><tr><td>2</td><td>&quot;bacteria&quot;</td><td>&quot;bacteria &lt;blast name&gt;&quot;</td><td>&quot;blast name&quot;</td></tr><tr><td>2</td><td>&quot;bacteria&quot;</td><td>&quot;bacteria &lt;genbank common name&gt;&quot;</td><td>&quot;genbank common name&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3476068</td><td>&quot;Comoliopsis Wurdack, 1984&quot;</td><td>&quot;&quot;</td><td>&quot;authority&quot;</td></tr><tr><td>3476159</td><td>&quot;Benthoplanidae Bezio, Paulay &amp;…</td><td>&quot;&quot;</td><td>&quot;authority&quot;</td></tr><tr><td>3476159</td><td>&quot;Benthoplanidae&quot;</td><td>&quot;&quot;</td><td>&quot;scientific name&quot;</td></tr><tr><td>3476160</td><td>&quot;Benthoplana Fricke &amp; Plante, 1…</td><td>&quot;&quot;</td><td>&quot;authority&quot;</td></tr><tr><td>3476160</td><td>&quot;Benthoplana&quot;</td><td>&quot;&quot;</td><td>&quot;scientific name&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (4_495_871, 4)\n",
       "┌─────────┬─────────────────────────────────┬────────────────────────────────┬─────────────────────┐\n",
       "│ tax_id  ┆ name_txt                        ┆ unique_name                    ┆ name_class          │\n",
       "│ ---     ┆ ---                             ┆ ---                            ┆ ---                 │\n",
       "│ i64     ┆ str                             ┆ str                            ┆ str                 │\n",
       "╞═════════╪═════════════════════════════════╪════════════════════════════════╪═════════════════════╡\n",
       "│ 1       ┆ all                             ┆                                ┆ synonym             │\n",
       "│ 1       ┆ root                            ┆                                ┆ scientific name     │\n",
       "│ 2       ┆ Bacteria                        ┆ Bacteria <bacteria>            ┆ scientific name     │\n",
       "│ 2       ┆ bacteria                        ┆ bacteria <blast name>          ┆ blast name          │\n",
       "│ 2       ┆ bacteria                        ┆ bacteria <genbank common name> ┆ genbank common name │\n",
       "│ …       ┆ …                               ┆ …                              ┆ …                   │\n",
       "│ 3476068 ┆ Comoliopsis Wurdack, 1984       ┆                                ┆ authority           │\n",
       "│ 3476159 ┆ Benthoplanidae Bezio, Paulay &… ┆                                ┆ authority           │\n",
       "│ 3476159 ┆ Benthoplanidae                  ┆                                ┆ scientific name     │\n",
       "│ 3476160 ┆ Benthoplana Fricke & Plante, 1… ┆                                ┆ authority           │\n",
       "│ 3476160 ┆ Benthoplana                     ┆                                ┆ scientific name     │\n",
       "└─────────┴─────────────────────────────────┴────────────────────────────────┴─────────────────────┘"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records = []\n",
    "with open(names_path, \"r\") as f:\n",
    "    nodes_lines = f.readlines()\n",
    "    for line in tqdm(nodes_lines):\n",
    "        line = line.rstrip(\"\\n\").rstrip(\"\\t|\")\n",
    "        fields = line.split(\"\\t|\\t\")\n",
    "        record = {\n",
    "            \"tax_id\": fields[0],\n",
    "            \"name_txt\": fields[1],\n",
    "            \"unique_name\": fields[2],\n",
    "            \"name_class\": fields[3],\n",
    "        }\n",
    "        records.append(record)\n",
    "names_df = pl.from_records(records).with_columns(\n",
    "    pl.col(\"tax_id\").cast(pl.Int64),\n",
    "    \n",
    ")\n",
    "names_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4750ccfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tax_id: 2231794\n",
      "name_txt: Massjukichlorella epiphytica\n",
      "unique_name: \n",
      "name_class: scientific name\n"
     ]
    }
   ],
   "source": [
    "test = names_df.sample(1) \n",
    "for col_name in names_df.columns:\n",
    "    print(f\"{col_name}: {test[col_name].first()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "edf3edfb",
   "metadata": {},
   "source": [
    "now let's form the lineage strings\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "02001c3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sci_name_df = names_df.filter(pl.col(\"name_class\") == \"scientific name\").select(\n",
    "    [\"tax_id\", \"name_txt\"]  ).rename({\"name_txt\": \"scientific_name\"}).unique()\n",
    "# sci_name_df.null_count()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "073c58f3",
   "metadata": {},
   "source": [
    "non tree accelrated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "64d20dcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building lineages for 2708826 tax_ids...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "655af61c12c84ae4bb5cc73a51d1ac18",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2708826 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Built 2708826 lineage strings\n",
      "\n",
      "Found 2708825 non-empty lineages\n",
      "\n",
      "Example lineages:\n",
      "shape: (10, 2)\n",
      "┌────────┬─────────────────────────────────┐\n",
      "│ tax_id ┆ lineage                         │\n",
      "│ ---    ┆ ---                             │\n",
      "│ i64    ┆ str                             │\n",
      "╞════════╪═════════════════════════════════╡\n",
      "│ 2      ┆ cellular organisms;Bacteria     │\n",
      "│ 6      ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 7      ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 9      ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 10     ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 11     ┆ cellular organisms;Bacteria;Ba… │\n",
      "│ 13     ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 14     ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 16     ┆ cellular organisms;Bacteria;Ps… │\n",
      "│ 17     ┆ cellular organisms;Bacteria;Ps… │\n",
      "└────────┴─────────────────────────────────┘\n"
     ]
    }
   ],
   "source": [
    "def build_all_lineages(nodes_df, sci_name_df):\n",
    "    \"\"\"Build lineage strings for all tax_ids efficiently using polars operations\"\"\"\n",
    "    \n",
    "    # Join nodes with scientific names to get name for each tax_id\n",
    "    nodes_with_names = nodes_df.join(\n",
    "        sci_name_df, \n",
    "        on=\"tax_id\", \n",
    "        how=\"left\"\n",
    "    ).select([\"tax_id\", \"parent_tax_id\", \"scientific_name\"])\n",
    "    \n",
    "    # Build a dictionary for faster lookups\n",
    "    tax_to_parent = dict(zip(\n",
    "        nodes_with_names[\"tax_id\"].to_list(),\n",
    "        nodes_with_names[\"parent_tax_id\"].to_list()\n",
    "    ))\n",
    "    tax_to_name = dict(zip(\n",
    "        nodes_with_names[\"tax_id\"].to_list(),\n",
    "        nodes_with_names[\"scientific_name\"].to_list()\n",
    "    ))\n",
    "    \n",
    "    def get_lineage(tax_id):\n",
    "        \"\"\"Get lineage for a single tax_id\"\"\"\n",
    "        if tax_id not in tax_to_name or tax_to_name[tax_id] is None:\n",
    "            return \"\"\n",
    "            \n",
    "        lineage = []\n",
    "        current = tax_id\n",
    "        visited = set()\n",
    "        \n",
    "        while current in tax_to_parent and current not in visited and current != 1:\n",
    "            visited.add(current)\n",
    "            name = tax_to_name.get(current)\n",
    "            if name:\n",
    "                lineage.append(name)\n",
    "            \n",
    "            parent = tax_to_parent[current]\n",
    "            if parent == current:  # Root case\n",
    "                break\n",
    "            current = parent\n",
    "        \n",
    "        return \";\".join(reversed(lineage))\n",
    "    \n",
    "    # Get all unique tax_ids\n",
    "    all_tax_ids = sci_name_df[\"tax_id\"].unique().sort().to_list()\n",
    "    print(f\"Building lineages for {len(all_tax_ids)} tax_ids...\")\n",
    "    \n",
    "    # Build lineages\n",
    "    lineages = []\n",
    "    for tax_id in tqdm(all_tax_ids):\n",
    "        lineage = get_lineage(tax_id)\n",
    "        lineages.append(lineage)\n",
    "    \n",
    "    # Create result DataFrame\n",
    "    result_df = pl.DataFrame({\n",
    "        \"tax_id\": all_tax_ids,\n",
    "        \"lineage\": lineages\n",
    "    })\n",
    "    \n",
    "    return result_df\n",
    "\n",
    "# Build all lineages\n",
    "lineage_df = build_all_lineages(nodes_df, sci_name_df)\n",
    "print(f\"Built {lineage_df.height} lineage strings\")\n",
    "\n",
    "# Show some examples with non-empty lineages\n",
    "non_empty_lineages = lineage_df.filter(pl.col(\"lineage\") != \"\")\n",
    "print(f\"\\nFound {non_empty_lineages.height} non-empty lineages\")\n",
    "print(\"\\nExample lineages:\")\n",
    "print(non_empty_lineages.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "fa6c5734",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating TaxonomyTree with NCBI nodes...\n"
     ]
    }
   ],
   "source": [
    "# Create taxonomy tree\n",
    "print(\"Creating TaxonomyTree with NCBI nodes...\")\n",
    "\n",
    "# Merge nodes with scientific names for the tree\n",
    "nodes_with_names = nodes_df.join(\n",
    "    sci_name_df, \n",
    "    on=\"tax_id\", \n",
    "    how=\"left\"\n",
    ").select([\"tax_id\", \"parent_tax_id\", \"rank\", \"scientific_name\"])\n",
    "\n",
    "# Create tree\n",
    "tax_tree = TaxonomyTree(nodes_with_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "1e1b6f00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16885"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mini_genebank[\"total_gene_count\"].null_count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "58300525",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (242_839, 41)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>assembly_accession</th><th>bioproject</th><th>biosample</th><th>wgs_master</th><th>refseq_category</th><th>ncbi_taxonid</th><th>species_taxid</th><th>organism_name</th><th>infraspecific_name</th><th>isolate</th><th>version_status</th><th>assembly_level</th><th>release_type</th><th>genome_rep</th><th>seq_rel_date</th><th>asm_name</th><th>asm_submitter</th><th>gbrs_paired_asm</th><th>paired_asm_comp</th><th>ftp_path</th><th>excluded_from_refseq</th><th>relation_to_type_material</th><th>asm_not_live_date</th><th>assembly_type</th><th>group</th><th>genome_size</th><th>genome_size_ungapped</th><th>gc_percent</th><th>replicon_count</th><th>scaffold_count</th><th>contig_count</th><th>annotation_provider</th><th>annotation_name</th><th>annotation_date</th><th>total_gene_count</th><th>protein_coding_gene_count</th><th>non_coding_gene_count</th><th>pubmed_id</th><th>md5sums_path</th><th>has_genes</th><th>has_proteins</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>f64</td><td>i64</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>i64</td><td>i64</td><td>str</td><td>str</td><td>bool</td><td>i32</td></tr></thead><tbody><tr><td>&quot;GCA_038318695.1&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>1949030</td><td>2955465</td><td>&quot;Influenza B virus (B/Illinois/…</td><td>&quot;strain=B/Illinois/22/2016&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Complete Genome&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2017-02-06&quot;</td><td>&quot;ASM3831869v1&quot;</td><td>&quot;WHO Collaborating Center for S…</td><td>null</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;viral&quot;</td><td>14398</td><td>14398</td><td>40.0</td><td>8</td><td>8</td><td>8</td><td>&quot;WHO Collaborating Center for S…</td><td>&quot;Annotation submitted by WHO Co…</td><td>&quot;2017-02-06&quot;</td><td>11</td><td>11</td><td>0</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_041200105.1&quot;</td><td>&quot;PRJNA1134181&quot;</td><td>&quot;SAMN42395754&quot;</td><td>null</td><td>null</td><td>3234140</td><td>3234140</td><td>&quot;Flavobacterium sp. WC2429&quot;</td><td>&quot;strain=WC2429&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Chromosome&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2024-08-11&quot;</td><td>&quot;ASM4120010v1&quot;</td><td>&quot;Wellesley College&quot;</td><td>&quot;GCF_041200105.1&quot;</td><td>&quot;identical&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>4043412</td><td>4043312</td><td>33.0</td><td>1</td><td>1</td><td>1</td><td>&quot;NCBI&quot;</td><td>&quot;NCBI Prokaryotic Genome Annota…</td><td>&quot;2024-08-01&quot;</td><td>3466</td><td>3375</td><td>78</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_007989705.1&quot;</td><td>&quot;PRJDB3962&quot;</td><td>&quot;SAMD00166011&quot;</td><td>&quot;BJUG00000000.1&quot;</td><td>null</td><td>417368</td><td>417368</td><td>&quot;Enterococcus thailandicus&quot;</td><td>&quot;strain=NBRC 101867&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Contig&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2019-08-01&quot;</td><td>&quot;ASM798970v1&quot;</td><td>&quot;National Institute of Technolo…</td><td>&quot;GCF_007989705.1&quot;</td><td>&quot;identical&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>&quot;assembly from type material&quot;</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>2783774</td><td>2783774</td><td>36.5</td><td>0</td><td>36</td><td>36</td><td>&quot;National Institute of Technolo…</td><td>&quot;Annotation submitted by Nation…</td><td>&quot;2023-08-24&quot;</td><td>2641</td><td>2597</td><td>44</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_001996065.1&quot;</td><td>&quot;PRJNA368798&quot;</td><td>&quot;SAMN06272993&quot;</td><td>&quot;MUFG00000000.1&quot;</td><td>null</td><td>1909453</td><td>1909453</td><td>&quot;Salinivibrio sp. MA351&quot;</td><td>&quot;strain=MA351&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Contig&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2017-02-14&quot;</td><td>&quot;ASM199606v1&quot;</td><td>&quot;Universidad de Sevilla&quot;</td><td>&quot;GCF_001996065.1&quot;</td><td>&quot;identical&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>3308854</td><td>3308854</td><td>49.5</td><td>0</td><td>59</td><td>59</td><td>&quot;NCBI&quot;</td><td>&quot;NCBI Prokaryotic Genome Annota…</td><td>&quot;2017-01-30&quot;</td><td>3058</td><td>2931</td><td>93</td><td>&quot;28684561&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_014163945.1&quot;</td><td>&quot;PRJNA648432&quot;</td><td>&quot;SAMN15642170&quot;</td><td>&quot;JACJEG000000000.1&quot;</td><td>null</td><td>2760948</td><td>2760948</td><td>&quot;Pseudoalteromonas sp. SR41-6&quot;</td><td>&quot;strain=SR41-6&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Scaffold&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2020-08-11&quot;</td><td>&quot;ASM1416394v1&quot;</td><td>&quot;University of Alaska Fairbanks&quot;</td><td>&quot;GCF_014163945.1&quot;</td><td>&quot;identical&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>5243740</td><td>5242440</td><td>40.5</td><td>0</td><td>284</td><td>297</td><td>&quot;NCBI&quot;</td><td>&quot;NCBI Prokaryotic Genome Annota…</td><td>&quot;2020-08-06&quot;</td><td>4775</td><td>4563</td><td>121</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;GCA_013107245.1&quot;</td><td>&quot;PRJNA596592&quot;</td><td>&quot;SAMN13631914&quot;</td><td>&quot;WVJT00000000.1&quot;</td><td>null</td><td>2690675</td><td>2690675</td><td>&quot;Roseobacter sp. HKCCD9117-2&quot;</td><td>&quot;strain=HKCCD9117-2&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Contig&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2020-05-15&quot;</td><td>&quot;ASM1310724v1&quot;</td><td>&quot;Chinese University of Hong Kon…</td><td>&quot;GCF_013107245.1&quot;</td><td>&quot;identical&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>4412902</td><td>4412902</td><td>60.0</td><td>0</td><td>76</td><td>76</td><td>&quot;NCBI&quot;</td><td>&quot;NCBI Prokaryotic Genome Annota…</td><td>&quot;2020-01-04&quot;</td><td>4404</td><td>4269</td><td>49</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_015107295.1&quot;</td><td>&quot;PRJNA550295&quot;</td><td>&quot;SAMN12123845&quot;</td><td>&quot;VVIH00000000.1&quot;</td><td>&quot;reference genome&quot;</td><td>452217</td><td>452217</td><td>&quot;Plecodus paradoxus&quot;</td><td>null</td><td>&quot;JDI4&quot;</td><td>&quot;latest&quot;</td><td>&quot;Scaffold&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2020-10-29&quot;</td><td>&quot;JDI4.scf&quot;</td><td>&quot;&quot;CICHLID~X&quot;&quot;</td><td>null</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;vertebrate_other&quot;</td><td>706454054</td><td>704756733</td><td>40.5</td><td>0</td><td>53948</td><td>119797</td><td>null</td><td>null</td><td>&quot;&quot;</td><td>0</td><td>0</td><td>0</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>false</td><td>0</td></tr><tr><td>&quot;GCA_038669045.1&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>1389833</td><td>2955291</td><td>&quot;Influenza A virus (A/Helsinki/…</td><td>&quot;strain=A/Helsinki/39/2013&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Complete Genome&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2013-09-16&quot;</td><td>&quot;ASM3866904v1&quot;</td><td>&quot;FIMM, University of Helsinki&quot;</td><td>null</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;viral&quot;</td><td>13417</td><td>13417</td><td>43.0</td><td>8</td><td>8</td><td>8</td><td>&quot;FIMM, University of Helsinki&quot;</td><td>&quot;Annotation submitted by FIMM, …</td><td>&quot;2014-03-06&quot;</td><td>12</td><td>11</td><td>0</td><td>&quot;24435855&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_000540015.1&quot;</td><td>&quot;PRJNA223818&quot;</td><td>&quot;SAMN02381559&quot;</td><td>&quot;JBQP00000000.1&quot;</td><td>null</td><td>1409498</td><td>1280</td><td>&quot;Staphylococcus aureus SJOS6027&quot;</td><td>&quot;strain=SJOS6027&quot;</td><td>null</td><td>&quot;latest&quot;</td><td>&quot;Scaffold&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2014-02-04&quot;</td><td>&quot;Stap_aure_SJOS6027_V1&quot;</td><td>&quot;Broad Institute&quot;</td><td>&quot;GCF_000540015.1&quot;</td><td>&quot;identical&quot;</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>null</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>2840434</td><td>2840334</td><td>33.0</td><td>0</td><td>36</td><td>37</td><td>&quot;Broad Institute&quot;</td><td>&quot;Annotation submitted by Broad …</td><td>&quot;2014-02-04&quot;</td><td>2727</td><td>2653</td><td>73</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>true</td><td>1</td></tr><tr><td>&quot;GCA_949504075.1&quot;</td><td>&quot;PRJEB32890&quot;</td><td>&quot;SAMEA112714571&quot;</td><td>&quot;CATBKG000000000.1&quot;</td><td>null</td><td>2320104</td><td>2320104</td><td>&quot;bacterium D16-54&quot;</td><td>null</td><td>&quot;G1-6A_111220.47_1677754016&quot;</td><td>&quot;latest&quot;</td><td>&quot;Contig&quot;</td><td>&quot;Major&quot;</td><td>&quot;Full&quot;</td><td>&quot;2023-04-04&quot;</td><td>&quot;G1-6A_111220.47_1677754016&quot;</td><td>&quot;HELMHOLTZ CENTRE FOR INFECTION…</td><td>null</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>&quot;derived from metagenome; genus…</td><td>null</td><td>null</td><td>&quot;haploid&quot;</td><td>&quot;bacteria&quot;</td><td>2599499</td><td>2599499</td><td>48.5</td><td>0</td><td>236</td><td>236</td><td>null</td><td>null</td><td>&quot;&quot;</td><td>0</td><td>0</td><td>0</td><td>null</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td><td>false</td><td>0</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (242_839, 41)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ assembly_ ┆ bioprojec ┆ biosample ┆ wgs_maste ┆ … ┆ pubmed_id ┆ md5sums_p ┆ has_genes ┆ has_prot │\n",
       "│ accession ┆ t         ┆ ---       ┆ r         ┆   ┆ ---       ┆ ath       ┆ ---       ┆ eins     │\n",
       "│ ---       ┆ ---       ┆ str       ┆ ---       ┆   ┆ str       ┆ ---       ┆ bool      ┆ ---      │\n",
       "│ str       ┆ str       ┆           ┆ str       ┆   ┆           ┆ str       ┆           ┆ i32      │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│ GCA_03831 ┆ null      ┆ null      ┆ null      ┆ … ┆ null      ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 8695.1    ┆           ┆           ┆           ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_04120 ┆ PRJNA1134 ┆ SAMN42395 ┆ null      ┆ … ┆ null      ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 0105.1    ┆ 181       ┆ 754       ┆           ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_00798 ┆ PRJDB3962 ┆ SAMD00166 ┆ BJUG00000 ┆ … ┆ null      ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 9705.1    ┆           ┆ 011       ┆ 000.1     ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_00199 ┆ PRJNA3687 ┆ SAMN06272 ┆ MUFG00000 ┆ … ┆ 28684561  ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 6065.1    ┆ 98        ┆ 993       ┆ 000.1     ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_01416 ┆ PRJNA6484 ┆ SAMN15642 ┆ JACJEG000 ┆ … ┆ null      ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 3945.1    ┆ 32        ┆ 170       ┆ 000000.1  ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │\n",
       "│ GCA_01310 ┆ PRJNA5965 ┆ SAMN13631 ┆ WVJT00000 ┆ … ┆ null      ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 7245.1    ┆ 92        ┆ 914       ┆ 000.1     ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_01510 ┆ PRJNA5502 ┆ SAMN12123 ┆ VVIH00000 ┆ … ┆ null      ┆ https://f ┆ false     ┆ 0        │\n",
       "│ 7295.1    ┆ 95        ┆ 845       ┆ 000.1     ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_03866 ┆ null      ┆ null      ┆ null      ┆ … ┆ 24435855  ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 9045.1    ┆           ┆           ┆           ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_00054 ┆ PRJNA2238 ┆ SAMN02381 ┆ JBQP00000 ┆ … ┆ null      ┆ https://f ┆ true      ┆ 1        │\n",
       "│ 0015.1    ┆ 18        ┆ 559       ┆ 000.1     ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "│ GCA_94950 ┆ PRJEB3289 ┆ SAMEA1127 ┆ CATBKG000 ┆ … ┆ null      ┆ https://f ┆ false     ┆ 0        │\n",
       "│ 4075.1    ┆ 0         ┆ 14571     ┆ 000000.1  ┆   ┆           ┆ tp.ncbi.n ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ lm.nih.go ┆           ┆          │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ v/g…      ┆           ┆          │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Set data availability based on your GenBank assembly data\n",
    "# Filter to only include entries with actual genome/transcriptome data\n",
    "\n",
    "data_available_df = mini_genebank.with_columns(\n",
    "    pl.when(\n",
    "        pl.col(\"total_gene_count\") > 0).then(pl.lit(True).alias(\"has_genes\")).otherwise(pl.lit(False).alias(\"has_genes\")),\n",
    "    pl.when(pl.col(\"protein_coding_gene_count\") > 0).then(pl.lit(1).alias(\"has_proteins\")).otherwise(pl.lit(0).alias(\"has_proteins\")),\n",
    ")\n",
    "            \n",
    "data_available_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "6bd569f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (165_650, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>tax_id</th><th>ftp_path</th></tr><tr><td>i64</td><td>str</td></tr></thead><tbody><tr><td>1949030</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>3234140</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>417368</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>1909453</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>2760948</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>1964747</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>3156266</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>2690675</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>1389833</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr><tr><td>1409498</td><td>&quot;https://ftp.ncbi.nlm.nih.gov/g…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (165_650, 2)\n",
       "┌─────────┬─────────────────────────────────┐\n",
       "│ tax_id  ┆ ftp_path                        │\n",
       "│ ---     ┆ ---                             │\n",
       "│ i64     ┆ str                             │\n",
       "╞═════════╪═════════════════════════════════╡\n",
       "│ 1949030 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 3234140 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 417368  ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 1909453 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 2760948 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ …       ┆ …                               │\n",
       "│ 1964747 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 3156266 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 2690675 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 1389833 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "│ 1409498 ┆ https://ftp.ncbi.nlm.nih.gov/g… │\n",
       "└─────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_available_df = data_available_df.filter(pl.col(\"has_genes\")).select([\"ncbi_taxonid\",\"ftp_path\"]).rename({\"ncbi_taxonid\": \"tax_id\"})\n",
    "data_available_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "4d22c6aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 165650 tax_ids with available assembly data\n",
      "Set data availability for 165650 tax_ids\n",
      "TaxonomyTree ready for neighbor finding!\n"
     ]
    }
   ],
   "source": [
    "print(f\"Found {data_available_df.height} tax_ids with available assembly data\")\n",
    "tax_tree._set_data_availability(data_df=data_available_df)\n",
    "\n",
    "print(\"TaxonomyTree ready for neighbor finding!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "fd5f63af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing with tax_ids: [0, 7, 9, 17, 21, 23, 24, 25, 33, 34, 35, 38, 41, 43, 45, 48, 51, 52, 54, 56, 61, 63, 65, 69, 83, 84, 86, 89, 94, 100, 102, 103, 104, 106, 108, 110, 114, 120, 122, 124, 128, 134, 139, 144, 148, 149, 150, 151, 152, 156]\n",
      "\n",
      "--- Tax ID 0 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 7 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Azorhizobium caulinodans ORS 571 (tax_id: 438753, distance: 1)\n",
      "\n",
      "--- Tax ID 9 ---\n",
      "✅ 9 has data directly!\n",
      "\n",
      "--- Tax ID 17 ---\n",
      "✅ 17 has data directly!\n",
      "\n",
      "--- Tax ID 21 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 23 ---\n",
      "✅ 23 has data directly!\n",
      "\n",
      "--- Tax ID 24 ---\n",
      "✅ 24 has data directly!\n",
      "\n",
      "--- Tax ID 25 ---\n",
      "✅ 25 has data directly!\n",
      "\n",
      "--- Tax ID 33 ---\n",
      "✅ 33 has data directly!\n",
      "\n",
      "--- Tax ID 34 ---\n",
      "✅ 34 has data directly!\n",
      "\n",
      "--- Tax ID 35 ---\n",
      "✅ 35 has data directly!\n",
      "\n",
      "--- Tax ID 38 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 41 ---\n",
      "✅ 41 has data directly!\n",
      "\n",
      "--- Tax ID 43 ---\n",
      "✅ 43 has data directly!\n",
      "\n",
      "--- Tax ID 45 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 48 ---\n",
      "✅ 48 has data directly!\n",
      "\n",
      "--- Tax ID 51 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Chondromyces apiculatus DSM 436 (tax_id: 1192034, distance: 1)\n",
      "\n",
      "--- Tax ID 52 ---\n",
      "✅ 52 has data directly!\n",
      "\n",
      "--- Tax ID 54 ---\n",
      "✅ 54 has data directly!\n",
      "\n",
      "--- Tax ID 56 ---\n",
      "✅ 56 has data directly!\n",
      "\n",
      "--- Tax ID 61 ---\n",
      "✅ 61 has data directly!\n",
      "\n",
      "--- Tax ID 63 ---\n",
      "✅ 63 has data directly!\n",
      "\n",
      "--- Tax ID 65 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Herpetosiphon aurantiacus DSM 785 (tax_id: 316274, distance: 1)\n",
      "\n",
      "--- Tax ID 69 ---\n",
      "✅ 69 has data directly!\n",
      "\n",
      "--- Tax ID 83 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 84 ---\n",
      "✅ 84 has data directly!\n",
      "\n",
      "--- Tax ID 86 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Hyphomonas jannaschiana VP2 (tax_id: 1280952, distance: 1)\n",
      "\n",
      "--- Tax ID 89 ---\n",
      "✅ 89 has data directly!\n",
      "\n",
      "--- Tax ID 94 ---\n",
      "✅ 94 has data directly!\n",
      "\n",
      "--- Tax ID 100 ---\n",
      "✅ 100 has data directly!\n",
      "\n",
      "--- Tax ID 102 ---\n",
      "✅ 102 has data directly!\n",
      "\n",
      "--- Tax ID 103 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 104 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Cyclobacterium marinum DSM 745 (tax_id: 880070, distance: 1)\n",
      "\n",
      "--- Tax ID 106 ---\n",
      "✅ 106 has data directly!\n",
      "\n",
      "--- Tax ID 108 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Spirosoma linguale DSM 74 (tax_id: 504472, distance: 1)\n",
      "\n",
      "--- Tax ID 110 ---\n",
      "✅ 110 has data directly!\n",
      "\n",
      "--- Tax ID 114 ---\n",
      "✅ 114 has data directly!\n",
      "\n",
      "--- Tax ID 120 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Planctopirus limnophila DSM 3776 (tax_id: 521674, distance: 1)\n",
      "\n",
      "--- Tax ID 122 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Gimesia maris DSM 8797 (tax_id: 344747, distance: 1)\n",
      "\n",
      "--- Tax ID 124 ---\n",
      "✅ 124 has data directly!\n",
      "\n",
      "--- Tax ID 128 ---\n",
      "🔽 Descendants with data (1):\n",
      "   - Isosphaera pallida ATCC 43644 (tax_id: 575540, distance: 1)\n",
      "\n",
      "--- Tax ID 134 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 139 ---\n",
      "✅ 139 has data directly!\n",
      "\n",
      "--- Tax ID 144 ---\n",
      "✅ 144 has data directly!\n",
      "\n",
      "--- Tax ID 148 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 149 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 150 ---\n",
      "✅ 150 has data directly!\n",
      "\n",
      "--- Tax ID 151 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 152 ---\n",
      "❌ No relatives with data found within distance 5\n",
      "\n",
      "--- Tax ID 156 ---\n",
      "✅ 156 has data directly!\n",
      "\n",
      "Tree structure ready for production use!\n"
     ]
    }
   ],
   "source": [
    "# Test the tree with some example queries\n",
    "# Get some test tax_ids from your rRNA data that might not have assemblies\n",
    "test_tax_ids = rrna_df[\"taxid\"].cast(pl.Int64).unique().limit(50).to_list()\n",
    "print(f\"Testing with tax_ids: {test_tax_ids}\")\n",
    "\n",
    "for tax_id in test_tax_ids:\n",
    "    print(f\"\\n--- Tax ID {tax_id} ---\")\n",
    "    try:\n",
    "        result = tax_tree.find_nearest_with_data(tax_id, max_distance=5)\n",
    "        \n",
    "        if result['self_has_data']:\n",
    "            print(f\"✅ {tax_id} has data directly!\")\n",
    "        else:\n",
    "            if result['ancestor']:\n",
    "                ancestor = result['ancestor']\n",
    "                print(f\"🔼 Nearest ancestor with data: {ancestor['name']} (tax_id: {ancestor['tax_id']}, distance: {ancestor['distance']})\")\n",
    "            \n",
    "            if result['descendants']:\n",
    "                print(f\"🔽 Descendants with data ({len(result['descendants'])}):\")\n",
    "                for desc in result['descendants'][:3]:  # Show first 3\n",
    "                    print(f\"   - {desc['name']} (tax_id: {desc['tax_id']}, distance: {desc['distance']})\")\n",
    "            \n",
    "            if not result['ancestor'] and not result['descendants']:\n",
    "                print(\"❌ No relatives with data found within distance 5\")\n",
    "                \n",
    "    except Exception as e:\n",
    "        print(f\"❌ Error processing tax_id {tax_id}: {e}\")\n",
    "\n",
    "print(\"\\nTree structure ready for production use!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "c191190e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ TaxonomyTree module reloaded with enhanced features!\n"
     ]
    }
   ],
   "source": [
    "# Reload the updated TaxonomyTree module\n",
    "import sys\n",
    "\n",
    "# Remove cached module\n",
    "if 'rolypoly.utils.bio.trees' in sys.modules:\n",
    "    del sys.modules['rolypoly.utils.bio.trees']\n",
    "    \n",
    "\n",
    "# Reimport\n",
    "from rolypoly.utils.bio.trees import TaxonomyTree\n",
    "\n",
    "print(\"✅ TaxonomyTree module reloaded with enhanced features!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "b261f197",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating enhanced TaxonomyTree with priority columns...\n",
      "Data available for 165650 tax_ids with metadata\n",
      "Built tree with 2708826 nodes, 2457509 leaf nodes\n",
      "Set data availability for 165650 tax_ids\n",
      "Stored metadata for 165650 tax_ids\n",
      "\n",
      "✅ Enhanced tree created!\n",
      "Built tree with 2708826 nodes, 2457509 leaf nodes\n",
      "Set data availability for 165650 tax_ids\n",
      "Stored metadata for 165650 tax_ids\n",
      "\n",
      "✅ Enhanced tree created!\n"
     ]
    }
   ],
   "source": [
    "# Recreate the tree with priority columns specified\n",
    "print(\"Creating enhanced TaxonomyTree with priority columns...\")\n",
    "\n",
    "# Prepare data with full metadata\n",
    "data_available_enhanced = mini_genebank.filter(\n",
    "    pl.col(\"total_gene_count\") > 0\n",
    ").select([\n",
    "    pl.col(\"ncbi_taxonid\").alias(\"tax_id\"),\n",
    "    \"ftp_path\",\n",
    "    \"protein_coding_gene_count\",\n",
    "    \"total_gene_count\",\n",
    "    \"genome_size\",\n",
    "    \"assembly_level\",\n",
    "    \"organism_name\"\n",
    "])\n",
    "\n",
    "print(f\"Data available for {data_available_enhanced.height} tax_ids with metadata\")\n",
    "\n",
    "# Create tree with priority columns\n",
    "tax_tree_enhanced = TaxonomyTree(\n",
    "    nodes_with_names,\n",
    "    data_availability_df=data_available_enhanced,\n",
    "    priority_columns=['protein_coding_gene_count', 'genome_size']  # Prioritize by gene count, then genome size\n",
    ")\n",
    "\n",
    "print(\"\\n✅ Enhanced tree created!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b00c0679",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "FEATURE 1: COMPREHENSIVE TREE STATISTICS\n",
      "======================================================================\n",
      "============================================================\n",
      "TAXONOMY TREE STATISTICS\n",
      "============================================================\n",
      "\n",
      "Tree Structure:\n",
      "  Total nodes:          2,708,826\n",
      "  Leaf nodes:           2,457,509\n",
      "  Internal nodes:       251,317\n",
      "\n",
      "Data Availability:\n",
      "  Nodes with data:      165,650\n",
      "  Leaves with data:     161,573\n",
      "  Leaves without data:  2,295,936\n",
      "  Internal nodes w/data:4,077\n",
      "\n",
      "Coverage:\n",
      "  All nodes:            6.12%\n",
      "  Leaf nodes only:      6.57%\n",
      "\n",
      "Cache:\n",
      "  Cached queries:       0\n",
      "\n",
      "Priority columns: protein_coding_gene_count, genome_size\n",
      "\n",
      "Top ranks by count:\n",
      "  species             : 2,223,790 total (83,775 with data,   3.8%)\n",
      "  no rank             :  253,736 total (58,417 with data,  23.0%)\n",
      "  genus               :  113,698 total (     0 with data,   0.0%)\n",
      "  strain              :   46,928 total (22,400 with data,  47.7%)\n",
      "  subspecies          :   30,788 total (   404 with data,   1.3%)\n",
      "  family              :   10,817 total (     0 with data,   0.0%)\n",
      "  varietas            :   10,355 total (    20 with data,   0.2%)\n",
      "  subfamily           :    3,349 total (     0 with data,   0.0%)\n",
      "  tribe               :    2,423 total (     0 with data,   0.0%)\n",
      "  order               :    2,069 total (     0 with data,   0.0%)\n",
      "============================================================\n",
      "============================================================\n",
      "TAXONOMY TREE STATISTICS\n",
      "============================================================\n",
      "\n",
      "Tree Structure:\n",
      "  Total nodes:          2,708,826\n",
      "  Leaf nodes:           2,457,509\n",
      "  Internal nodes:       251,317\n",
      "\n",
      "Data Availability:\n",
      "  Nodes with data:      165,650\n",
      "  Leaves with data:     161,573\n",
      "  Leaves without data:  2,295,936\n",
      "  Internal nodes w/data:4,077\n",
      "\n",
      "Coverage:\n",
      "  All nodes:            6.12%\n",
      "  Leaf nodes only:      6.57%\n",
      "\n",
      "Cache:\n",
      "  Cached queries:       0\n",
      "\n",
      "Priority columns: protein_coding_gene_count, genome_size\n",
      "\n",
      "Top ranks by count:\n",
      "  species             : 2,223,790 total (83,775 with data,   3.8%)\n",
      "  no rank             :  253,736 total (58,417 with data,  23.0%)\n",
      "  genus               :  113,698 total (     0 with data,   0.0%)\n",
      "  strain              :   46,928 total (22,400 with data,  47.7%)\n",
      "  subspecies          :   30,788 total (   404 with data,   1.3%)\n",
      "  family              :   10,817 total (     0 with data,   0.0%)\n",
      "  varietas            :   10,355 total (    20 with data,   0.2%)\n",
      "  subfamily           :    3,349 total (     0 with data,   0.0%)\n",
      "  tribe               :    2,423 total (     0 with data,   0.0%)\n",
      "  order               :    2,069 total (     0 with data,   0.0%)\n",
      "============================================================\n"
     ]
    }
   ],
   "source": [
    "# Feature 1: Get comprehensive tree statistics\n",
    "print(\"=\"*70)\n",
    "print(\"FEATURE 1: COMPREHENSIVE TREE STATISTICS\")\n",
    "print(\"=\"*70)\n",
    "\n",
    "tax_tree_enhanced.print_stats()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "83e42aee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======================================================================\n",
      "FEATURE 2: BATCH QUERIES WITH STATISTICS\n",
      "======================================================================\n",
      "\n",
      "Querying 100 tax_ids from rRNA data...\n",
      "\n",
      "📊 Batch Query Statistics:\n",
      "  Total queried:        100\n",
      "  Self has data:        19\n",
      "  Ancestor found:       0\n",
      "  Descendant found:     2\n",
      "  No match found:       79\n",
      "\n",
      "📏 Distance Distribution:\n",
      "  Distance 0: 19 matches\n",
      "  Distance 1: 2 matches\n",
      "\n",
      "❌ Unmatched tax_ids (first 10): [413816, 244960, 870734, 304739, 1973154, 1769086, 979731, 573973, 639593, 297705]\n"
     ]
    }
   ],
   "source": [
    "# Feature 2: Batch query for multiple tax_ids with statistics\n",
    "print(\"\\n\" + \"=\"*70)\n",
    "print(\"FEATURE 2: BATCH QUERIES WITH STATISTICS\")\n",
    "print(\"=\"*70)\n",
    "\n",
    "# Get a sample of tax_ids from rrna_df\n",
    "sample_tax_ids = rrna_df[\"taxid\"].cast(pl.Int64).unique().sample(100).to_list()\n",
    "\n",
    "print(f\"\\nQuerying {len(sample_tax_ids)} tax_ids from rRNA data...\")\n",
    "\n",
    "# Batch query with stats\n",
    "results, stats = tax_tree_enhanced.find_nearest_with_data_batch(\n",
    "    tax_ids=sample_tax_ids,\n",
    "    max_distance=10,\n",
    "    leaves_only=False,  # Include all nodes\n",
    "    return_stats=True\n",
    ")\n",
    "\n",
    "print(\"\\n📊 Batch Query Statistics:\")\n",
    "print(f\"  Total queried:        {stats['total_queried']:,}\")\n",
    "print(f\"  Self has data:        {stats['self_has_data']:,}\")\n",
    "print(f\"  Ancestor found:       {stats['ancestor_found']:,}\")\n",
    "print(f\"  Descendant found:     {stats['descendant_found']:,}\")\n",
    "print(f\"  No match found:       {stats['no_match_found']:,}\")\n",
    "\n",
    "print(\"\\n📏 Distance Distribution:\")\n",
    "for distance in sorted(stats['distance_distribution'].keys())[:10]:\n",
    "    count = stats['distance_distribution'][distance]\n",
    "    print(f\"  Distance {distance}: {count:,} matches\")\n",
    "\n",
    "if stats['unmatched_tax_ids']:\n",
    "    print(f\"\\n❌ Unmatched tax_ids (first 10): {stats['unmatched_tax_ids'][:10]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "103d9fe0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======================================================================\n",
      "FEATURE 3: PRIORITY-BASED NEIGHBOR SELECTION\n",
      "======================================================================\n",
      "\n",
      "Testing with tax_id: 413816\n",
      "Name: Halorubrum halophilum\n",
      "\n",
      "🔍 Neighbors with DEFAULT priority (protein_coding_gene_count, genome_size):\n",
      "\n",
      "🔍 Neighbors with OVERRIDE priority (genome_size, protein_coding_gene_count):\n"
     ]
    }
   ],
   "source": [
    "# Feature 3: Priority-based neighbor selection\n",
    "print(\"\\n\" + \"=\"*70)\n",
    "print(\"FEATURE 3: PRIORITY-BASED NEIGHBOR SELECTION\")\n",
    "print(\"=\"*70)\n",
    "\n",
    "# Find a tax_id without data to demonstrate priority selection\n",
    "test_tax_id = None\n",
    "for tid in sample_tax_ids[:50]:\n",
    "    if tid not in tax_tree_enhanced.data_available and tid in tax_tree_enhanced.parent_map:\n",
    "        test_tax_id = tid\n",
    "        break\n",
    "\n",
    "if test_tax_id:\n",
    "    print(f\"\\nTesting with tax_id: {test_tax_id}\")\n",
    "    print(f\"Name: {tax_tree_enhanced.name_map.get(test_tax_id, 'Unknown')}\")\n",
    "    \n",
    "    # Find neighbors with default priority\n",
    "    result_default = tax_tree_enhanced.find_nearest_with_data(test_tax_id, max_distance=5)\n",
    "    \n",
    "    print(\"\\n🔍 Neighbors with DEFAULT priority (protein_coding_gene_count, genome_size):\")\n",
    "    if result_default['ancestor']:\n",
    "        anc = result_default['ancestor']\n",
    "        meta = anc.get('metadata', {})\n",
    "        print(f\"\\n  Ancestor: {anc['name']} (distance: {anc['distance']})\")\n",
    "        print(f\"    Protein genes: {meta.get('protein_coding_gene_count', 'N/A'):,}\")\n",
    "        print(f\"    Genome size:   {meta.get('genome_size', 'N/A'):,}\")\n",
    "    \n",
    "    if result_default['descendants']:\n",
    "        print(f\"\\n  Descendants ({len(result_default['descendants'])} found):\")\n",
    "        for desc in result_default['descendants'][:3]:\n",
    "            meta = desc.get('metadata', {})\n",
    "            print(f\"    {desc['name']} (distance: {desc['distance']})\")\n",
    "            print(f\"      Protein genes: {meta.get('protein_coding_gene_count', 'N/A'):,}\")\n",
    "            print(f\"      Genome size:   {meta.get('genome_size', 'N/A'):,}\")\n",
    "    \n",
    "    # Override priority to prioritize genome_size over gene count\n",
    "    result_override = tax_tree_enhanced.find_nearest_with_data(\n",
    "        test_tax_id, \n",
    "        max_distance=5, \n",
    "        priority_override=['genome_size', 'protein_coding_gene_count']\n",
    "    )\n",
    "    \n",
    "    print(\"\\n🔍 Neighbors with OVERRIDE priority (genome_size, protein_coding_gene_count):\")\n",
    "    if result_override['ancestor']:\n",
    "        anc = result_override['ancestor']\n",
    "        meta = anc.get('metadata', {})\n",
    "        print(f\"\\n  Ancestor: {anc['name']} (distance: {anc['distance']})\")\n",
    "        print(f\"    Genome size:   {meta.get('genome_size', 'N/A'):,}\")\n",
    "        print(f\"    Protein genes: {meta.get('protein_coding_gene_count', 'N/A'):,}\")\n",
    "else:\n",
    "    print(\"Could not find a suitable test tax_id without data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "7690d055",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======================================================================\n",
      "FEATURE 4: LEAF-ONLY PROCESSING\n",
      "======================================================================\n",
      "\n",
      "Processing only leaf nodes from rRNA data...\n",
      "Total rRNA tax_ids: 105,793\n",
      "Leaf nodes in rRNA: 96,898 (91.6%)\n",
      "\n",
      "Querying sample of leaf nodes...\n",
      "\n",
      "📊 Leaf-Only Query Statistics:\n",
      "  Total queried:        200\n",
      "  Self has data:        152\n",
      "  Ancestor found:       1\n",
      "  Descendant found:     0\n",
      "  No match found:       47\n",
      "\n",
      "📏 Distance Distribution for Leaves:\n",
      "  Distance 0: 152 matches\n",
      "  Distance 1: 1 matches\n",
      "\n",
      "✅ Coverage: 76.5% of rRNA leaf nodes have data or found ancestors with data\n"
     ]
    }
   ],
   "source": [
    "# Feature 4: Leaf-only processing (optimized for rRNA mapping use case)\n",
    "print(\"\\n\" + \"=\"*70)\n",
    "print(\"FEATURE 4: LEAF-ONLY PROCESSING\")\n",
    "print(\"=\"*70)\n",
    "\n",
    "# Since rRNA sequences can only come from extant organisms (leaves),\n",
    "# we can optimize by only processing leaf nodes\n",
    "\n",
    "print(\"\\nProcessing only leaf nodes from rRNA data...\")\n",
    "\n",
    "# Get all unique tax_ids from rRNA data\n",
    "all_rrna_tax_ids = rrna_df[\"taxid\"].cast(pl.Int64).unique().to_list()\n",
    "print(f\"Total rRNA tax_ids: {len(all_rrna_tax_ids):,}\")\n",
    "\n",
    "# Check how many are leaves\n",
    "leaves_in_rrna = [tid for tid in all_rrna_tax_ids if tid in tax_tree_enhanced.leaf_nodes]\n",
    "print(f\"Leaf nodes in rRNA: {len(leaves_in_rrna):,} ({len(leaves_in_rrna)/len(all_rrna_tax_ids)*100:.1f}%)\")\n",
    "\n",
    "# Batch query leaves only with statistics\n",
    "print(\"\\nQuerying sample of leaf nodes...\")\n",
    "sample_leaves = leaves_in_rrna[:200] if len(leaves_in_rrna) > 200 else leaves_in_rrna\n",
    "\n",
    "results_leaves, stats_leaves = tax_tree_enhanced.find_nearest_with_data_batch(\n",
    "    tax_ids=sample_leaves,\n",
    "    max_distance=10,\n",
    "    leaves_only=True,  # Only process leaves\n",
    "    return_stats=True\n",
    ")\n",
    "\n",
    "print(\"\\n📊 Leaf-Only Query Statistics:\")\n",
    "print(f\"  Total queried:        {stats_leaves['total_queried']:,}\")\n",
    "print(f\"  Self has data:        {stats_leaves['self_has_data']:,}\")\n",
    "print(f\"  Ancestor found:       {stats_leaves['ancestor_found']:,}\")\n",
    "print(f\"  Descendant found:     {stats_leaves['descendant_found']:,}\")\n",
    "print(f\"  No match found:       {stats_leaves['no_match_found']:,}\")\n",
    "\n",
    "print(\"\\n📏 Distance Distribution for Leaves:\")\n",
    "for distance in sorted(stats_leaves['distance_distribution'].keys())[:10]:\n",
    "    count = stats_leaves['distance_distribution'][distance]\n",
    "    print(f\"  Distance {distance}: {count:,} matches\")\n",
    "\n",
    "# Calculate coverage for rRNA data\n",
    "coverage_pct = ((stats_leaves['self_has_data'] + stats_leaves['ancestor_found']) / \n",
    "                stats_leaves['total_queried'] * 100)\n",
    "print(f\"\\n✅ Coverage: {coverage_pct:.1f}% of rRNA leaf nodes have data or found ancestors with data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "43e98bb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "======================================================================\n",
      "PRACTICAL APPLICATION: rRNA -> GENOME MAPPING TABLE\n",
      "======================================================================\n",
      "\n",
      "Processing all rRNA tax_ids to build mapping table...\n"
     ]
    }
   ],
   "source": [
    "# Practical example: Build lookup table for rRNA sequences -> genomes\n",
    "print(\"\\n\" + \"=\"*70)\n",
    "print(\"PRACTICAL APPLICATION: rRNA -> GENOME MAPPING TABLE\")\n",
    "print(\"=\"*70)\n",
    "\n",
    "# Process ALL rRNA tax_ids (this might take a moment)\n",
    "print(\"\\nProcessing all rRNA tax_ids to build mapping table...\")\n",
    "\n",
    "all_results, all_stats = tax_tree_enhanced.find_nearest_with_data_batch(\n",
    "    tax_ids=all_rrna_tax_ids,\n",
    "    max_distance=4,\n",
    "    leaves_only=False,  # Process all to see full picture\n",
    "    return_stats=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "2c65d598",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[261], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m mapping_records \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m tax_id, result \u001b[38;5;129;01min\u001b[39;00m all_results\u001b[38;5;241m.\u001b[39mitems():\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;66;03m# Get rRNA info\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m     rrna_info \u001b[38;5;241m=\u001b[39m \u001b[43mrrna_df\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilter\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcol\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtaxid\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcast\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mInt64\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mtax_id\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241m.\u001b[39mrow(\u001b[38;5;241m0\u001b[39m, named\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m) \u001b[38;5;28;01mif\u001b[39;00m tax_id \u001b[38;5;129;01min\u001b[39;00m rrna_df[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtaxid\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mcast(pl\u001b[38;5;241m.\u001b[39mInt64) \u001b[38;5;28;01melse\u001b[39;00m {}\n\u001b[1;32m      7\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m result[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mself_has_data\u001b[39m\u001b[38;5;124m'\u001b[39m]:\n\u001b[1;32m      8\u001b[0m         \u001b[38;5;66;03m# Direct mapping\u001b[39;00m\n\u001b[1;32m      9\u001b[0m         mapping_records\u001b[38;5;241m.\u001b[39mappend({\n\u001b[1;32m     10\u001b[0m             \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrrna_tax_id\u001b[39m\u001b[38;5;124m'\u001b[39m: tax_id,\n\u001b[1;32m     11\u001b[0m             \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrrna_organism\u001b[39m\u001b[38;5;124m'\u001b[39m: tax_tree_enhanced\u001b[38;5;241m.\u001b[39mname_map\u001b[38;5;241m.\u001b[39mget(tax_id, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUnknown\u001b[39m\u001b[38;5;124m'\u001b[39m),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     18\u001b[0m             \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome_size\u001b[39m\u001b[38;5;124m'\u001b[39m: result[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mancestor\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmetadata\u001b[39m\u001b[38;5;124m'\u001b[39m, {})\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgenome_size\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m     19\u001b[0m         })\n",
      "File \u001b[0;32m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/polars/dataframe/frame.py:5179\u001b[0m, in \u001b[0;36mDataFrame.filter\u001b[0;34m(self, *predicates, **constraints)\u001b[0m\n\u001b[1;32m   5021\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   5022\u001b[0m \u001b[38;5;124;03mFilter rows, retaining those that match the given predicate expression(s).\u001b[39;00m\n\u001b[1;32m   5023\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   5174\u001b[0m \u001b[38;5;124;03m└──────┴──────┴─────┘\u001b[39;00m\n\u001b[1;32m   5175\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m   5176\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mpolars\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mlazyframe\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mopt_flags\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m QueryOptFlags\n\u001b[1;32m   5178\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m (\n\u001b[0;32m-> 5179\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mlazy\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5180\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfilter\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mpredicates\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mconstraints\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5181\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcollect\u001b[49m\u001b[43m(\u001b[49m\u001b[43moptimizations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mQueryOptFlags\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_eager\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   5182\u001b[0m )\n",
      "File \u001b[0;32m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/polars/_utils/deprecation.py:97\u001b[0m, in \u001b[0;36mdeprecate_streaming_parameter.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     93\u001b[0m         kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mengine\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124min-memory\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m     95\u001b[0m     \u001b[38;5;28;01mdel\u001b[39;00m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstreaming\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m---> 97\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/polars/lazyframe/opt_flags.py:330\u001b[0m, in \u001b[0;36mforward_old_opt_flags.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    327\u001b[0m         optflags \u001b[38;5;241m=\u001b[39m cb(optflags, kwargs\u001b[38;5;241m.\u001b[39mpop(key))  \u001b[38;5;66;03m# type: ignore[no-untyped-call,unused-ignore]\u001b[39;00m\n\u001b[1;32m    329\u001b[0m kwargs[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124moptimizations\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m optflags\n\u001b[0;32m--> 330\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunction\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/polars/lazyframe/frame.py:2332\u001b[0m, in \u001b[0;36mLazyFrame.collect\u001b[0;34m(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, cluster_with_columns, collapse_joins, no_optimization, engine, background, optimizations, **_kwargs)\u001b[0m\n\u001b[1;32m   2330\u001b[0m \u001b[38;5;66;03m# Only for testing purposes\u001b[39;00m\n\u001b[1;32m   2331\u001b[0m callback \u001b[38;5;241m=\u001b[39m _kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpost_opt_callback\u001b[39m\u001b[38;5;124m\"\u001b[39m, callback)\n\u001b[0;32m-> 2332\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m wrap_df(\u001b[43mldf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcollect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcallback\u001b[49m\u001b[43m)\u001b[49m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Build a mapping table\n",
    "mapping_records = []\n",
    "for tax_id, result in all_results.items():\n",
    "    # Get rRNA info\n",
    "    rrna_info = rrna_df.filter(pl.col(\"taxid\").cast(pl.Int64) == tax_id).row(0, named=True) if tax_id in rrna_df[\"taxid\"].cast(pl.Int64) else {}\n",
    "    \n",
    "    if result['self_has_data']:\n",
    "        # Direct mapping\n",
    "        mapping_records.append({\n",
    "            'rrna_tax_id': tax_id,\n",
    "            'rrna_organism': tax_tree_enhanced.name_map.get(tax_id, 'Unknown'),\n",
    "            'genome_tax_id': tax_id,\n",
    "            'genome_organism': tax_tree_enhanced.name_map.get(tax_id, 'Unknown'),\n",
    "            'relationship': 'self',\n",
    "            'distance': 0,\n",
    "            'ftp_path': result['ancestor'].get('metadata', {}).get('ftp_path', ''),\n",
    "            'protein_coding_genes': result['ancestor'].get('metadata', {}).get('protein_coding_gene_count', 0),\n",
    "            'genome_size': result['ancestor'].get('metadata', {}).get('genome_size', 0)\n",
    "        })\n",
    "    elif result['ancestor']:\n",
    "        # Map to ancestor\n",
    "        anc = result['ancestor']\n",
    "        mapping_records.append({\n",
    "            'rrna_tax_id': tax_id,\n",
    "            'rrna_organism': tax_tree_enhanced.name_map.get(tax_id, 'Unknown'),\n",
    "            'genome_tax_id': anc['tax_id'],\n",
    "            'genome_organism': anc['name'],\n",
    "            'relationship': 'ancestor',\n",
    "            'distance': anc['distance'],\n",
    "            'ftp_path': anc.get('metadata', {}).get('ftp_path', ''),\n",
    "            'protein_coding_genes': anc.get('metadata', {}).get('protein_coding_gene_count', 0),\n",
    "            'genome_size': anc.get('metadata', {}).get('genome_size', 0)\n",
    "        })\n",
    "    elif result['descendants']:\n",
    "        # Map to best descendant\n",
    "        desc = result['descendants'][0]  # Already prioritized\n",
    "        mapping_records.append({\n",
    "            'rrna_tax_id': tax_id,\n",
    "            'rrna_organism': tax_tree_enhanced.name_map.get(tax_id, 'Unknown'),\n",
    "            'genome_tax_id': desc['tax_id'],\n",
    "            'genome_organism': desc['name'],\n",
    "            'relationship': 'descendant',\n",
    "            'distance': desc['distance'],\n",
    "            'ftp_path': desc.get('metadata', {}).get('ftp_path', ''),\n",
    "            'protein_coding_genes': desc.get('metadata', {}).get('protein_coding_gene_count', 0),\n",
    "            'genome_size': desc.get('metadata', {}).get('genome_size', 0)\n",
    "        })\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5924b4a3",
   "metadata": {},
   "source": [
    "## Alternative: Using bigtree library\n",
    "\n",
    "Exploring the `bigtree` library which provides native Polars support for tree construction and manipulation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "d30e9a73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ bigtree version 0.30.1 is installed\n"
     ]
    }
   ],
   "source": [
    "# Try using bigtree library with polars_to_tree\n",
    "# First, check if bigtree is installed, if not, install it\n",
    "try:\n",
    "    import bigtree  \n",
    "    print(f\"✅ bigtree version {bigtree.__version__} is installed\")\n",
    "except ImportError:\n",
    "    print(\"📦 Installing bigtree...\")\n",
    "    import subprocess\n",
    "    subprocess.run([\"pip\", \"install\", \"bigtree\"], check=True)\n",
    "    import bigtree\n",
    "    print(f\"✅ bigtree version {bigtree.__version__} installed successfully\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "1352f703",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing data for bigtree's polars_to_tree...\n",
      "Starting with 2,708,826 nodes\n",
      "Collected 148,236 nodes (including ancestors)\n",
      "Filtered DataFrame has 148,236 rows\n",
      "Collected 148,236 nodes (including ancestors)\n",
      "Filtered DataFrame has 148,236 rows\n"
     ]
    }
   ],
   "source": [
    "# Prepare data for bigtree\n",
    "# bigtree expects a path column that shows the full hierarchical path\n",
    "# We'll use our nodes_with_names DataFrame\n",
    "\n",
    "print(\"Preparing data for bigtree's polars_to_tree...\")\n",
    "print(f\"Starting with {nodes_with_names.height:,} nodes\")\n",
    "\n",
    "# Let's use nodes related to our rRNA data\n",
    "tax_ids_bigtree = rrna_df[\"taxid\"].cast(pl.Int64).unique().to_list()\n",
    "\n",
    "# Get all ancestors for these nodes to build complete paths\n",
    "def get_all_ancestors(tax_id, parent_map, max_depth=50):\n",
    "    \"\"\"Get all ancestors up to root\"\"\"\n",
    "    ancestors = []\n",
    "    current = tax_id\n",
    "    depth = 0\n",
    "    visited = set()\n",
    "    \n",
    "    while current in parent_map and depth < max_depth and current not in visited:\n",
    "        visited.add(current)\n",
    "        ancestors.append(current)\n",
    "        parent = parent_map[current]\n",
    "        if parent == current:  # Root\n",
    "            break\n",
    "        current = parent\n",
    "        depth += 1\n",
    "    \n",
    "    return ancestors\n",
    "\n",
    "# Build parent map for quick lookups\n",
    "parent_map_bigtree = dict(zip(nodes_with_names[\"tax_id\"], nodes_with_names[\"parent_tax_id\"]))\n",
    "\n",
    "# Get all nodes needed \n",
    "all_needed_ids = set()\n",
    "for tid in tax_ids_bigtree:\n",
    "    if tid in parent_map_bigtree:\n",
    "        ancestors = get_all_ancestors(tid, parent_map_bigtree)\n",
    "        all_needed_ids.update(ancestors)\n",
    "\n",
    "print(f\"Collected {len(all_needed_ids):,} nodes (including ancestors)\")\n",
    "\n",
    "# Filter to only needed nodes\n",
    "bigtree_df = nodes_with_names.filter(pl.col(\"tax_id\").is_in(all_needed_ids))\n",
    "print(f\"Filtered DataFrame has {bigtree_df.height:,} rows\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "0a2fe555",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (148_236, 4)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>tax_id</th><th>parent_tax_id</th><th>rank</th><th>scientific_name</th></tr><tr><td>i64</td><td>i64</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>1</td><td>1</td><td>&quot;no rank&quot;</td><td>&quot;root&quot;</td></tr><tr><td>2</td><td>131567</td><td>&quot;domain&quot;</td><td>&quot;Bacteria&quot;</td></tr><tr><td>6</td><td>335928</td><td>&quot;genus&quot;</td><td>&quot;Azorhizobium&quot;</td></tr><tr><td>7</td><td>6</td><td>&quot;species&quot;</td><td>&quot;Azorhizobium caulinodans&quot;</td></tr><tr><td>9</td><td>32199</td><td>&quot;species&quot;</td><td>&quot;Buchnera aphidicola&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>3467296</td><td>1760</td><td>&quot;order&quot;</td><td>&quot;Allobogoriellales&quot;</td></tr><tr><td>3471053</td><td>2841720</td><td>&quot;genus&quot;</td><td>&quot;Tibetiodes&quot;</td></tr><tr><td>3472296</td><td>159987</td><td>&quot;family&quot;</td><td>&quot;Neocamarosporiaceae&quot;</td></tr><tr><td>3472734</td><td>983542</td><td>&quot;genus&quot;</td><td>&quot;Paraleonurus&quot;</td></tr><tr><td>3476159</td><td>140471</td><td>&quot;family&quot;</td><td>&quot;Benthoplanidae&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (148_236, 4)\n",
       "┌─────────┬───────────────┬─────────┬──────────────────────────┐\n",
       "│ tax_id  ┆ parent_tax_id ┆ rank    ┆ scientific_name          │\n",
       "│ ---     ┆ ---           ┆ ---     ┆ ---                      │\n",
       "│ i64     ┆ i64           ┆ str     ┆ str                      │\n",
       "╞═════════╪═══════════════╪═════════╪══════════════════════════╡\n",
       "│ 1       ┆ 1             ┆ no rank ┆ root                     │\n",
       "│ 2       ┆ 131567        ┆ domain  ┆ Bacteria                 │\n",
       "│ 6       ┆ 335928        ┆ genus   ┆ Azorhizobium             │\n",
       "│ 7       ┆ 6             ┆ species ┆ Azorhizobium caulinodans │\n",
       "│ 9       ┆ 32199         ┆ species ┆ Buchnera aphidicola      │\n",
       "│ …       ┆ …             ┆ …       ┆ …                        │\n",
       "│ 3467296 ┆ 1760          ┆ order   ┆ Allobogoriellales        │\n",
       "│ 3471053 ┆ 2841720       ┆ genus   ┆ Tibetiodes               │\n",
       "│ 3472296 ┆ 159987        ┆ family  ┆ Neocamarosporiaceae      │\n",
       "│ 3472734 ┆ 983542        ┆ genus   ┆ Paraleonurus             │\n",
       "│ 3476159 ┆ 140471        ┆ family  ┆ Benthoplanidae           │\n",
       "└─────────┴───────────────┴─────────┴──────────────────────────┘"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bigtree_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae8959cb",
   "metadata": {},
   "source": [
    "bigtree expects the root node parent to be null, not itself"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4397e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "bigtree_df = bigtree_df.with_columns(\n",
    "    pl.when(pl.col(\"tax_id\") == 1)\n",
    "      .then(pl.lit(None))\n",
    "      .otherwise(pl.col(\"parent_tax_id\"))\n",
    "      .alias(\"parent_tax_id\")\n",
    ") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22967496",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bigtree as bt\n",
    "test_tree = bt.polars_to_tree_by_relation(\n",
    "    allow_duplicates=True,\n",
    "    child_col=\"tax_id\",\n",
    "    parent_col=\"parent_tax_id\",\n",
    "    data=bigtree_df.select([\"tax_id\",\"parent_tax_id\"]),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "4ce12a6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Node(/1, )"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c55c8bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Alternative: Using NetworkX for more advanced tree operations\n",
    "# # NetworkX is excellent for complex graph algorithms and visualization\n",
    "\n",
    "# import networkx as nx\n",
    "# nx.from_nested_tuple\n",
    "# def create_networkx_taxonomy_tree(nodes_df, data_availability_set):\n",
    "#     \"\"\"Create NetworkX directed graph for taxonomy tree\"\"\"\n",
    "#     # G = nx.DiGraph(incoming_graph_data=)\n",
    "    \n",
    "#     # Add nodes with attributes\n",
    "#     for row in nodes_df.iter_rows(named=True):\n",
    "#         tax_id = row['tax_id']\n",
    "#         G.add_node(tax_id, \n",
    "#                     parent=row['parent_tax_id'],\n",
    "#                     rank=row.get('rank', ''),\n",
    "#                     name=row.get('scientific_name', ''),\n",
    "#                     has_data=tax_id in data_availability_set)\n",
    "        \n",
    "#         # Add edges (parent -> child relationships)\n",
    "#         parent_id = row['parent_tax_id']\n",
    "#         if parent_id != tax_id:  # Avoid self-loops\n",
    "#             G.add_edge(parent_id, tax_id)\n",
    "    \n",
    "#     return G\n",
    "\n",
    "# def find_nearest_data_nx(G, tax_id, max_distance=10):\n",
    "#     \"\"\"Find nearest neighbors with data using NetworkX algorithms\"\"\"\n",
    "#     if not G.has_node(tax_id):\n",
    "#         return None\n",
    "        \n",
    "#     if G.nodes[tax_id]['has_data']:\n",
    "#         return {'self': tax_id, 'distance': 0}\n",
    "    \n",
    "#     # Find ancestors (predecessors in directed path to root)\n",
    "#     ancestors = []\n",
    "#     current = tax_id\n",
    "#     distance = 0\n",
    "#     while distance < max_distance:\n",
    "#         predecessors = list(G.predecessors(current))\n",
    "#         if not predecessors:\n",
    "#             break\n",
    "#         parent = predecessors[0]  # Should be only one parent\n",
    "#         distance += 1\n",
    "#         if G.nodes[parent]['has_data']:\n",
    "#             ancestors.append({'tax_id': parent, 'distance': distance, \n",
    "#                             'name': G.nodes[parent]['name']})\n",
    "#             break\n",
    "#         current = parent\n",
    "    \n",
    "#     # Find descendants using BFS\n",
    "#     descendants = []\n",
    "#     for node in nx.bfs_tree(G, tax_id, max_depth=max_distance):\n",
    "#         if node != tax_id and G.nodes[node]['has_data']:\n",
    "#             distance = nx.shortest_path_length(G, tax_id, node)\n",
    "#             descendants.append({'tax_id': node, 'distance': distance,\n",
    "#                                 'name': G.nodes[node]['name']})\n",
    "    \n",
    "#     return {\n",
    "#         'ancestors': ancestors,\n",
    "#         'descendants': sorted(descendants, key=lambda x: x['distance'])\n",
    "#     }\n",
    "\n",
    "# test = create_networkx_taxonomy_tree(nodes_df,data_available_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f283c467",
   "metadata": {},
   "source": [
    "## Alternative: Using tskit (Tree Sequence Kit)\n",
    "\n",
    "`tskit` is a high-performance library originally designed for population genetics, but works great for large taxonomy trees. It uses efficient C implementations and can handle millions of nodes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "59ff0081",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing taxonomy tree for tskit...\n",
      "Created mapping for 148,237 unique nodes\n",
      "Prepared 148,236 edges for tskit\n",
      "Sample edges:\n",
      "shape: (3, 4)\n",
      "┌───────┬────────┬─────────────────┬─────────┐\n",
      "│ child ┆ parent ┆ scientific_name ┆ rank    │\n",
      "│ ---   ┆ ---    ┆ ---             ┆ ---     │\n",
      "│ u32   ┆ u32    ┆ str             ┆ str     │\n",
      "╞═══════╪════════╪═════════════════╪═════════╡\n",
      "│ 1     ┆ null   ┆ root            ┆ no rank │\n",
      "│ 2     ┆ 28572  ┆ Bacteria        ┆ domain  │\n",
      "│ 3     ┆ 55375  ┆ Azorhizobium    ┆ genus   │\n",
      "└───────┴────────┴─────────────────┴─────────┘\n"
     ]
    }
   ],
   "source": [
    "# Import tskit and prepare taxonomy data for tskit format\n",
    "import tskit\n",
    "import numpy as np\n",
    "\n",
    "print(\"Preparing taxonomy tree for tskit...\")\n",
    "\n",
    "# Use the same subset as bigtree for comparison\n",
    "tskit_df = bigtree_df.select([\"tax_id\", \"parent_tax_id\", \"scientific_name\", \"rank\"])\n",
    "\n",
    "# Create ID mapping: tskit requires integer node IDs starting from 0\n",
    "all_tax_ids = pl.concat([\n",
    "    tskit_df[\"tax_id\"], \n",
    "    tskit_df[\"parent_tax_id\"]\n",
    "]).unique().sort()\n",
    "\n",
    "# Create mapping from tax_id to sequential integer\n",
    "id_map = all_tax_ids.to_frame(\"tax_id\").with_row_index(\"node_id\")\n",
    "print(f\"Created mapping for {id_map.height:,} unique nodes\")\n",
    "\n",
    "# Join to get integer IDs - do in two separate steps to avoid column conflicts\n",
    "# First join for child IDs\n",
    "tskit_edges = tskit_df.join(id_map, left_on=\"tax_id\", right_on=\"tax_id\", how=\"left\")\n",
    "tskit_edges = tskit_edges.rename({\"node_id\": \"child\"})\n",
    "\n",
    "# Second join for parent IDs\n",
    "parent_map = id_map.rename({\"node_id\": \"parent\", \"tax_id\": \"parent_tax_id\"})\n",
    "tskit_edges = tskit_edges.join(parent_map, on=\"parent_tax_id\", how=\"left\")\n",
    "\n",
    "# Select final columns\n",
    "tskit_edges = tskit_edges.select([\"child\", \"parent\", \"scientific_name\", \"rank\"])\n",
    "\n",
    "print(f\"Prepared {tskit_edges.height:,} edges for tskit\")\n",
    "print(f\"Sample edges:\\n{tskit_edges.head(3)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "736bb34c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Building tskit tree structure...\n",
      "Computing node depths...\n",
      "Found 1 root node(s): [1]\n",
      "Depth range: 0 to 36\n",
      "Time range: 0.0 to 36.0\n",
      "Nodes at depth 0 (roots): 1\n",
      "Added 148,237 nodes\n",
      "Edges before filtering: 148,236\n",
      "Edges after filtering: 148,235\n",
      "Added 148,235 edges to tree\n",
      "Sorting tables...\n",
      "Building index...\n",
      "\n",
      "✅ Created tskit TreeSequence:\n",
      "   Nodes: 148,237\n",
      "   Edges: 148,235\n",
      "   Trees: 1\n",
      "Depth range: 0 to 36\n",
      "Time range: 0.0 to 36.0\n",
      "Nodes at depth 0 (roots): 1\n",
      "Added 148,237 nodes\n",
      "Edges before filtering: 148,236\n",
      "Edges after filtering: 148,235\n",
      "Added 148,235 edges to tree\n",
      "Sorting tables...\n",
      "Building index...\n",
      "\n",
      "✅ Created tskit TreeSequence:\n",
      "   Nodes: 148,237\n",
      "   Edges: 148,235\n",
      "   Trees: 1\n"
     ]
    }
   ],
   "source": [
    "# Build tskit TableCollection and TreeSequence (with explicit root handling)\n",
    "print(\"Building tskit tree structure...\")\n",
    "\n",
    "# Initialize TableCollection\n",
    "tables = tskit.TableCollection(sequence_length=1.0)\n",
    "\n",
    "print(\"Computing node depths...\")\n",
    "\n",
    "# Build parent-child map, excluding self-loops\n",
    "parent_child_map = defaultdict(list)\n",
    "all_nodes = set(range(id_map.height))\n",
    "children_set = set()\n",
    "\n",
    "for row in tskit_edges.filter(pl.col(\"parent\").is_not_null()).iter_rows(named=True):\n",
    "    parent = row['parent']\n",
    "    child = row['child']\n",
    "    if parent != child:  # Exclude self-loops (this is the root)\n",
    "        parent_child_map[parent].append(child)\n",
    "        children_set.add(child)\n",
    "\n",
    "# Find roots: nodes that are never children OR nodes where parent == child\n",
    "roots = []\n",
    "for row in tskit_edges.iter_rows(named=True):\n",
    "    child = row['child']\n",
    "    parent = row['parent']\n",
    "    # Root if: no parent, self-parent, or never appears as a child\n",
    "    if parent is None or parent == child or child not in children_set:\n",
    "        roots.append(child)\n",
    "\n",
    "roots = list(set(roots))  # Deduplicate\n",
    "print(f\"Found {len(roots)} root node(s): {roots}\")\n",
    "\n",
    "if not roots:\n",
    "    # Fallback: just use node 0 as root\n",
    "    print(\"Warning: No clear root found, using node 0\")\n",
    "    roots = [0]\n",
    "\n",
    "# BFS to assign depths - start from roots at depth 0\n",
    "depth_map = {}\n",
    "queue = deque([(root, 0) for root in roots])\n",
    "visited = set()\n",
    "\n",
    "while queue:\n",
    "    node, depth = queue.popleft()\n",
    "    if node in visited:\n",
    "        continue\n",
    "    visited.add(node)\n",
    "    depth_map[node] = depth\n",
    "    \n",
    "    # Add children at next depth\n",
    "    for child in parent_child_map.get(node, []):\n",
    "        if child not in visited:\n",
    "            queue.append((child, depth + 1))\n",
    "\n",
    "# Assign depth to any unvisited nodes (orphans)\n",
    "max_depth = max(depth_map.values()) if depth_map else 0\n",
    "for node_id in all_nodes:\n",
    "    if node_id not in depth_map:\n",
    "        depth_map[node_id] = max_depth + 1\n",
    "\n",
    "# Create time array: time = max_depth - depth (so root has highest time)\n",
    "max_depth = max(depth_map.values())\n",
    "times = np.array([max_depth - depth_map.get(i, 0) for i in range(id_map.height)], dtype=np.float64)\n",
    "\n",
    "print(f\"Depth range: 0 to {max_depth}\")\n",
    "print(f\"Time range: {times.min():.1f} to {times.max():.1f}\")\n",
    "print(f\"Nodes at depth 0 (roots): {sum(1 for d in depth_map.values() if d == 0)}\")\n",
    "\n",
    "# Add nodes\n",
    "tables.nodes.set_columns(\n",
    "    flags=np.zeros(id_map.height, dtype=np.uint32),\n",
    "    time=times\n",
    "\n",
    ")\n",
    "# tables.nodes.set_columns(\n",
    "#     time=np.zeros(id_map.height), # Placeholder time\n",
    "#     flags=np.zeros(id_map.height, dtype=np.uint32)\n",
    "# )\n",
    "\n",
    "print(f\"Added {id_map.height:,} nodes\")\n",
    "\n",
    "# Add edges (exclude self-loops)\n",
    "edges_df = tskit_edges.filter(\n",
    "    pl.col(\"parent\").is_not_null() & \n",
    "    (pl.col(\"parent\") != pl.col(\"child\"))\n",
    ")\n",
    "\n",
    "print(f\"Edges before filtering: {tskit_edges.height:,}\")\n",
    "print(f\"Edges after filtering: {edges_df.height:,}\")\n",
    "\n",
    "if edges_df.height > 0:\n",
    "    tables.edges.set_columns(\n",
    "        left=np.zeros(edges_df.height, dtype=np.float64),\n",
    "        right=np.ones(edges_df.height, dtype=np.float64),\n",
    "        parent=edges_df[\"parent\"].cast(pl.Int32).to_numpy(),\n",
    "        child=edges_df[\"child\"].cast(pl.Int32).to_numpy()\n",
    "    )\n",
    "    print(f\"Added {edges_df.height:,} edges to tree\")\n",
    "\n",
    "# Sort and build index\n",
    "print(\"Sorting tables...\")\n",
    "tables.sort()\n",
    "print(\"Building index...\")\n",
    "tables.build_index()\n",
    "\n",
    "# Create tree sequence\n",
    "ts = tables.tree_sequence()\n",
    "\n",
    "print(\"\\n✅ Created tskit TreeSequence:\")\n",
    "print(f\"   Nodes: {ts.num_nodes:,}\")\n",
    "print(f\"   Edges: {ts.num_edges:,}\")\n",
    "print(f\"   Trees: {ts.num_trees}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946f8e24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======================================================================\n",
      "TSKIT TREE OPERATIONS\n",
      "======================================================================\n",
      "\n",
      "1️⃣ Tree properties:\n",
      "   Root: -1\n",
      "   Num roots: 0\n",
      "   Total branch length: 0.00\n",
      "\n",
      "2️⃣ MRCA of nodes [5, 10, 15]: 757\n",
      "\n",
      "3️⃣ Tree traversal examples:\n",
      "   Root node -1 has 1 children: (28572,)\n"
     ]
    }
   ],
   "source": [
    "# Explore tskit tree operations\n",
    "print(\"=\"*70)\n",
    "print(\"TSKIT TREE OPERATIONS\")\n",
    "print(\"=\"*70)\n",
    "\n",
    "# Get the first (and only) tree\n",
    "tree = ts.first()\n",
    "\n",
    "print(\"\\n1️⃣ Tree properties:\")\n",
    "print(f\"   Root: {tree.root}\")\n",
    "print(f\"   Num roots: {tree.num_roots}\")\n",
    "print(f\"   Total branch length: {tree.total_branch_length:.2f}\")\n",
    "\n",
    "# Find MRCA (Most Recent Common Ancestor) - useful for taxonomy!\n",
    "sample_nodes = [5, 10, 15]  # Example node IDs\n",
    "if all(n < ts.num_nodes for n in sample_nodes):\n",
    "    mrca = tree.mrca(*sample_nodes)\n",
    "    print(f\"\\n2️⃣ MRCA of nodes {sample_nodes}: {mrca}\")\n",
    "\n",
    "# Get parent/child relationships\n",
    "print(\"\\n3️⃣ Tree traversal examples:\")\n",
    "# Get children of root\n",
    "root_children = tree.get_children(1)\n",
    "print(f\"   Root node {tree.root} has {len(root_children)} children: {root_children[:5]}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ada1f2dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   Node 10 has parent: 6457\n",
      "\n",
      "4️⃣ Samples (leaves) under root:\n",
      "   Found 0 leaf nodes\n",
      "   First 10: []\n",
      "\n",
      "5️⃣ Path from node 20 to root (lineage):\n",
      "   Path (7 nodes): [20, 6461, 139261, 147261, 2, 28572, 1]\n",
      "\n",
      "✅ tskit provides fast tree traversal operations!\n"
     ]
    }
   ],
   "source": [
    "# Get parent of a node\n",
    "sample_node = 10\n",
    "if sample_node < ts.num_nodes:\n",
    "    parent = tree.parent(sample_node)\n",
    "    print(f\"   Node {sample_node} has parent: {parent}\")\n",
    "\n",
    "# Get all samples under a node\n",
    "print(\"\\n4️⃣ Samples (leaves) under root:\")\n",
    "samples_under_root = list(tree.samples(tree.root))\n",
    "print(f\"   Found {len(samples_under_root)} leaf nodes\")\n",
    "print(f\"   First 10: {samples_under_root[:10]}\")\n",
    "\n",
    "# Path from node to root (lineage!)\n",
    "sample_node = 20\n",
    "if sample_node < ts.num_nodes:\n",
    "    print(f\"\\n5️⃣ Path from node {sample_node} to root (lineage):\")\n",
    "    current = sample_node\n",
    "    path = []\n",
    "    while current != tskit.NULL:\n",
    "        path.append(current)\n",
    "        current = tree.parent(current)\n",
    "    print(f\"   Path ({len(path)} nodes): {path[:10]}{'...' if len(path) > 10 else ''}\")\n",
    "\n",
    "print(\"\\n✅ tskit provides fast tree traversal operations!\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
