{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "222afdb2",
   "metadata": {},
   "source": [
    "# How much gets filtered when removing rRNA, tRNA, and mtDNA/RNA?\n",
    "# Notebook 1: Actual tests\n",
    "\n",
    "I'm making this notebook as a n intermission from preparing rolypoly external database. So far I've used a combination of NCBI +SILVA rRNAs to remove contaminating rRNA reads from RNA-seq data. Subsequently, the organisms whose rRNAs were most matched, are then also fetched (or their transcriptomes, if available) to remove any remaining reads that may have come from those organisms.  \n",
    "THis is messy, requires NCBI taxdump, taxonkit, and ncbi-datasets. So I started removing this dependency by using a set of rRNAs for which I can generate a prebuilt table containing the FTP addresses of the hosts' genomes/transcriptomes.  \n",
    "While doing so, I realised the step above could be split - quick rRNA mapping to get rough taxonomic breakdown of the sample, and then a more thorough removal of rRNA, tRNA, and mtDNA/RNA using a more comprehensive database. The question is then how much will these diffrent combinations filter, how much more time, and would masking the fasta for subsequences shared with RNA viruses change the results significantly?\n",
    "\n",
    "To test these, below are the actual mapping/search/etc filtering of these sequence types, and some graphs and so on."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b6f89d",
   "metadata": {},
   "source": [
    "*Note*: Parts of this script are from [build_data.py](../src/rolypoly/commands/misc/build_data.py) and [filter_reads.py](../src/rolypoly/commands/reads/filter_reads.py) scripts in the rolypoly repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57818007",
   "metadata": {},
   "source": [
    "Loading libraries and defining paths to sets already created/downloaded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a12c02e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import logging\n",
    "import shutil\n",
    "import subprocess\n",
    "import aria2p\n",
    "import tempfile\n",
    "import os\n",
    "import time\n",
    "from ftplib import FTP\n",
    "import glob\n",
    "\n",
    "import polars as pl\n",
    "from tqdm.notebook import tqdm\n",
    "from bbmapy import bbduk, bbmask, kcompress\n",
    "\n",
    "from rolypoly.utils.bio.sequences import (\n",
    "    filter_fasta_by_headers,\n",
    "    write_fasta_file,\n",
    "    remove_duplicates\n",
    ")\n",
    "\n",
    "from rolypoly.utils.bio.polars_fastx import from_fastx_eager\n",
    "\n",
    "from rolypoly.utils.logging.loggit import  setup_logging\n",
    "from rolypoly.utils.various import fetch_and_extract, run_command_comp, extract\n",
    "\n",
    "### DEBUG ARGS (for manually building, not entering via CLI):\n",
    "threads = 6\n",
    "log_file = \"notebooks/Exprimental/trrna.log\"\n",
    "data_dir = \"/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\"\n",
    "\n",
    "global rrna_dir\n",
    "global contam_dir\n",
    "\n",
    "logger = setup_logging(log_file) # for some functions from build_data.py that require/are easier with a logger...\n",
    "\n",
    "contam_dir = os.path.join(data_dir, \"contam\")\n",
    "rrna_dir = os.path.join(contam_dir, \"rrna\")\n",
    "trna_dir =  os.path.join(contam_dir, \"trna\")\n",
    "trna_seqs = os.path.join(trna_dir, \"tRNA_sequences_deduplicated_filtered.fasta\")\n",
    "masking_dir = os.path.join(contam_dir, \"masking\")\n",
    "taxonomy_dir = os.path.join(data_dir, \"taxdump\")\n",
    "reference_seqs = os.path.join(data_dir, \"reference_seqs\")\n",
    "mmseqs_ref_dir = os.path.join(reference_seqs, \"mmseqs\")\n",
    "rvmt_dir = os.path.join(reference_seqs, \"RVMT\")\n",
    "ncbi_ribovirus_dir = os.path.join(reference_seqs, \"ncbi_ribovirus\")\n",
    "\n",
    "\n",
    "# Masking sequences \n",
    "rvmt_fasta_path = os.path.join(data_dir, \"reference_seqs\", \"RVMT\", \"RVMT_cleaned_contigs.fasta\")\n",
    "ncbi_ribovirus_fasta_path = os.path.join(data_dir,\"reference_seqs\",\"ncbi_ribovirus\",\"refseq_ribovirus_genomes.fasta\")\n",
    "rna_viruses_entropy_masked_path = os.path.join(masking_dir, \"combined_entropy_masked.fasta\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2aceda4",
   "metadata": {},
   "source": [
    "## Reading fetched/processed data (SILVA, NCBI rRNAs, tRNAs, mtDNAs):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "689a1346",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Silva\n",
    "silva_release = \"138.2\"\n",
    "silva_ssu_path = os.path.join(rrna_dir, f\"SILVA_{silva_release}_SSURef_NR99_tax_silva.fasta\")\n",
    "silva_lsu_path = os.path.join(rrna_dir, f\"SILVA_{silva_release}_LSURef_NR99_tax_silva.fasta\")\n",
    "silva_fasta_df = pl.read_parquet(os.path.join(rrna_dir, \"silva99_fasta.parquet\"))\n",
    "silva_df = pl.read_parquet(os.path.join(rrna_dir, \"silva_rrna_sequences.parquet\"))\n",
    "\n",
    "# organelles\n",
    "mito_fasta = os.path.join( os.path.join(data_dir, \"reference_seqs\", \"mito_refseq\"), \"combined_mito_refseq.fasta\")\n",
    "plastid_fasta = os.path.join( os.path.join(data_dir, \"reference_seqs\", \"plastid_refseq\"), \"combined_plastid_refseq.fasta\")\n",
    "\n",
    "# rrna (ITS, 16S, 18S,23S, LSU, SSU from blastdbs)\n",
    "rrna_df = pl.scan_csv(\n",
    "    os.path.join(rrna_dir, \"*.tab\"),\n",
    "    separator=\"@\",\n",
    "    has_header=False,\n",
    "    null_values = [\"N/A\"],\n",
    "    new_columns=[\"taxid\", \"header\",\"name\",\"sequence\"],\n",
    "    include_file_paths=\"type\"\n",
    ").collect()\n",
    "rrna_df = rrna_df.drop(\"name\")\n",
    "rrna_df = rrna_df.with_columns(\n",
    "    rRNA_type=pl.col(\"type\").str.extract(r\"([^/]+)\\.tab$\", 1)\n",
    ").drop(\"type\")\n",
    "\n",
    "for rna_type in rrna_df[\"rRNA_type\"].unique():\n",
    "    temp_fasta = os.path.join(rrna_dir, f\"{rna_type}_ncbi.fasta\")\n",
    "    temp_df = rrna_df.filter(pl.col([\"rRNA_type\"]) == rna_type)\n",
    "    write_fasta_file(\n",
    "        seqs=temp_df[\"sequence\"].to_list(),\n",
    "        headers=temp_df[\"header\"].to_list(),\n",
    "        output_file=temp_fasta,\n",
    "    )\n",
    "    del temp_df\n",
    "\n",
    "# tRNAs\n",
    "trna_fasta = os.path.join( os.path.join(data_dir, \"\", \"trna_refseq\"), \"combined_trna_refseq.fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c76cfca2",
   "metadata": {},
   "source": [
    "### create entropy masked, RNAvirus masked, dust masked variants for each of the sequences to test as filters\n",
    "also need to remember to create a kcomressed version (so reduced, compacted version of kmers of length something)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c17f5ea6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (13, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>name</th><th>raw_fasta_file</th></tr><tr><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;ITS&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;16S&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;18S&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;23S&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;LSU&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;plastid&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;silva_SSU&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;silva_LSU&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;RefSeq_RNA_viruses&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr><tr><td>&quot;RVMT&quot;</td><td>&quot;/clusterfs/jgi/scratch/science…</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (13, 2)\n",
       "┌────────────────────┬─────────────────────────────────┐\n",
       "│ name               ┆ raw_fasta_file                  │\n",
       "│ ---                ┆ ---                             │\n",
       "│ str                ┆ str                             │\n",
       "╞════════════════════╪═════════════════════════════════╡\n",
       "│ ITS                ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ 16S                ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ 18S                ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ 23S                ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ LSU                ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ …                  ┆ …                               │\n",
       "│ plastid            ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ silva_SSU          ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ silva_LSU          ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ RefSeq_RNA_viruses ┆ /clusterfs/jgi/scratch/science… │\n",
       "│ RVMT               ┆ /clusterfs/jgi/scratch/science… │\n",
       "└────────────────────┴─────────────────────────────────┘"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# # for simplicity, will make a dataframe with the info on all sequence types to be tested\n",
    "test_track_df = pl.DataFrame({\n",
    "    \"name\" : [\"ITS\",\"16S\",\"18S\",\"23S\",\"LSU\",\"SSU\",\"tRNA\", \"mito\",\"plastid\",\"silva_SSU\",\"silva_LSU\",\"RefSeq_RNA_viruses\",\"RVMT\"],\n",
    "    \"raw_fasta_file\" : [\n",
    "        os.path.join(rrna_dir, f\"ITS_ncbi_rRNA_all_sequences.fasta\"),\n",
    "        os.path.join(rrna_dir, f\"16S_ncbi_rRNA_all_sequences.fasta\"),\n",
    "        os.path.join(rrna_dir, f\"18S_ncbi_rRNA_all_sequences.fasta\"),\n",
    "        os.path.join(rrna_dir, f\"23S_ncbi_rRNA_all_sequences.fasta\"),\n",
    "        os.path.join(rrna_dir, f\"LSU_ncbi_rRNA_all_sequences.fasta\"),\n",
    "        os.path.join(rrna_dir, f\"SSU_ncbi_rRNA_all_sequences.fasta\"),\n",
    "        trna_fasta,\n",
    "        mito_fasta,\n",
    "        plastid_fasta,\n",
    "        silva_ssu_path,\n",
    "        silva_lsu_path,\n",
    "        ncbi_ribovirus_fasta_path,\n",
    "        rvmt_fasta_path\n",
    "    ]})\n",
    "test_track_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40bc1550",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge SILVA sequences and apply entropy masking\n",
    "print(\"Merging and masking SILVA sequences\")\n",
    "\n",
    "# # Apply entropy masking\n",
    "# bbduk(\n",
    "#     in1=silva_merged,\n",
    "#     out=silva_masked,\n",
    "#     entropy=0.6,\n",
    "#     entropyk=4,\n",
    "#     entropywindow=24,\n",
    "#     maskentropy=True,\n",
    "#     ziplevel=9,\n",
    "# )\n",
    "\n",
    "# print(f\"Created masked SILVA rRNA database: {silva_masked}\")\n",
    "\n",
    "# # clean up\n",
    "# try:\n",
    "#     os.remove(deduplicated_fasta)\n",
    "#     os.remove(compressed_path)\n",
    "# except Exception as e:\n",
    "#     logger.warning(f\"Could not remove intermediate files: {e}\")\n",
    "\n",
    "print(f\"Masking sequences prepared in {masking_dir}\")\n",
    "\n",
    "# ##### Create rRNA DB #####\n",
    "# cd $rolypoly/data/\n",
    "# mkdir rRNA\n",
    "# cd rRNA\n",
    "# gzip SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz\n",
    "# gzip SILVA_138.1_LSURef_NR99_tax_silva.fasta.gz\n",
    "# cat *fasta > merged.fas\n",
    "\n",
    "# bbduk.sh -Xmx1g in=merged.fas out=merged_masked.fa zl=9 entropy=0.6 entropyk=4 entropywindow=24 maskentropy\n",
    "# bbduk.sh -Xmx1g in=rmdup_rRNA_ncbi.fasta  out=rmdup_rRNA_ncbi_masked.fa zl=9 entropy=0.6 entropyk=4 entropywindow=24 maskentropy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe943ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote combined rRNA fasta to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ncbi_rRNA_all_sequences.fasta\n"
     ]
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
