{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "222afdb2",
   "metadata": {},
   "source": [
    "# How much gets filtered when removing rRNA, tRNA, and mtDNA/RNA? \n",
    "# Notebook 1: fetching the data\n",
    "I'm making this notebook as a n intermission from preparing rolypoly external database. So far I've used a combination of NCBI +SILVA rRNAs to remove contaminating rRNA reads from RNA-seq data. Subsequently, the organisms whose rRNAs were most matched, are then also fetched (or their transcriptomes, if available) to remove any remaining reads that may have come from those organisms.  \n",
    "THis is messy, requires NCBI taxdump, taxonkit, and ncbi-datasets. So I started removing this dependency by using a set of rRNAs for which I can generate a prebuilt table containing the FTP addresses of the hosts' genomes/transcriptomes.  \n",
    "While doing so, I realised the step above could be split - quick rRNA mapping to get rough taxonomic breakdown of the sample, and then a more thorough removal of rRNA, tRNA, and mtDNA/RNA using a more comprehensive database. The question is then how much will these diffrent combinations filter, how much more time, and would masking the fasta for subsequences shared with RNA viruses change the results significantly?\n",
    "\n",
    "To test these, below are how I got the data, how I named the sets. For simplicity, the [trrna_fetch.ipynb](trrna_fetch.ipynb) has the code for the actual filtering using these sequence typs, graphs and so on. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b6f89d",
   "metadata": {},
   "source": [
    "*Note*: Parts of this script are from [build_data.py](../src/rolypoly/commands/misc/build_data.py) and [filter_reads.py](../src/rolypoly/commands/reads/filter_reads.py) scripts in the rolypoly repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57818007",
   "metadata": {},
   "source": [
    "Loading libraries and defining paths to sets already created/downloaded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a12c02e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting data preparation to : /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\n"
     ]
    }
   ],
   "source": [
    "import datetime\n",
    "import json\n",
    "import logging\n",
    "import shutil\n",
    "import subprocess\n",
    "import tarfile\n",
    "from pathlib import Path as pt\n",
    "import aria2p\n",
    "import tempfile\n",
    "import os\n",
    "import time\n",
    "import ftplib\n",
    "from ftplib import FTP\n",
    "import glob\n",
    "from collections import defaultdict, deque\n",
    "import heapq\n",
    "\n",
    "import polars as pl\n",
    "import requests\n",
    "from tqdm.notebook import tqdm\n",
    "from bbmapy import bbduk, bbmask, kcompress\n",
    "from rich.console import Console\n",
    "from rich_click import command, option\n",
    "\n",
    "from rolypoly.utils.bio.alignments import (\n",
    "    hmmdb_from_directory,\n",
    "    mmseqs_profile_db_from_directory,\n",
    ")\n",
    "from rolypoly.utils.bio.sequences import (\n",
    "    filter_fasta_by_headers,\n",
    "    write_fasta_file,\n",
    "    remove_duplicates\n",
    ")\n",
    "\n",
    "from rolypoly.utils.bio.trees import TaxonomyTree\n",
    "\n",
    "from rolypoly.utils.bio.polars_fastx import from_fastx_eager\n",
    "\n",
    "from rolypoly.utils.logging.citation_reminder import remind_citations\n",
    "from rolypoly.utils.logging.loggit import get_version_info, setup_logging\n",
    "from rolypoly.utils.various import fetch_and_extract, run_command_comp,extract\n",
    "\n",
    "### DEBUG ARGS (for manually building, not entering via CLI):\n",
    "threads = 6\n",
    "log_file = \"notebooks/Exprimental/trrna.log\"\n",
    "data_dir = \"/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\"\n",
    "\n",
    "global rrna_dir\n",
    "global contam_dir\n",
    "\n",
    "logger = setup_logging(log_file)\n",
    "print(f\"Starting data preparation to : {data_dir}\")\n",
    "\n",
    "contam_dir = os.path.join(data_dir, \"contam\")\n",
    "os.makedirs(contam_dir, exist_ok=True)\n",
    "\n",
    "rrna_dir = os.path.join(contam_dir, \"rrna\")\n",
    "os.makedirs(rrna_dir, exist_ok=True)\n",
    "\n",
    "trna_dir = os.path.join(contam_dir, \"trna\")\n",
    "os.makedirs(trna_dir, exist_ok=True)\n",
    "\n",
    "masking_dir = os.path.join(contam_dir, \"masking\")\n",
    "os.makedirs(masking_dir, exist_ok=True)\n",
    "\n",
    "# taxonomy_dir = os.path.join(data_dir, \"taxdump\")\n",
    "# os.makedirs(taxonomy_dir, exist_ok=True)\n",
    "\n",
    "reference_seqs = os.path.join(data_dir, \"reference_seqs\")\n",
    "os.makedirs(reference_seqs, exist_ok=True)\n",
    "\n",
    "mmseqs_ref_dir = os.path.join(reference_seqs, \"mmseqs\")\n",
    "os.makedirs(mmseqs_ref_dir, exist_ok=True)\n",
    "\n",
    "rvmt_dir = os.path.join(reference_seqs, \"RVMT\")\n",
    "os.makedirs(rvmt_dir, exist_ok=True)\n",
    "\n",
    "ncbi_ribovirus_dir = os.path.join(reference_seqs, \"ncbi_ribovirus\")\n",
    "os.makedirs(ncbi_ribovirus_dir, exist_ok=True)\n",
    "\n",
    "# Masking sequences preparation\n",
    "rvmt_fasta_path = os.path.join(\n",
    "    data_dir, \"reference_seqs\", \"RVMT\", \"RVMT_cleaned_contigs.fasta\"\n",
    ")\n",
    "ncbi_ribovirus_fasta_path = os.path.join(\n",
    "    data_dir,\n",
    "    \"reference_seqs\",\n",
    "    \"ncbi_ribovirus\",\n",
    "    \"refseq_ribovirus_genomes.fasta\",\n",
    ")\n",
    "\n",
    "rna_viruses_entropy_masked_path = os.path.join(\n",
    "    masking_dir, \"combined_entropy_masked.fasta\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2aceda4",
   "metadata": {},
   "source": [
    "## Fetching data (SILVA, NCBI rRNAs, tRNAs, mtDNAs):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "689a1346",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:03:55</span> - Downloading                                                                                              <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_SSURef_NR99_tax_silva.fasta.gz</span> to         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:03:55\u001b[0m - Downloading                                                                                              \u001b]8;id=364914;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=664156;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[4;94mhttps://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_SSURef_NR99_tax_silva.fasta.gz\u001b[0m to         \u001b[2m              \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:15</span> - Successfully downloaded to                                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:15\u001b[0m - Successfully downloaded to                                                                               \u001b]8;id=58595;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=229357;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:16</span> - Extracting <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta.gz</span>   <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">255</span></a>\n",
       "         to <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">rrna</span>                                                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:16\u001b[0m - Extracting \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta.gz\u001b[0m   \u001b]8;id=107698;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=689189;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\u001b\\\u001b[2m255\u001b[0m\u001b]8;;\u001b\\\n",
       "         to \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/\u001b[0m\u001b[95mrrna\u001b[0m                                                  \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:20</span> - Successfully extracted to                                                                                <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">288</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta</span>                                       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:20\u001b[0m - Successfully extracted to                                                                                \u001b]8;id=665527;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=456689;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\u001b\\\u001b[2m288\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta\u001b[0m                                       \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:20</span> - Downloading                                                                                              <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_LSURef_NR99_tax_silva.fasta.gz</span> to         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:20\u001b[0m - Downloading                                                                                              \u001b]8;id=271910;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=99333;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[4;94mhttps://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_LSURef_NR99_tax_silva.fasta.gz\u001b[0m to         \u001b[2m              \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:37</span> - Successfully downloaded to                                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:37\u001b[0m - Successfully downloaded to                                                                               \u001b]8;id=829389;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=472864;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:37</span> - Extracting <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta.gz</span>   <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">255</span></a>\n",
       "         to <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">rrna</span>                                                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:37\u001b[0m - Extracting \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta.gz\u001b[0m   \u001b]8;id=356328;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=146753;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\u001b\\\u001b[2m255\u001b[0m\u001b]8;;\u001b\\\n",
       "         to \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/\u001b[0m\u001b[95mrrna\u001b[0m                                                  \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:04:38</span> - Successfully extracted to                                                                                <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">288</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta</span>                                       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:04:38\u001b[0m - Successfully extracted to                                                                                \u001b]8;id=577559;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=80398;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\u001b\\\u001b[2m288\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta\u001b[0m                                       \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total SILVA99 sequences 605774\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (4, 3)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>header</th><th>sequence</th><th>rRNA_type</th></tr><tr><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;AY846379.1.1791 Eukaryota;Arch…</td><td>&quot;AACCUGGUUGAUCCUGCCAGUAGUCAUAUG…</td><td>&quot;SSU&quot;</td></tr><tr><td>&quot;AY846382.1.1778 Eukaryota;Arch…</td><td>&quot;GUUGAUCCUGCCAGUAGUCAUAUGCUUGUC…</td><td>&quot;SSU&quot;</td></tr><tr><td>&quot;AB000393.1.1510 Bacteria;Pseud…</td><td>&quot;UGGCUCAGAUUGAACGCUGGCGGCAGGCCU…</td><td>&quot;SSU&quot;</td></tr><tr><td>&quot;AY909590.1.2352 Eukaryota;Arch…</td><td>&quot;GACUAAGCCAUGCAUGUCUAAGUAUAAACG…</td><td>&quot;SSU&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (4, 3)\n",
       "┌─────────────────────────────────┬─────────────────────────────────┬───────────┐\n",
       "│ header                          ┆ sequence                        ┆ rRNA_type │\n",
       "│ ---                             ┆ ---                             ┆ ---       │\n",
       "│ str                             ┆ str                             ┆ str       │\n",
       "╞═════════════════════════════════╪═════════════════════════════════╪═══════════╡\n",
       "│ AY846379.1.1791 Eukaryota;Arch… ┆ AACCUGGUUGAUCCUGCCAGUAGUCAUAUG… ┆ SSU       │\n",
       "│ AY846382.1.1778 Eukaryota;Arch… ┆ GUUGAUCCUGCCAGUAGUCAUAUGCUUGUC… ┆ SSU       │\n",
       "│ AB000393.1.1510 Bacteria;Pseud… ┆ UGGCUCAGAUUGAACGCUGGCGGCAGGCCU… ┆ SSU       │\n",
       "│ AY909590.1.2352 Eukaryota;Arch… ┆ GACUAAGCCAUGCAUGUCUAAGUAUAAACG… ┆ SSU       │\n",
       "└─────────────────────────────────┴─────────────────────────────────┴───────────┘"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silva_release = \"138.2\"\n",
    "\n",
    "# Download SILVA rRNA sequences (SSU and LSU)\n",
    "silva_ssu_path = os.path.join(\n",
    "    rrna_dir, f\"SILVA_{silva_release}_SSURef_NR99_tax_silva.fasta\"\n",
    ")\n",
    "silva_lsu_path = os.path.join(\n",
    "    rrna_dir, f\"SILVA_{silva_release}_LSURef_NR99_tax_silva.fasta\"\n",
    ")\n",
    "# skipping downloading (already done)\n",
    "fetch_and_extract(\n",
    "    f\"https://www.arb-silva.de/fileadmin/silva_databases/release_{silva_release.replace('.', '_')}/Exports/SILVA_{silva_release}_SSURef_NR99_tax_silva.fasta.gz\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"tmp_ssu.fasta.gz\"),\n",
    "    extract_to=rrna_dir,\n",
    "    rename_extracted=silva_ssu_path,\n",
    "    logger=logger,\n",
    ")\n",
    "fetch_and_extract(\n",
    "    f\"https://www.arb-silva.de/fileadmin/silva_databases/release_{silva_release.replace('.', '_')}/Exports/SILVA_{silva_release}_LSURef_NR99_tax_silva.fasta.gz\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"tmp_lsu.fasta.gz\"),\n",
    "    extract_to=rrna_dir,\n",
    "    rename_extracted=silva_lsu_path,\n",
    "    logger=logger,\n",
    ")\n",
    "\n",
    "# Download SILVA taxonomy mappings (maps accessions to NCBI taxids)\n",
    "silva_ssu_taxmap = pl.read_csv(\n",
    "    \"https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/taxonomy/ncbi/taxmap_embl-ebi_ena_ssu_ref_nr99_138.2.txt.gz\",\n",
    "    truncate_ragged_lines=True,\n",
    "    separator=\"\\t\",\n",
    "    infer_schema_length=123123,\n",
    ")\n",
    "silva_lsu_taxmap = pl.read_csv(\n",
    "    \"https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/taxonomy/ncbi/taxmap_embl-ebi_ena_lsu_ref_nr99_138.2.txt.gz\",\n",
    "    truncate_ragged_lines=True,\n",
    "    separator=\"\\t\",\n",
    "    infer_schema_length=123123,\n",
    ")\n",
    "silva_taxmap = pl.concat([silva_lsu_taxmap, silva_ssu_taxmap])\n",
    "silva_taxmap.write_parquet(os.path.join(rrna_dir, \"silva_taxmap_embl-ebi_ena.parquet\"))\n",
    "\n",
    "# Parse SILVA headers and extract accessions\n",
    "silva_fasta_df = pl.concat(\n",
    "    [\n",
    "        from_fastx_eager(silva_ssu_path).with_columns(\n",
    "            pl.lit(\"SSU\").alias(\"rRNA_type\")\n",
    "        ),\n",
    "        from_fastx_eager(silva_lsu_path).with_columns(\n",
    "            pl.lit(\"LSU\").alias(\"rRNA_type\")\n",
    "        ),\n",
    "    ]\n",
    ")\n",
    "silva_fasta_df.write_parquet(os.path.join(rrna_dir, \"silva99_fasta.parquet\"))\n",
    "print(f\"total SILVA99 sequences {silva_fasta_df.height}\")\n",
    "silva_fasta_df.head(4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e52b38f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged taxonomy for 605774 SILVA sequences\n",
      "Total of 105645 unique NCBI taxids found in SILVA sequences\n"
     ]
    }
   ],
   "source": [
    "# Extract accession from header (format: >accession.version rest_of_header)\n",
    "silva_fasta_df = silva_fasta_df.with_columns(\n",
    "    primaryAccession=pl.col(\"header\").str.extract(\n",
    "        r\"^([A-Za-z0-9_]+)(?:\\.\\d+)*\", 1\n",
    "    ),  # DQ150555.1.2478 -> DQ150555\n",
    "    accession=pl.col(\"header\").str.extract(\n",
    "        r\"^([A-Za-z0-9_]+(?:\\.\\d+)?)\", 1\n",
    "    ),  # AY846379 or DQ150555.1\n",
    "    taxonomy_raw=pl.col(\"header\").str.replace(r\"^\\S+\\s+\", \"\"),\n",
    ")\n",
    "    # silva_fasta_df = silva_fasta_df.with_columns(\n",
    "    #     pl.col(\"sequence\").str.len_chars().alias(\"seq_length\")\n",
    "    # )\n",
    "    # silva_taxmap = silva_taxmap.with_columns(\n",
    "    #     (pl.col(\"stop\") - pl.col(\"start\")).alias(\"seq_length\")\n",
    "    # )\n",
    "\n",
    "silva_df = silva_fasta_df.join(\n",
    "    silva_taxmap.select(\n",
    "        [\"primaryAccession\", \"ncbi_taxonid\", \"submitted_path\"]\n",
    "    ).unique(),  # seq_length\n",
    "    on=[\"primaryAccession\"],\n",
    "    how=\"inner\",\n",
    ")\n",
    "silva_df.write_parquet(os.path.join(rrna_dir, \"silva_rrna_sequences.parquet\"))\n",
    "# silva_df.height\n",
    "silva_df[\"ncbi_taxonid\"].null_count()\n",
    "\n",
    "# Load SILVA taxonomy mappings\n",
    "print(\n",
    "    f\"Merged taxonomy for {silva_df.filter(pl.col('ncbi_taxonid').is_not_null()).height} SILVA sequences\"\n",
    ")\n",
    "\n",
    "unique_taxids = (\n",
    "    silva_df.filter(pl.col(\"ncbi_taxonid\").is_not_null())\n",
    "    .select(\"ncbi_taxonid\")\n",
    "    .unique()[\"ncbi_taxonid\"]\n",
    "    .to_list()\n",
    ")\n",
    "print(\n",
    "    f\"Total of {len(unique_taxids)} unique NCBI taxids found in SILVA sequences\"\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fca0a78d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# silva_df = pl.read_parquet(os.path.join(rrna_dir, \"silva_rrna_sequences.parquet\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff910e05",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-12-01 14:15:15,584 - rolypoly.utils.logging.loggit - INFO - Downloading https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/assembly_summary_genbank.txt.gz\n",
      "2025-12-01 14:15:45,465 - rolypoly.utils.logging.loggit - INFO - Successfully downloaded to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/assembly_summary_genbank.txt.gz\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">01</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">14:15:45</span> - Loading NCBI GenBank assembly summary                                                                   <a href=\"file:///tmp/ipykernel_2638087/3124954572.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">3124954572.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///tmp/ipykernel_2638087/3124954572.py#7\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">7</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m01\u001b[0m \u001b[1;92m14:15:45\u001b[0m - Loading NCBI GenBank assembly summary                                                                   \u001b]8;id=290238;file:///tmp/ipykernel_2638087/3124954572.py\u001b\\\u001b[2m3124954572.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=190348;file:///tmp/ipykernel_2638087/3124954572.py#7\u001b\\\u001b[2m7\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Generate FTP download URLs for host genomes/transcriptomes\n",
    "fetch_and_extract(\n",
    "    url=\"https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"assembly_summary_genbank.txt.gz\"),\n",
    "    extract=False,\n",
    ")\n",
    "print(\"Loading NCBI GenBank assembly summary\")\n",
    "# genbank_summary = pl.read_csv(os.path.join(rrna_dir, \"assembly_summary_genbank.txt.gz\",),\n",
    "# infer_schema_length=100020, separator=\"\\t\", skip_rows=1,\n",
    "# null_values=[\"na\",\"NA\",\"-\"],ignore_errors=True,\n",
    "# has_header=True)\n",
    "# polars failed me, so using line by line iterator\n",
    "from gzip import open as gz_open\n",
    "with gz_open(\n",
    "    os.path.join(rrna_dir, \"assembly_summary_genbank.txt.gz\"), \"r\"\n",
    ") as f:\n",
    "    header = None\n",
    "    records = []\n",
    "    i = 0\n",
    "    for line in f:\n",
    "        if i == 0:\n",
    "            i += 1\n",
    "            continue\n",
    "        line = line.rstrip(b\"\\n\")\n",
    "        if i == 1:\n",
    "            header = line.decode()[1:].strip().split(\"\\t\")\n",
    "            i += 1\n",
    "            continue\n",
    "        fields = line.decode().strip().split(\"\\t\")\n",
    "        record = dict(zip(header, fields))\n",
    "        records.append(record)\n",
    "genbank_summary = pl.from_records(records).rename({\"taxid\": \"ncbi_taxonid\"})\n",
    "genbank_summary.collect_schema()\n",
    "    # Schema([('assembly_accession', String),\n",
    "#         ('bioproject', String),\n",
    "#         ('biosample', String),\n",
    "#         ('wgs_master', String),\n",
    "#         ('refseq_category', String),\n",
    "#         ('ncbi_taxonid', String),\n",
    "#         ('species_taxid', String),\n",
    "#         ('organism_name', String),\n",
    "#         ('infraspecific_name', String),\n",
    "#         ('isolate', String),\n",
    "#         ('version_status', String),\n",
    "#         ('assembly_level', String),\n",
    "#         ('release_type', String),\n",
    "#         ('genome_rep', String),\n",
    "#         ('seq_rel_date', String),\n",
    "#         ('asm_name', String),\n",
    "#         ('asm_submitter', String),\n",
    "#         ('gbrs_paired_asm', String),\n",
    "#         ('paired_asm_comp', String),\n",
    "#         ('ftp_path', String),\n",
    "#         ('excluded_from_refseq', String),\n",
    "#         ('relation_to_type_material', String),\n",
    "#         ('asm_not_live_date', String),\n",
    "#         ('assembly_type', String),\n",
    "#         ('group', String),\n",
    "#         ('genome_size', String),\n",
    "#         ('genome_size_ungapped', String),\n",
    "#         ('gc_percent', String),\n",
    "#         ('replicon_count', String),\n",
    "#         ('scaffold_count', String),\n",
    "#         ('contig_count', String),\n",
    "#         ('annotation_provider', String),\n",
    "#         ('annotation_name', String),\n",
    "#         ('annotation_date', String),\n",
    "#         ('total_gene_count', String),\n",
    "#         ('protein_coding_gene_count', String),\n",
    "#         ('non_coding_gene_count', String),\n",
    "#         ('pubmed_id', String)])\n",
    "\n",
    "genbank_summary.write_csv(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.tsv\"), separator=\"\\t\"\n",
    ")\n",
    "genbank_summary = pl.read_csv(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.tsv\"),\n",
    "    infer_schema_length=100020,\n",
    "    separator=\"\\t\",\n",
    "    null_values=[\"na\", \"NA\", \"-\"],\n",
    "    ignore_errors=True,\n",
    "    has_header=True,\n",
    ")\n",
    "\n",
    "# In [91]: genbank_summary.collect_schema()\n",
    "# Out[91]: \n",
    "# Schema([('assembly_accession', String),\n",
    "#         ('bioproject', String),\n",
    "#         ('biosample', String),\n",
    "#         ('wgs_master', String),\n",
    "#         ('refseq_category', String),\n",
    "#         ('ncbi_taxonid', Int64),\n",
    "#         ('species_taxid', Int64),\n",
    "#         ('organism_name', String),\n",
    "#         ('infraspecific_name', String),\n",
    "#         ('isolate', String),\n",
    "#         ('version_status', String),\n",
    "#         ('assembly_level', String),\n",
    "#         ('release_type', String),\n",
    "#         ('genome_rep', String),\n",
    "#         ('seq_rel_date', String),\n",
    "#         ('asm_name', String),\n",
    "#         ('asm_submitter', String),\n",
    "#         ('gbrs_paired_asm', String),\n",
    "#         ('paired_asm_comp', String),\n",
    "#         ('ftp_path', String),\n",
    "#         ('excluded_from_refseq', String),\n",
    "#         ('relation_to_type_material', String),\n",
    "#         ('asm_not_live_date', String),\n",
    "#         ('assembly_type', String),\n",
    "#         ('group', String),\n",
    "#         ('genome_size', Int64),\n",
    "#         ('genome_size_ungapped', Int64),\n",
    "#         ('gc_percent', Float64),\n",
    "#         ('replicon_count', Int64),\n",
    "#         ('scaffold_count', Int64),\n",
    "#         ('contig_count', Int64),\n",
    "#         ('annotation_provider', String),\n",
    "#         ('annotation_name', String),\n",
    "#         ('annotation_date', String),\n",
    "#         ('total_gene_count', Int64),\n",
    "#         ('protein_coding_gene_count', Int64),\n",
    "#         ('non_coding_gene_count', Int64),\n",
    "#         ('pubmed_id', String)])\n",
    "\n",
    "genbank_summary.write_parquet(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.parquet\")\n",
    ")\n",
    "genbank_summary.write_csv(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.tsv\"), separator=\"\\t\"\n",
    ")\n",
    "print(genbank_summary.head(4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c4571252",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "shape: (4, 38)\n",
      "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
      "│ assembly_ ┆ bioprojec ┆ biosample ┆ wgs_maste ┆ … ┆ total_gen ┆ protein_c ┆ non_codin ┆ pubmed_i │\n",
      "│ accession ┆ t         ┆ ---       ┆ r         ┆   ┆ e_count   ┆ oding_gen ┆ g_gene_co ┆ d        │\n",
      "│ ---       ┆ ---       ┆ str       ┆ ---       ┆   ┆ ---       ┆ e_count   ┆ unt       ┆ ---      │\n",
      "│ str       ┆ str       ┆           ┆ str       ┆   ┆ i64       ┆ ---       ┆ ---       ┆ str      │\n",
      "│           ┆           ┆           ┆           ┆   ┆           ┆ i64       ┆ i64       ┆          │\n",
      "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
      "│ GCA_00000 ┆ PRJNA1381 ┆ SAMN02803 ┆ null      ┆ … ┆ 17872     ┆ 13962     ┆ 3543      ┆ 10731132 │\n",
      "│ 1215.4    ┆ 2         ┆ 731       ┆           ┆   ┆           ┆           ┆           ┆ ;1253756 │\n",
      "│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ 8;125375 │\n",
      "│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ 72;125…  │\n",
      "│ GCA_00000 ┆ PRJNA3125 ┆ null      ┆ null      ┆ … ┆ 37        ┆ 13        ┆ 24        ┆ 7219534; │\n",
      "│ 1405.29   ┆ 7         ┆           ┆           ┆   ┆           ┆           ┆           ┆ 10508508 │\n",
      "│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ ;1083095 │\n",
      "│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ 3;1123…  │\n",
      "│ GCA_00000 ┆ PRJNA1318 ┆ SAMN02981 ┆ AACZ00000 ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 16136131 │\n",
      "│ 1515.5    ┆ 4         ┆ 217       ┆ 000.4     ┆   ┆           ┆           ┆           ┆ ;1613613 │\n",
      "│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ 4        │\n",
      "│ GCA_00000 ┆ PRJNA2086 ┆ SAMN02981 ┆ ABGA00000 ┆ … ┆ 0         ┆ 0         ┆ 0         ┆ 21270892 │\n",
      "│ 1545.3    ┆ 9         ┆ 238       ┆ 000.1     ┆   ┆           ┆           ┆           ┆          │\n",
      "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘\n"
     ]
    }
   ],
   "source": [
    "genbank_summary = pl.read_parquet(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.parquet\")\n",
    ")\n",
    "print(genbank_summary.head(4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "18135e03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered GenBank summary to 242839 unique taxid entries for SILVA sequences\n",
      "30503\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1340307/1389620562.py:15: DeprecationWarning: `is_in` with a collection of the same datatype is ambiguous and deprecated.\n",
      "Please use `implode` to return to previous behavior.\n",
      "\n",
      "See https://github.com/pola-rs/polars/issues/22149 for more information.\n",
      "  in_silva = mini_genebank.filter(pl.col(\"ncbi_taxonid\").is_in(silva_taxmap[\"ncbi_taxonid\"])).unique()\n"
     ]
    }
   ],
   "source": [
    "# next, for every unique ncbi_taxonid, we select the one that has the most protein_coding_gene_count, then refseq_category, then tie breaking with non_coding_gene_count, tie breaking by latest assembly (by seq_rel_date).\n",
    "mini_genebank = genbank_summary.sort(\n",
    "    by=[\n",
    "        pl.col(\"refseq_category\").reverse(),\n",
    "        pl.col(\"protein_coding_gene_count\").cast(pl.Int64).reverse(),\n",
    "        pl.col(\"non_coding_gene_count\").cast(pl.Int64).reverse(),\n",
    "        pl.col(\"seq_rel_date\").reverse(),\n",
    "        pl.col(\"genome_size\").reverse(),\n",
    "    ]\n",
    ").unique(subset=[\"ncbi_taxonid\"], keep=\"first\")\n",
    "print(\n",
    "    f\"Filtered GenBank summary to {mini_genebank.height} unique taxid entries for SILVA sequences\"\n",
    ")\n",
    "\n",
    "in_silva = mini_genebank.filter(pl.col(\"ncbi_taxonid\").is_in(silva_taxmap[\"ncbi_taxonid\"])).unique()\n",
    "print(in_silva.height)\n",
    "# print(temp_genbank.height)\n",
    "# only 30503 out of ~240k?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d6f0c4dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'9a59e7252082bf546b2f84a6212753ac  ./annotation_hashes.txt\\nca4b9086529f58d55f3d10c3816aabd2  ./GCA_038318695.1_ASM3831869v1_assembly_report.txt\\n397cdca0cf495a861533c2fda3008de9  ./GCA_038318695.1_ASM3831869v1_assembly_stats.txt\\nfaad4a79de2e96ac4ba06ed789db6d16  ./GCA_038318695.1_ASM3831869v1_cds_from_genomic.fna.gz\\n770f9d368a473e0f00ce627f0ef0db6d  ./GCA_038318695.1_ASM3831869v1_feature_count.txt\\n0b259388786231db1f212e9cc1011f9a  ./GCA_038318695.1_ASM3831869v1_feature_table.txt.gz\\n03be2534ed6cbf9de002af189cf2459f  ./GCA_038318695.1_ASM3831869v1_genomic.fna.gz\\ne410d6ee6a2050ea00c31b78ad12ff1e  ./GCA_038318695.1_ASM3831869v1_genomic.gbff.gz\\ne69e781c501af782ae27053603c2e7a2  ./GCA_038318695.1_ASM3831869v1_genomic.gff.gz\\ne57ed066b6be493266857fb24fea6058  ./GCA_038318695.1_ASM3831869v1_genomic.gtf.gz\\n069d32032455e11ffbf4cc3aa489e1f9  ./GCA_038318695.1_ASM3831869v1_protein.faa.gz\\nae3785cd2f95bf874bb47924f542f9e4  ./GCA_038318695.1_ASM3831869v1_protein.gpff.gz\\na2e7f7b60c56b39c78902c708a3ac160  ./GCA_038318695.1_ASM3831869v1_translated_cds.faa.gz\\n'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# next, for every unique entry, we create paths for the expected location of the transcripts/transcriptome fasta, the rna fasta, and the genome.\n",
    "mini_genebank = mini_genebank.with_columns(\n",
    "        (pl.col(\"ftp_path\") + \"/md5checksums.txt\").alias(\"md5sums_path\")\n",
    ")\n",
    "# fetching the md5sums files for a subset to test time\n",
    "test_df = mini_genebank[:10].with_columns(\n",
    "        pl.col(\"ftp_path\").map_elements(\n",
    "            lambda path: requests.get(path + \"/md5checksums.txt\").text,return_dtype=pl.Utf8\n",
    "        ).alias(\"md5sums_content\")\n",
    ")\n",
    "test_df[\"md5sums_content\"].first()\n",
    "# ~2.2s, too slow, so will do it with aria2 using FTP \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f26b83b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing to download 242833 MD5 checksum files in parallel using FTP protocol\n",
      "Sample FTP URL: ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/018/101/015/GCA_018101015.1_ASM1810101v1/md5checksums.txt\n",
      "Queuing FTP downloads...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0889c7dc75af4dfc929b8dfd0187df5d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Queuing FTP downloads:   0%|          | 0/242833 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Queued 242833 FTP downloads. Waiting for aria2 to complete all downloads...\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Create a temporary directory for downloads\n",
    "temp_dir = tempfile.mkdtemp()\n",
    "\n",
    "# Start aria2c daemon (assumes aria2c is running with RPC enabled)\n",
    "api = aria2p.API(\n",
    "    aria2p.Client(\n",
    "        host=\"http://localhost\",\n",
    "        port=6800,\n",
    "        secret=\"\"\n",
    "    )\n",
    ")\n",
    "# print(\"Make sure aria2c is running with: aria2c --enable-rpc --rpc-listen-all=true --rpc-allow-origin-all\")\n",
    "\n",
    "def convert_to_ftp_url(http_ftp_path):\n",
    "    \"\"\"Convert NCBI HTTP FTP path to proper FTP URL\"\"\"\n",
    "    if http_ftp_path.startswith(\"https://ftp.ncbi.nlm.nih.gov/\"):\n",
    "        # Convert https://ftp.ncbi.nlm.nih.gov/path to ftp://ftp.ncbi.nlm.nih.gov/path\n",
    "        return http_ftp_path.replace(\"https://\", \"ftp://\")\n",
    "    elif http_ftp_path.startswith(\"http://ftp.ncbi.nlm.nih.gov/\"):\n",
    "        # Convert http://ftp.ncbi.nlm.nih.gov/path to ftp://ftp.ncbi.nlm.nih.gov/path\n",
    "        return http_ftp_path.replace(\"http://\", \"ftp://\")\n",
    "    elif http_ftp_path.startswith(\"ftp://\"):\n",
    "        return http_ftp_path\n",
    "    else:\n",
    "        # If it doesn't start with a protocol, assume it's a path that needs ftp:// prefix\n",
    "        if http_ftp_path.startswith(\"/\"):\n",
    "            return f\"ftp://ftp.ncbi.nlm.nih.gov{http_ftp_path}\"\n",
    "        else:\n",
    "            return f\"ftp://ftp.ncbi.nlm.nih.gov/{http_ftp_path}\"\n",
    "\n",
    "# Create download URLs for md5checksum files with unique filenames using FTP protocol\n",
    "md5_urls = []\n",
    "url_to_filename = {}  # Map URLs to their unique filenames\n",
    "filename_to_ftp_path = {}  # Map filenames back to FTP paths\n",
    "\n",
    "for i, ftp_path in enumerate(mini_genebank[\"ftp_path\"]):  # [:100] Test with first 100 entries - 3.3s, quick, so running on all\n",
    "    if ftp_path and ftp_path != \"na\":\n",
    "        # Convert to proper FTP URL and append md5checksums.txt\n",
    "        ftp_url = convert_to_ftp_url(ftp_path)\n",
    "        md5_url = ftp_url + \"/md5checksums.txt\"\n",
    "        \n",
    "        # Create unique filename using index to avoid conflicts\n",
    "        unique_filename = f\"md5checksums_{i:06d}.txt\"\n",
    "        \n",
    "        md5_urls.append(md5_url)\n",
    "        url_to_filename[md5_url] = unique_filename\n",
    "        filename_to_ftp_path[unique_filename] = ftp_path\n",
    "\n",
    "print(f\"Preparing to download {len(md5_urls)} MD5 checksum files in parallel using FTP protocol\")\n",
    "print(f\"Sample FTP URL: {md5_urls[0] if md5_urls else 'No URLs generated'}\")\n",
    "\n",
    "# FTP-optimized aria2 options\n",
    "ftp_options = {\n",
    "    \"dir\": temp_dir,\n",
    "    \"max-connection-per-server\": \"10\",  # FTP servers often limit connections\n",
    "    \"split\": \"1\",  # Don't split small text files\n",
    "    \"ftp-reuse-connection\": \"true\",  # Reuse FTP connections for efficiency\n",
    "    \"ftp-pasv\": \"true\",  # Use passive FTP mode (better for firewalls)\n",
    "    \"timeout\": \"60\",  # 60 second timeout for FTP connections\n",
    "    \"retry-wait\": \"3\",  # Wait 3 seconds between retries\n",
    "    \"max-tries\": \"3\",  # Try up to 3 times per file\n",
    "    \"connect-timeout\": \"15\",  # 30 second connection timeout\n",
    "    \"max-concurrent-downloads\": \"10\",  # Limit total concurrent downloads\n",
    "}\n",
    "\n",
    "# Queue all downloads with progress bar\n",
    "downloads = []\n",
    "print(\"Queuing FTP downloads...\")\n",
    "for url in tqdm(md5_urls, desc=\"Queuing FTP downloads\"):\n",
    "    try:\n",
    "        unique_filename = url_to_filename[url]\n",
    "        download_options = ftp_options.copy()\n",
    "        download_options[\"out\"] = unique_filename  # Specify unique output filename\n",
    "        download = api.add_uris([url], options=download_options)\n",
    "        downloads.append((url, download, unique_filename))\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to queue download for {url}: {e}\")\n",
    "\n",
    "print(f\"Queued {len(downloads)} FTP downloads. Waiting for aria2 to complete all downloads...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ee270fa3",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'temp_dir' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m test_df \u001b[38;5;241m=\u001b[39m pl\u001b[38;5;241m.\u001b[39mscan_csv(\u001b[43mtemp_dir\u001b[49m\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/*.txt\u001b[39m\u001b[38;5;124m\"\u001b[39m,  separator\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m, include_file_paths\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile_path\u001b[39m\u001b[38;5;124m\"\u001b[39m, has_header\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,schema\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmd5\u001b[39m\u001b[38;5;124m\"\u001b[39m:pl\u001b[38;5;241m.\u001b[39mUtf8,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdropme\u001b[39m\u001b[38;5;124m\"\u001b[39m:pl\u001b[38;5;241m.\u001b[39mBoolean,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mfile_path_on_ftp\u001b[39m\u001b[38;5;124m\"\u001b[39m:pl\u001b[38;5;241m.\u001b[39mUtf8}\n\u001b[1;32m      2\u001b[0m )\u001b[38;5;241m.\u001b[39mcollect()\n\u001b[1;32m      3\u001b[0m test_df \u001b[38;5;241m=\u001b[39m test_df\u001b[38;5;241m.\u001b[39mdrop(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdropme\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      4\u001b[0m test_df\n",
      "\u001b[0;31mNameError\u001b[0m: name 'temp_dir' is not defined"
     ]
    }
   ],
   "source": [
    "# test_df = pl.scan_csv(temp_dir+\"/*.txt\",  separator=\" \", include_file_paths=\"file_path\", has_header=False,schema={\"md5\":pl.Utf8,\"dropme\":pl.Boolean,\"file_path_on_ftp\":pl.Utf8}\n",
    "# ).collect()\n",
    "# test_df = test_df.drop(\"dropme\")\n",
    "# test_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d79da07e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Wait for all aria2 downloads to complete\n",
    "# # Check aria2 global stats to see when all downloads are done\n",
    "# for thing in tqdm(enumerate(downloads), desc=\"waiting for downloads to complete\"):\n",
    "#     stats = api.get_stats()\n",
    "#     active = int(stats.num_active)\n",
    "#     waiting = int(stats.num_waiting)\n",
    "    \n",
    "#     if active == 0 and waiting == 0:\n",
    "#         print(\"All downloads completed!\")\n",
    "#         break\n",
    "#     tqdm.update()\n",
    "#     time.sleep(2)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "9b4534b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing downloaded files...\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "94f309fe9963405babba092f62b0f242",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Reading downloaded files:   0%|          | 0/242833 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "✅ Successfully processed 60579 MD5 checksum files\n",
      "❌ Failed/missing files: 182254\n",
      "❌ Failed/missing example url: ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/038/506/625/GCA_038506625.1_ASM3850662v1/md5checksums.txt\n",
      "📊 Total attempted: 242833\n",
      "📋 Sample of first entry: ba7bb432c803cec66a415cf43ae54ef6  ./annotation_hashes.txt\n",
      "d364bab756fa59130f83f23bface8fc1  ./GCA_018101015.1_ASM1810101v1_assembly_report.txt\n",
      "f501bb76bb9b9d7bf21d5acf8c373195  ./GCA_018101015.1_ASM18\n",
      "🔗 Mapping verification: 60579 entries in final dictionary\n"
     ]
    }
   ],
   "source": [
    "# Now read all the downloaded files\n",
    "ftp_path_md5_content = {}\n",
    "completed_count = 0\n",
    "failed_count = 0\n",
    "failed_exmp =\"\"\n",
    "print(\"Processing downloaded files...\")\n",
    "for url, download, unique_filename in tqdm(downloads, desc=\"Reading downloaded files\"):\n",
    "    file_path = os.path.join(temp_dir, unique_filename)\n",
    "    \n",
    "    if os.path.exists(file_path):\n",
    "        try:\n",
    "            with open(file_path, 'r') as f:\n",
    "                content = f.read()\n",
    "            \n",
    "            # Map FTP path to content using our mapping\n",
    "            ftp_path = filename_to_ftp_path[unique_filename]\n",
    "            ftp_path_md5_content[ftp_path] = content\n",
    "            completed_count += 1\n",
    "        except Exception as e:\n",
    "            print(f\"Error reading downloaded file {file_path}: {e}\")\n",
    "            failed_count += 1\n",
    "            # print(url)\n",
    "    else:\n",
    "        # print(f\"Downloaded file not found: {file_path}\")\n",
    "        failed_count += 1\n",
    "        failed_exmp = url\n",
    "        # print(url)\n",
    "\n",
    "# # Clean up temporary files\n",
    "# try:\n",
    "#     import shutil\n",
    "#     shutil.rmtree(temp_dir)\n",
    "# except Exception as e:\n",
    "#     print(f\"Failed to clean up temp directory: {e}\")\n",
    "\n",
    "print(f\"\\n✅ Successfully processed {completed_count} MD5 checksum files\")\n",
    "print(f\"❌ Failed/missing files: {failed_count}\")\n",
    "print(f\"❌ Failed/missing example url: {failed_exmp}\")\n",
    "print(f\"📊 Total attempted: {len(md5_urls)}\")\n",
    "print(f\"📋 Sample of first entry: {list(ftp_path_md5_content.values())[0][:200] if ftp_path_md5_content else 'No content'}\")\n",
    "print(f\"🔗 Mapping verification: {len(ftp_path_md5_content)} entries in final dictionary\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a68a739",
   "metadata": {},
   "outputs": [],
   "source": [
    "mini_genebank = mini_genebank.with_columns(\n",
    "        (pl.col(\"ftp_path\") + \"/md5checksums.txt\").alias(\"md5sums_path\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41c17114",
   "metadata": {},
   "outputs": [],
   "source": [
    "minimini= mini_genebank[:10]\n",
    "test_df = [pl.scan_csv(md5sums_path, sep=\"\\t\", has_header=False) for md5sums_path in minimini[\"md5sums_path\"].to_list()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c1ae24b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ftp_path_md5_content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40bc1550",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge SILVA sequences and apply entropy masking\n",
    "print(\"Merging and masking SILVA sequences\")\n",
    "silva_merged = os.path.join(rrna_dir, \"SILVA_merged.fasta\")\n",
    "silva_masked = os.path.join(rrna_dir, \"SILVA_merged_masked.fasta\")\n",
    "\n",
    "# Concatenate SILVA files\n",
    "run_command_comp(\n",
    "    base_cmd=\"cat\",\n",
    "    positional_args=[silva_ssu_path, silva_lsu_path],\n",
    "    positional_args_location=\"end\",\n",
    "    params={},\n",
    "    output_file=silva_merged,\n",
    "    logger=logger,\n",
    ")\n",
    "\n",
    "# Apply entropy masking\n",
    "bbduk(\n",
    "    in1=silva_merged,\n",
    "    out=silva_masked,\n",
    "    entropy=0.6,\n",
    "    entropyk=4,\n",
    "    entropywindow=24,\n",
    "    maskentropy=True,\n",
    "    ziplevel=9,\n",
    ")\n",
    "\n",
    "print(f\"Created masked SILVA rRNA database: {silva_masked}\")\n",
    "\n",
    "# # clean up\n",
    "# try:\n",
    "#     os.remove(deduplicated_fasta)\n",
    "#     os.remove(compressed_path)\n",
    "# except Exception as e:\n",
    "#     logger.warning(f\"Could not remove intermediate files: {e}\")\n",
    "\n",
    "print(f\"Masking sequences prepared in {masking_dir}\")\n",
    "\n",
    "# ##### Create rRNA DB #####\n",
    "# cd $rolypoly/data/\n",
    "# mkdir rRNA\n",
    "# cd rRNA\n",
    "# gzip SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz\n",
    "# gzip SILVA_138.1_LSURef_NR99_tax_silva.fasta.gz\n",
    "# cat *fasta > merged.fas\n",
    "\n",
    "# bbduk.sh -Xmx1g in=merged.fas out=merged_masked.fa zl=9 entropy=0.6 entropyk=4 entropywindow=24 maskentropy\n",
    "# bbduk.sh -Xmx1g in=rmdup_rRNA_ncbi.fasta  out=rmdup_rRNA_ncbi_masked.fa zl=9 entropy=0.6 entropyk=4 entropywindow=24 maskentropy\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a058d83",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "4008b431",
   "metadata": {},
   "source": [
    "## rRNAs, tRNA, mtDNA, and plastid-DNA from NCBI BLAST DBs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9bf454",
   "metadata": {},
   "source": [
    "This is a little hacky, but works?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ed29977",
   "metadata": {},
   "source": [
    "rRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b6991ebb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Connected to ftp.ncbi.nlm.nih.gov\n",
      "\n",
      "--- Directory Contents ---\n",
      "Downloading 18S_fungal_sequences.tar.gz\n",
      "Downloading 18S_fungal_sequences.tar.gz.md5\n",
      "Downloading 28S_fungal_sequences.tar.gz\n",
      "Downloading 28S_fungal_sequences.tar.gz.md5\n",
      "Downloading ITS_RefSeq_Fungi.tar.gz\n",
      "Downloading ITS_RefSeq_Fungi.tar.gz.md5\n",
      "Downloading ITS_eukaryote_sequences.tar.gz\n",
      "Downloading ITS_eukaryote_sequences.tar.gz.md5\n",
      "Downloading LSU_eukaryote_rRNA.tar.gz\n",
      "Downloading LSU_eukaryote_rRNA.tar.gz.md5\n",
      "Downloading LSU_prokaryote_rRNA.tar.gz\n",
      "Downloading LSU_prokaryote_rRNA.tar.gz.md5\n",
      "Downloading SSU_eukaryote_rRNA.tar.gz\n",
      "Downloading SSU_eukaryote_rRNA.tar.gz.md5\n",
      "Downloading 16S_ribosomal_RNA.tar.gz\n",
      "Downloading 16S_ribosomal_RNA.tar.gz.md5\n",
      "Downloading 16S_ribosomal_RNA-nucl-metadata.json\n",
      "Downloading 18S_fungal_sequences-nucl-metadata.json\n",
      "Downloading 28S_fungal_sequences-nucl-metadata.json\n",
      "Downloading ITS_RefSeq_Fungi-nucl-metadata.json\n",
      "Downloading ITS_eukaryote_sequences-nucl-metadata.json\n",
      "Downloading LSU_eukaryote_rRNA-nucl-metadata.json\n",
      "Downloading LSU_prokaryote_rRNA-nucl-metadata.json\n",
      "Downloading SSU_eukaryote_rRNA-nucl-metadata.json\n",
      "Downloaded 24 files matching terms: ['16S', '18S', '28S', 'ITS', 'LSU', 'SSU', 'ribosomal']\n"
     ]
    }
   ],
   "source": [
    "# first, getting the contents of https://ftp.ncbi.nlm.nih.gov/blast/db/ and filtering to files containing \"16S\", \"18S\", \"28S\", \"ITS\" \n",
    "# LSU, SSU\n",
    "\n",
    "terms = [\"16S\", \"18S\", \"28S\", \"ITS\", \"LSU\", \"SSU\", \"ribosomal\"]\n",
    "files_2_download = []\n",
    "FTP_HOST = \"ftp.ncbi.nlm.nih.gov\"\n",
    "FTP_DIR = \"/blast/db/\"\n",
    "\n",
    "# List to store the file and directory names\n",
    "file_list = []\n",
    "\n",
    "# Connect to the FTP host\n",
    "with FTP(FTP_HOST) as ftp:\n",
    "    #  Log in anonymously (default behavior for .login())\n",
    "    ftp.login()\n",
    "    print(f\"Connected to {FTP_HOST}\")\n",
    "    ftp.cwd(FTP_DIR)\n",
    "    logger.debug(f\"Changed directory to {FTP_DIR}\")\n",
    "\n",
    "    # nlst() returns only names, not detailed information\n",
    "    file_list = ftp.nlst()\n",
    "    \n",
    "    print(\"\\n--- Directory Contents ---\")\n",
    "    for item in file_list:\n",
    "        if any(term in item for term in terms):\n",
    "            files_2_download.append(item)\n",
    "            logger.debug(item)\n",
    "            \n",
    "            # Download file directly using FTP\n",
    "            local_path = os.path.join(rrna_dir, item)\n",
    "            print(f\"Downloading {item}\")\n",
    "            with open(local_path, 'wb') as local_file:\n",
    "                ftp.retrbinary(f'RETR {item}', local_file.write)\n",
    "            logger.debug(f\"Successfully downloaded {item}\")\n",
    "\n",
    "print(f\"Downloaded {len(files_2_download)} files matching terms: {terms}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ca9f924a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 18S_fungal_sequences.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/18S_fungal_sequences.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/18S_fungal_sequences.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:09</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">18S_fungal_sequences</span> -out                          <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">18S_fungal_sequences.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:09\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=79157;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=776658;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95m18S_fungal_sequences\u001b[0m -out                          \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95m18S_fungal_sequences.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m     \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/18S_fungal_sequences.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">5700653</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/18S_fungal_sequences.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m5700653\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 28S_fungal_sequences.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/28S_fungal_sequences.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/28S_fungal_sequences.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:14</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">28S_fungal_sequences</span> -out                          <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">28S_fungal_sequences.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:14\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=744864;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=166510;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95m28S_fungal_sequences\u001b[0m -out                          \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95m28S_fungal_sequences.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m     \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/28S_fungal_sequences.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">12514460</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/28S_fungal_sequences.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m12514460\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing ITS_RefSeq_Fungi.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_RefSeq_Fungi.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_RefSeq_Fungi.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:19</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">ITS_RefSeq_Fungi</span> -out                              <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">ITS_RefSeq_Fungi.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:19\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=207760;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=341108;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mITS_RefSeq_Fungi\u001b[0m -out                              \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mITS_RefSeq_Fungi.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m         \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_RefSeq_Fungi.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">13138739</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_RefSeq_Fungi.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m13138739\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing ITS_eukaryote_sequences.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_eukaryote_sequences.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_eukaryote_sequences.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:26</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">ITS_eukaryote_sequences</span> -out                       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">ITS_eukaryote_sequences.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:26\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=332926;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=361501;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mITS_eukaryote_sequences\u001b[0m -out                       \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mITS_eukaryote_sequences.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m  \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_eukaryote_sequences.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">is </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">64900379</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ITS_eukaryote_sequences.tab'\u001b[0m\u001b[32m size \u001b[0m\n",
       "\u001b[32mis \u001b[0m\u001b[1;32m64900379\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing LSU_eukaryote_rRNA.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_eukaryote_rRNA.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_eukaryote_rRNA.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:32</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">LSU_eukaryote_rRNA</span> -out                            <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">LSU_eukaryote_rRNA.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:32\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=767122;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=354467;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mLSU_eukaryote_rRNA\u001b[0m -out                            \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mLSU_eukaryote_rRNA.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m       \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_eukaryote_rRNA.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">14610798</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_eukaryote_rRNA.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m14610798\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing LSU_prokaryote_rRNA.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_prokaryote_rRNA.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_prokaryote_rRNA.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:35</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">LSU_prokaryote_rRNA</span> -out                           <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">LSU_prokaryote_rRNA.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:35\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=663831;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=4494;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mLSU_prokaryote_rRNA\u001b[0m -out                           \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mLSU_prokaryote_rRNA.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m      \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_prokaryote_rRNA.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">12127093</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/LSU_prokaryote_rRNA.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m12127093\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing SSU_eukaryote_rRNA.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/SSU_eukaryote_rRNA.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/SSU_eukaryote_rRNA.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:38</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">SSU_eukaryote_rRNA</span> -out                            <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">SSU_eukaryote_rRNA.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:38\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=776650;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=697132;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mSSU_eukaryote_rRNA\u001b[0m -out                            \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mSSU_eukaryote_rRNA.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m       \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/SSU_eukaryote_rRNA.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">16578954</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/SSU_eukaryote_rRNA.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m16578954\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 16S_ribosomal_RNA.tar.gz...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">Successfully decompressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/16S_ribosomal_RNA.tar.gz'</span><span style=\"color: #008000; text-decoration-color: #008000\"> to </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mSuccessfully decompressed \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/16S_ribosomal_RNA.tar.gz'\u001b[0m\u001b[32m to \u001b[0m\n",
       "\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna'\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">04</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">15:08:43</span> - Running command: blastdbcmd   -entry all -db                                                            <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">1257</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">16S_ribosomal_RNA</span> -out                             <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">16S_ribosomal_RNA.tab</span> -outfmt <span style=\"color: #008000; text-decoration-color: #008000\">\"%T@%t@%K@%s\"</span>        <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m04\u001b[0m \u001b[1;92m15:08:43\u001b[0m - Running command: blastdbcmd   -entry all -db                                                            \u001b]8;id=593720;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=580490;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#1257\u001b\\\u001b[2m1257\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95m16S_ribosomal_RNA\u001b[0m -out                             \u001b[2m               \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95m16S_ribosomal_RNA.tab\u001b[0m -outfmt \u001b[32m\"%T@%t@%K@%s\"\u001b[0m        \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/16S_ribosomal_RNA.tab'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">42549204</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/16S_ribosomal_RNA.tab'\u001b[0m\u001b[32m size is \u001b[0m\n",
       "\u001b[1;32m42549204\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# for each of these, will run blastdbcmd to extract sequences into fasta \n",
    "for item in files_2_download:\n",
    "    if not item.endswith('.tar.gz'):\n",
    "        continue\n",
    "    print(f\"Processing {item}...\")\n",
    "    extract(archive_path=os.path.join(rrna_dir, item),\n",
    "        extract_to=rrna_dir,\n",
    "    )\n",
    "    # blastdbcmd -entry all -db LSU_prokaryote_rRNA  -out reference.fasta -outfmt \"%T;%t;%s\" # taxid;header;sequence\n",
    "    run_command_comp(\n",
    "        base_cmd=\"blastdbcmd\",\n",
    "        positional_args=[\n",
    "            \"-entry\",\n",
    "            \"all\",\n",
    "            \"-db\",\n",
    "            os.path.join(rrna_dir, item.replace('.tar.gz','')),\n",
    "            \"-out\",\n",
    "            os.path.join(rrna_dir, item.replace('.tar.gz','').replace('.tar','') + '.tab'),\n",
    "            \"-outfmt\",\n",
    "            str(f'\"%T@%t@%K@%s\"') ,\n",
    "        ],\n",
    "        positional_args_location=\"end\",\n",
    "        params={},\n",
    "        logger=logger,\n",
    "        output_file=os.path.join(rrna_dir, item.replace('.tar.gz','').replace('.tar','') + '.tab')\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "78a669bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 99 files to remove\n",
      "Keeping 23 files (.tab, .fasta, .parquet, .tsv, .gz)\n",
      "Removed: 16S_ribosomal_RNA-nucl-metadata.json\n",
      "Removed: 16S_ribosomal_RNA.ndb\n",
      "Removed: 16S_ribosomal_RNA.nhr\n",
      "Removed: 16S_ribosomal_RNA.nin\n",
      "Removed: 16S_ribosomal_RNA.nnd\n",
      "Removed: 16S_ribosomal_RNA.nni\n",
      "Removed: 16S_ribosomal_RNA.nog\n",
      "Removed: 16S_ribosomal_RNA.nos\n",
      "Removed: 16S_ribosomal_RNA.not\n",
      "Removed: 16S_ribosomal_RNA.nsq\n",
      "Removed: 16S_ribosomal_RNA.ntf\n",
      "Removed: 16S_ribosomal_RNA.nto\n",
      "Removed: 16S_ribosomal_RNA.tar.gz.md5\n",
      "Removed: 18S_fungal_sequences-nucl-metadata.json\n",
      "Removed: 18S_fungal_sequences.ndb\n",
      "Removed: 18S_fungal_sequences.nhr\n",
      "Removed: 18S_fungal_sequences.nin\n",
      "Removed: 18S_fungal_sequences.nnd\n",
      "Removed: 18S_fungal_sequences.nni\n",
      "Removed: 18S_fungal_sequences.nog\n",
      "Removed: 18S_fungal_sequences.nos\n",
      "Removed: 18S_fungal_sequences.not\n",
      "Removed: 18S_fungal_sequences.nsq\n",
      "Removed: 18S_fungal_sequences.ntf\n",
      "Removed: 18S_fungal_sequences.nto\n",
      "Removed: 18S_fungal_sequences.tar.gz.md5\n",
      "Removed: 28S_fungal_sequences-nucl-metadata.json\n",
      "Removed: 28S_fungal_sequences.ndb\n",
      "Removed: 28S_fungal_sequences.nhr\n",
      "Removed: 28S_fungal_sequences.nin\n",
      "Removed: 28S_fungal_sequences.nnd\n",
      "Removed: 28S_fungal_sequences.nni\n",
      "Removed: 28S_fungal_sequences.nog\n",
      "Removed: 28S_fungal_sequences.nos\n",
      "Removed: 28S_fungal_sequences.not\n",
      "Removed: 28S_fungal_sequences.nsq\n",
      "Removed: 28S_fungal_sequences.ntf\n",
      "Removed: 28S_fungal_sequences.nto\n",
      "Removed: 28S_fungal_sequences.tar.gz.md5\n",
      "Removed: ITS_RefSeq_Fungi-nucl-metadata.json\n",
      "Removed: ITS_RefSeq_Fungi.ndb\n",
      "Removed: ITS_RefSeq_Fungi.nhr\n",
      "Removed: ITS_RefSeq_Fungi.nin\n",
      "Removed: ITS_RefSeq_Fungi.nnd\n",
      "Removed: ITS_RefSeq_Fungi.nni\n",
      "Removed: ITS_RefSeq_Fungi.nog\n",
      "Removed: ITS_RefSeq_Fungi.nos\n",
      "Removed: ITS_RefSeq_Fungi.not\n",
      "Removed: ITS_RefSeq_Fungi.nsq\n",
      "Removed: ITS_RefSeq_Fungi.ntf\n",
      "Removed: ITS_RefSeq_Fungi.nto\n",
      "Removed: ITS_RefSeq_Fungi.tar.gz.md5\n",
      "Removed: ITS_eukaryote_sequences-nucl-metadata.json\n",
      "Removed: ITS_eukaryote_sequences.ndb\n",
      "Removed: ITS_eukaryote_sequences.nhr\n",
      "Removed: ITS_eukaryote_sequences.nin\n",
      "Removed: ITS_eukaryote_sequences.nog\n",
      "Removed: ITS_eukaryote_sequences.nos\n",
      "Removed: ITS_eukaryote_sequences.not\n",
      "Removed: ITS_eukaryote_sequences.nsq\n",
      "Removed: ITS_eukaryote_sequences.ntf\n",
      "Removed: ITS_eukaryote_sequences.nto\n",
      "Removed: ITS_eukaryote_sequences.tar.gz.md5\n",
      "Removed: LSU_eukaryote_rRNA-nucl-metadata.json\n",
      "Removed: LSU_eukaryote_rRNA.ndb\n",
      "Removed: LSU_eukaryote_rRNA.nhr\n",
      "Removed: LSU_eukaryote_rRNA.nin\n",
      "Removed: LSU_eukaryote_rRNA.nog\n",
      "Removed: LSU_eukaryote_rRNA.nos\n",
      "Removed: LSU_eukaryote_rRNA.not\n",
      "Removed: LSU_eukaryote_rRNA.nsq\n",
      "Removed: LSU_eukaryote_rRNA.ntf\n",
      "Removed: LSU_eukaryote_rRNA.nto\n",
      "Removed: LSU_eukaryote_rRNA.tar.gz.md5\n",
      "Removed: LSU_prokaryote_rRNA-nucl-metadata.json\n",
      "Removed: LSU_prokaryote_rRNA.ndb\n",
      "Removed: LSU_prokaryote_rRNA.nhr\n",
      "Removed: LSU_prokaryote_rRNA.nin\n",
      "Removed: LSU_prokaryote_rRNA.nog\n",
      "Removed: LSU_prokaryote_rRNA.nos\n",
      "Removed: LSU_prokaryote_rRNA.not\n",
      "Removed: LSU_prokaryote_rRNA.nsq\n",
      "Removed: LSU_prokaryote_rRNA.ntf\n",
      "Removed: LSU_prokaryote_rRNA.nto\n",
      "Removed: LSU_prokaryote_rRNA.tar.gz.md5\n",
      "Removed: SSU_eukaryote_rRNA-nucl-metadata.json\n",
      "Removed: SSU_eukaryote_rRNA.ndb\n",
      "Removed: SSU_eukaryote_rRNA.nhr\n",
      "Removed: SSU_eukaryote_rRNA.nin\n",
      "Removed: SSU_eukaryote_rRNA.nog\n",
      "Removed: SSU_eukaryote_rRNA.nos\n",
      "Removed: SSU_eukaryote_rRNA.not\n",
      "Removed: SSU_eukaryote_rRNA.nsq\n",
      "Removed: SSU_eukaryote_rRNA.ntf\n",
      "Removed: SSU_eukaryote_rRNA.nto\n",
      "Removed: SSU_eukaryote_rRNA.tar.gz.md5\n",
      "Removed: taxdb.btd\n",
      "Removed: taxdb.bti\n",
      "Removed: taxonomy4blast.sqlite3\n",
      "Cleanup complete\n"
     ]
    }
   ],
   "source": [
    "# get all files to remove (everything but the .tab files)\n",
    "all_files_in_rrna = glob.glob(os.path.join(rrna_dir, \"*\"))\n",
    "all_tab_files = glob.glob(os.path.join(rrna_dir, \"*.tab\"))\n",
    "all_fasta_files = glob.glob(os.path.join(rrna_dir, \"*.fasta\"))\n",
    "\n",
    "# Files to keep (tab files, fasta files, and parquet/tsv files)\n",
    "files_to_keep = set(all_tab_files + all_fasta_files + \n",
    "                   glob.glob(os.path.join(rrna_dir, \"*.parquet\")) +\n",
    "                   glob.glob(os.path.join(rrna_dir, \"*.tsv\")) +\n",
    "                   glob.glob(os.path.join(rrna_dir, \"*.gz\")))\n",
    "\n",
    "# Get all files to remove (exclude directories and files to keep)\n",
    "files_2_remove = [f for f in all_files_in_rrna if os.path.isfile(f) and f not in files_to_keep]\n",
    "\n",
    "print(f\"Found {len(files_2_remove)} files to remove\")\n",
    "print(f\"Keeping {len(files_to_keep)} files (.tab, .fasta, .parquet, .tsv, .gz)\")\n",
    "\n",
    "# Remove the files\n",
    "for file_path in files_2_remove:\n",
    "    try:\n",
    "        os.remove(file_path)\n",
    "        print(f\"Removed: {os.path.basename(file_path)}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to remove {os.path.basename(file_path)}: {e}\")\n",
    "\n",
    "print(\"Cleanup complete\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c9b854f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 159996 rRNA sequences from\n"
     ]
    }
   ],
   "source": [
    "rrna_df = pl.scan_csv(\n",
    "    os.path.join(rrna_dir, \"*.tab\"),\n",
    "    separator=\"@\",\n",
    "    has_header=False,\n",
    "    null_values = [\"N/A\"],\n",
    "    new_columns=[\"taxid\", \"header\",\"name\",\"sequence\"],\n",
    "    include_file_paths=\"type\"\n",
    ").collect()\n",
    "rrna_df = rrna_df.drop(\"name\")\n",
    "rrna_df = rrna_df.with_columns(\n",
    "    rRNA_type=pl.col(\"type\").str.extract(r\"([^/]+)\\.tab$\", 1)\n",
    ")\n",
    "rrna_df = rrna_df.drop(\"type\")\n",
    "print(f\"Read {rrna_df.height} rRNA sequences from\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "143e16a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>rRNA_type</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;ITS_eukaryote_sequences&quot;</td><td>77582</td></tr><tr><td>&quot;16S_ribosomal_RNA&quot;</td><td>27708</td></tr><tr><td>&quot;ITS_RefSeq_Fungi&quot;</td><td>19541</td></tr><tr><td>&quot;28S_fungal_sequences&quot;</td><td>11754</td></tr><tr><td>&quot;SSU_eukaryote_rRNA&quot;</td><td>8784</td></tr><tr><td>&quot;LSU_eukaryote_rRNA&quot;</td><td>6575</td></tr><tr><td>&quot;LSU_prokaryote_rRNA&quot;</td><td>4047</td></tr><tr><td>&quot;18S_fungal_sequences&quot;</td><td>4005</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8, 2)\n",
       "┌─────────────────────────┬───────┐\n",
       "│ rRNA_type               ┆ count │\n",
       "│ ---                     ┆ ---   │\n",
       "│ str                     ┆ u32   │\n",
       "╞═════════════════════════╪═══════╡\n",
       "│ ITS_eukaryote_sequences ┆ 77582 │\n",
       "│ 16S_ribosomal_RNA       ┆ 27708 │\n",
       "│ ITS_RefSeq_Fungi        ┆ 19541 │\n",
       "│ 28S_fungal_sequences    ┆ 11754 │\n",
       "│ SSU_eukaryote_rRNA      ┆ 8784  │\n",
       "│ LSU_eukaryote_rRNA      ┆ 6575  │\n",
       "│ LSU_prokaryote_rRNA     ┆ 4047  │\n",
       "│ 18S_fungal_sequences    ┆ 4005  │\n",
       "└─────────────────────────┴───────┘"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrna_df[\"rRNA_type\"].value_counts(sort = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cfe943ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote combined rRNA fasta to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ncbi_rRNA_all_sequences.fasta\n"
     ]
    }
   ],
   "source": [
    "rrna_fasta_path = os.path.join(rrna_dir, \"ncbi_rRNA_all_sequences.fasta\")\n",
    "write_fasta_file(\n",
    "    seqs=rrna_df[\"sequence\"].to_list(),\n",
    "    headers=rrna_df[\"header\"].to_list(),\n",
    "    output_file=rrna_fasta_path,\n",
    ")\n",
    "print(f\"Wrote combined rRNA fasta to {rrna_fasta_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "33db56b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (11, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>group</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;bacteria&quot;</td><td>125828</td></tr><tr><td>&quot;viral&quot;</td><td>88342</td></tr><tr><td>&quot;fungi&quot;</td><td>7870</td></tr><tr><td>&quot;invertebrate&quot;</td><td>7069</td></tr><tr><td>&quot;vertebrate_other&quot;</td><td>4354</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;plant&quot;</td><td>3385</td></tr><tr><td>&quot;vertebrate_mammalian&quot;</td><td>1124</td></tr><tr><td>&quot;protozoa&quot;</td><td>1123</td></tr><tr><td>&quot;metagenomes&quot;</td><td>133</td></tr><tr><td>&quot;other&quot;</td><td>21</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (11, 2)\n",
       "┌──────────────────────┬────────┐\n",
       "│ group                ┆ count  │\n",
       "│ ---                  ┆ ---    │\n",
       "│ str                  ┆ u32    │\n",
       "╞══════════════════════╪════════╡\n",
       "│ bacteria             ┆ 125828 │\n",
       "│ viral                ┆ 88342  │\n",
       "│ fungi                ┆ 7870   │\n",
       "│ invertebrate         ┆ 7069   │\n",
       "│ vertebrate_other     ┆ 4354   │\n",
       "│ …                    ┆ …      │\n",
       "│ plant                ┆ 3385   │\n",
       "│ vertebrate_mammalian ┆ 1124   │\n",
       "│ protozoa             ┆ 1123   │\n",
       "│ metagenomes          ┆ 133    │\n",
       "│ other                ┆ 21     │\n",
       "└──────────────────────┴────────┘"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mini_genebank[\"group\"].value_counts(sort=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f3cdd589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (25_099, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>organism_name</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>null</td><td>117964</td></tr><tr><td>&quot;Giardia duodenalis&quot;</td><td>127</td></tr><tr><td>&quot;Globisporangium heterothallicu…</td><td>93</td></tr><tr><td>&quot;Babesia bovis&quot;</td><td>65</td></tr><tr><td>&quot;Globisporangium sylvaticum&quot;</td><td>61</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;Andreaea rupestris&quot;</td><td>1</td></tr><tr><td>&quot;Tetraphis pellucida&quot;</td><td>1</td></tr><tr><td>&quot;Anncaliia algerae&quot;</td><td>1</td></tr><tr><td>&quot;Dictyocoela muelleri&quot;</td><td>1</td></tr><tr><td>&quot;Vairimorpha necatrix&quot;</td><td>1</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (25_099, 2)\n",
       "┌─────────────────────────────────┬────────┐\n",
       "│ organism_name                   ┆ count  │\n",
       "│ ---                             ┆ ---    │\n",
       "│ str                             ┆ u32    │\n",
       "╞═════════════════════════════════╪════════╡\n",
       "│ null                            ┆ 117964 │\n",
       "│ Giardia duodenalis              ┆ 127    │\n",
       "│ Globisporangium heterothallicu… ┆ 93     │\n",
       "│ Babesia bovis                   ┆ 65     │\n",
       "│ Globisporangium sylvaticum      ┆ 61     │\n",
       "│ …                               ┆ …      │\n",
       "│ Andreaea rupestris              ┆ 1      │\n",
       "│ Tetraphis pellucida             ┆ 1      │\n",
       "│ Anncaliia algerae               ┆ 1      │\n",
       "│ Dictyocoela muelleri            ┆ 1      │\n",
       "│ Vairimorpha necatrix            ┆ 1      │\n",
       "└─────────────────────────────────┴────────┘"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df = rrna_df.join(\n",
    "    mini_genebank.select(\n",
    "        [\"ncbi_taxonid\", \"organism_name\"]\n",
    "    ).unique(),  # seq_length\n",
    "    left_on=[\"taxid\"],\n",
    "    right_on=[\"ncbi_taxonid\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "temp_df[\"organism_name\"].value_counts(sort=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12737bc5",
   "metadata": {},
   "source": [
    "some 117k nulls, these are the ones that will need to patch to link to cloest related organism that does have genome/transcriptome data on genbank.  \n",
    "That is done in [rrna_genome_mapping_taxonomy.ipynb](rrna_genome_mapping_taxonomy.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7a6e65e",
   "metadata": {},
   "source": [
    "## Now tRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b6d2cb41",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-12-04 11:45:32,833 - rolypoly.utils.logging.loggit - INFO - Downloading https://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF00005.fa.gz to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna/RF00005.fa.gz\n",
      "2025-12-04 11:46:10,628 - rolypoly.utils.logging.loggit - INFO - Successfully downloaded to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna/RF00005.fa.gz\n",
      "2025-12-04 11:46:10,632 - rolypoly.utils.logging.loggit - INFO - Extracting /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna/RF00005.fa.gz to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna\n",
      "2025-12-04 11:46:14,336 - rolypoly.utils.logging.loggit - INFO - Successfully extracted to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna/tRNA_sequences.fasta\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded tRNA sequences to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna/tRNA_sequences.fasta\n"
     ]
    }
   ],
   "source": [
    "file_url = \"https://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF00005.fa.gz\"\n",
    "trna_seqs = os.path.join(trna_dir, \"tRNA_sequences.fasta\")\n",
    "gz_filename = \"RF00005.fa.gz\"\n",
    "fetch_and_extract(\n",
    "                url=file_url,\n",
    "                fetched_to=os.path.join(trna_dir, gz_filename),\n",
    "                extract_to=trna_dir,\n",
    "                expected_file=trna_seqs,\n",
    "            )\n",
    "print(f\"Downloaded tRNA sequences to {trna_seqs}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "515356d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-12-04 11:46:25,181 - rolypoly.utils.logging.loggit - INFO - Processed 5335982 records: 3028115 unique, 2307867 duplicates removed\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'total_records': 5335982,\n",
       " 'unique_records': 3028115,\n",
       " 'duplicates_removed': 2307867}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove duplicates\n",
    "deduplicated_fasta = os.path.join(trna_dir, \"tRNA_sequences_deduplicated.fasta\")\n",
    "remove_duplicates(\n",
    "    input_file=trna_seqs,\n",
    "    output_file=deduplicated_fasta,\n",
    "    return_stats=True,\n",
    "    by = \"seq\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2d02b442",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'min_length': 33,\n",
       " 'max_length': 271,\n",
       " 'mean_length': 72.89565752951918,\n",
       " 'median_length': 72.0,\n",
       " 'std_length': 8.179689610946584,\n",
       " 'total_length': 220736434,\n",
       " 'min_gc': 0.08823529411764706,\n",
       " 'max_gc': 0.813953488372093,\n",
       " 'mean_gc': 0.5516366297753061,\n",
       " 'median_gc': 0.5535714285714286,\n",
       " 'std_gc': 0.057055637448841706,\n",
       " 'total_sequences': 3028115}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from rolypoly.utils.bio.polars_fastx import compute_aggregate_stats,fasta_stats\n",
    "info_table = fasta_stats(deduplicated_fasta)\n",
    "# print(info_table)\n",
    "\n",
    "some_stats = compute_aggregate_stats(info_table,fields=[\"length\", \"gc_content\", \"n_content\"])\n",
    "some_stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d21a27f",
   "metadata": {},
   "source": [
    "I don't trust this (min len 30, min gc 8% ), so will do some light filteration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f927d279",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'min_length': 50,\n",
       " 'max_length': 248,\n",
       " 'mean_length': 72.91280093980956,\n",
       " 'median_length': 72.0,\n",
       " 'std_length': 8.152576143845652,\n",
       " 'total_length': 220644708,\n",
       " 'min_gc': 0.08823529411764706,\n",
       " 'max_gc': 0.813953488372093,\n",
       " 'mean_gc': 0.5516374569020241,\n",
       " 'median_gc': 0.5535714285714286,\n",
       " 'std_gc': 0.05704648222966359,\n",
       " 'total_sequences': 3026145}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_table = info_table.filter(\n",
    "    pl.col(\"length\").is_between(50,250),\n",
    "    pl.col(\"gc_content\") >= 0.01,\n",
    ")\n",
    "some_stats_filtered = compute_aggregate_stats(info_table,fields=[\"length\", \"gc_content\", \"n_content\"])\n",
    "some_stats_filtered\n",
    "# info_table.write_parquet(os.path.join(trna_dir, \"tRNA_sequences_deduplicated_stats.parquet\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d1bb6c0",
   "metadata": {},
   "source": [
    "looks a little better, will just accept that very low or very high GC content tRNAs exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bbd4ccc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote filtered tRNA sequences to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/trna/tRNA_sequences_deduplicated_filtered.fasta\n"
     ]
    }
   ],
   "source": [
    "write_fasta_file(\n",
    "    seqs=info_table[\"sequence\"].to_list(),\n",
    "    headers=info_table[\"header\"].to_list(),\n",
    "    output_file=os.path.join(trna_dir, \"tRNA_sequences_deduplicated_filtered.fasta\"),\n",
    ")\n",
    "print(f\"Wrote filtered tRNA sequences to {os.path.join(trna_dir, 'tRNA_sequences_deduplicated_filtered.fasta')}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b21a1429",
   "metadata": {},
   "source": [
    "## plastid and mitochondria genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "545d92dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rolypoly.commands.misc.build_data import prepare_plastid_data, prepare_mito_data\n",
    "prepare_plastid_data(data_dir, logger)\n",
    "prepare_mito_data(data_dir, logger)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b21a1429",
   "metadata": {},
   "source": [
    "## gene2accession (NCBI)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ffb2dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fetch_and_extract( url=\"http://ftp.ncbi.nlm.nih.gov/gene/DATA/gene2accession.gz\",\n",
    "#     fetched_to=os.path.join(rrna_dir, \"gene2accession.gz\"),\n",
    "#     extract=False,\n",
    "# )\n",
    "\n",
    "silva_df = silva_df.with_columns(\n",
    "    ncbi_taxonid=pl.col(\"ncbi_taxonid\").cast(pl.String)\n",
    ")\n",
    "\n",
    "silva_df1 = silva_df.join(\n",
    "    genbank_summary.select([\"ncbi_taxonid\", \"ftp_path\"]),\n",
    "    on=[\"ncbi_taxonid\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "silva_df1\n",
    "\n",
    "silva_df = silva_df.with_columns(\n",
    "    genome_ftp_url=pl.when(pl.col(\"ncbi_taxonid\").is_not_null())\n",
    "    .then(\n",
    "        pl.format(\n",
    "            \"https://ftp.ncbi.nlm.nih.gov/genomes/all/refseq/taxid_{}/\",\n",
    "            pl.col(\"ncbi_taxonid\"),\n",
    "        )\n",
    "    )\n",
    "    .otherwise(None),\n",
    "    datasets_api_url=pl.when(pl.col(\"ncbi_taxonid\").is_not_null())\n",
    "    .then(\n",
    "        pl.format(\n",
    "            \"https://api.ncbi.nlm.nih.gov/datasets/v2alpha/genome/taxon/{}/download?include_annotation_type=GENOME_FASTA,RNA_FASTA\",\n",
    "            pl.col(\"ncbi_taxonid\"),\n",
    "        )\n",
    "    )\n",
    "    .otherwise(None),\n",
    ")\n",
    "\n",
    "# Save metadata table\n",
    "metadata_output = os.path.join(rrna_dir, \"rrna_metadata.tsv\")\n",
    "silva_df.write_csv(metadata_output, separator=\"\\t\")\n",
    "print(\n",
    "    f\"Saved rRNA metadata table with {len(silva_df)} entries to {metadata_output}\"\n",
    ")\n",
    "\n",
    "gene2accession = pl.read_csv(\n",
    "    os.path.join(rrna_dir, \"gene2accession.gz\"),\n",
    "    separator=\"\\t\",\n",
    "    # skip_rows=1,\n",
    "    # infer_schema_length=100020,\n",
    "    null_values=[\"na\", \"NA\", \"-\"],\n",
    "    ignore_errors=True,\n",
    "    has_header=True,\n",
    "    # n_rows=100\n",
    ")\n",
    "gene2accession.write_parquet(os.path.join(rrna_dir, \"gene2accession.parquet\"))\n",
    "# gene2accession = pl.read_parquet(os.path.join(rrna_dir, \"gene2accession.parquet\"))\n",
    "# gene2accession.collect_schema()\n",
    "# Schema([('#tax_id', Int64),\n",
    "#     ('GeneID', Int64),\n",
    "#     ('status', String),\n",
    "#     ('RNA_nucleotide_accession.version', String),\n",
    "#     ('RNA_nucleotide_gi', String),\n",
    "#     ('protein_accession.version', String),\n",
    "#     ('protein_gi', Int64),\n",
    "#     ('genomic_nucleotide_accession.version', String),\n",
    "#     ('genomic_nucleotide_gi', Int64),\n",
    "#     ('start_position_on_the_genomic_accession', Int64),\n",
    "#     ('end_position_on_the_genomic_accession', Int64),\n",
    "#     ('orientation', String),\n",
    "#     ('assembly', String),\n",
    "#     ('mature_peptide_accession.version', String),\n",
    "#     ('mature_peptide_gi', String),\n",
    "#     ('Symbol', String)])\n",
    "gene2accession = gene2accession.rename({\"#tax_id\": \"ncbi_taxonid\"})\n",
    "test_df = gene2accession.filter(pl.col(\"ncbi_taxonid\").is_in(unique_taxids))\n",
    "test_df.height # 148449745\n",
    "test_df2 = gene2accession.select([\"ncbi_taxonid\",\"assembly\"]).unique()\n",
    "test_df2.height # 52548\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
