{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "222afdb2",
   "metadata": {},
   "source": [
    "# How much gets filtered when removing rRNA, tRNA, and mtDNA/RNA? \n",
    "# Notebook 1: fetching the data\n",
    "I'm making this notebook as a n intermission from preparing rolypoly external database. So far I've used a combination of NCBI +SILVA rRNAs to remove contaminating rRNA reads from RNA-seq data. Subsequently, the organisms whose rRNAs were most matched, are then also fetched (or their transcriptomes, if available) to remove any remaining reads that may have come from those organisms.  \n",
    "THis is messy, requires NCBI taxdump, taxonkit, and ncbi-datasets. So I started removing this dependency by using a set of rRNAs for which I can generate a prebuilt table containing the FTP addresses of the hosts' genomes/transcriptomes.  \n",
    "While doing so, I realised the step above could be split - quick rRNA mapping to get rough taxonomic breakdown of the sample, and then a more thorough removal of rRNA, tRNA, and mtDNA/RNA using a more comprehensive database. The question is then how much will these diffrent combinations filter, how much more time, and would masking the fasta for subsequences shared with RNA viruses change the results significantly?\n",
    "\n",
    "To test these, below are how I got the data, how I named the sets. For simplicity, the [trrna_fetch.ipynb](trrna_fetch.ipynb) has the code for the actual filtering using these sequence typs, graphs and so on. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47b6f89d",
   "metadata": {},
   "source": [
    "*Note*: Parts of this script are from [build_data.py](../src/rolypoly/commands/misc/build_data.py) and [filter_reads.py](../src/rolypoly/commands/reads/filter_reads.py) scripts in the rolypoly repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57818007",
   "metadata": {},
   "source": [
    "Loading libraries and defining paths to sets already created/downloaded:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a12c02e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n",
      "Starting data preparation to : /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import hashlib\n",
    "import os\n",
    "from ftplib import FTP\n",
    "import glob\n",
    "\n",
    "import polars as pl\n",
    "from bbmapy import bbduk\n",
    "\n",
    "from rolypoly.utils.bio.sequences import write_fasta_file, remove_duplicates\n",
    "\n",
    "\n",
    "from rolypoly.utils.bio.polars_fastx import from_fastx_eager\n",
    "\n",
    "from rolypoly.utils.logging.loggit import setup_logging\n",
    "from rolypoly.utils.various import fetch_and_extract, run_command_comp, extract\n",
    "\n",
    "### DEBUG ARGS (for manually building, not entering via CLI):\n",
    "threads = 6\n",
    "log_file = \"notebooks/Exprimental/trrna.log\"\n",
    "global data_dir\n",
    "data_dir = \"/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data\"\n",
    "os.environ[\"ROLYPOLY_DATA_DIR\"] = data_dir\n",
    "\n",
    "global rrna_dir\n",
    "global contam_dir\n",
    "\n",
    "logger = setup_logging(log_file)\n",
    "print(f\"Starting data preparation to : {data_dir}\")\n",
    "\n",
    "contam_dir = os.path.join(data_dir, \"contam\")\n",
    "os.makedirs(contam_dir, exist_ok=True)\n",
    "\n",
    "rrna_dir = os.path.join(contam_dir, \"rrna\")\n",
    "os.makedirs(rrna_dir, exist_ok=True)\n",
    "\n",
    "trna_dir = os.path.join(contam_dir, \"trna\")\n",
    "os.makedirs(trna_dir, exist_ok=True)\n",
    "\n",
    "masking_dir = os.path.join(contam_dir, \"masking\")\n",
    "os.makedirs(masking_dir, exist_ok=True)\n",
    "\n",
    "reference_seqs = os.path.join(data_dir, \"reference_seqs\")\n",
    "os.makedirs(reference_seqs, exist_ok=True)\n",
    "\n",
    "mmseqs_ref_dir = os.path.join(reference_seqs, \"mmseqs\")\n",
    "os.makedirs(mmseqs_ref_dir, exist_ok=True)\n",
    "\n",
    "rvmt_dir = os.path.join(reference_seqs, \"RVMT\")\n",
    "os.makedirs(rvmt_dir, exist_ok=True)\n",
    "\n",
    "ncbi_ribovirus_dir = os.path.join(reference_seqs, \"ncbi_ribovirus\")\n",
    "os.makedirs(ncbi_ribovirus_dir, exist_ok=True)\n",
    "\n",
    "# Masking sequences preparation\n",
    "rvmt_fasta_path = os.path.join(\n",
    "    data_dir, \"reference_seqs\", \"RVMT\", \"RVMT_cleaned_contigs.fasta\"\n",
    ")\n",
    "ncbi_ribovirus_fasta_path = os.path.join(\n",
    "    data_dir,\n",
    "    \"reference_seqs\",\n",
    "    \"ncbi_ribovirus\",\n",
    "    \"refseq_ribovirus_genomes.fasta\",\n",
    ")\n",
    "\n",
    "rna_viruses_entropy_masked_path = os.path.join(\n",
    "    masking_dir, \"combined_entropy_masked.fasta\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2aceda4",
   "metadata": {},
   "source": [
    "## Fetching data (SILVA, NCBI rRNAs, tRNAs, mtDNAs):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "689a1346",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:53:33</span> - Downloading                                                                                              <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_SSURef_NR99_tax_silva.fasta.gz</span> to         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:53:33\u001b[0m - Downloading                                                                                              \u001b]8;id=664730;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=737937;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[4;94mhttps://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_SSURef_NR99_tax_silva.fasta.gz\u001b[0m to         \u001b[2m              \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:01</span> - Successfully downloaded to                                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:01\u001b[0m - Successfully downloaded to                                                                               \u001b]8;id=74975;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=787093;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:01</span> - Extracting <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta.gz</span>   <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">255</span></a>\n",
       "         to <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">rrna</span>                                                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:01\u001b[0m - Extracting \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta.gz\u001b[0m   \u001b]8;id=479798;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=958292;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\u001b\\\u001b[2m255\u001b[0m\u001b]8;;\u001b\\\n",
       "         to \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/\u001b[0m\u001b[95mrrna\u001b[0m                                                  \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:05</span> - Successfully extracted to                                                                                <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">288</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_ssu.fasta</span>                                       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:05\u001b[0m - Successfully extracted to                                                                                \u001b]8;id=463229;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=894040;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\u001b\\\u001b[2m288\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_ssu.fasta\u001b[0m                                       \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:05</span> - Downloading                                                                                              <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_LSURef_NR99_tax_silva.fasta.gz</span> to         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:05\u001b[0m - Downloading                                                                                              \u001b]8;id=306869;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=962956;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[4;94mhttps://www.arb-silva.de/fileadmin/silva_databases/release_138_2/Exports/SILVA_138.2_LSURef_NR99_tax_silva.fasta.gz\u001b[0m to         \u001b[2m              \u001b[0m\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:13</span> - Successfully downloaded to                                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta.gz</span>                                    <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:13\u001b[0m - Successfully downloaded to                                                                               \u001b]8;id=789217;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=399592;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta.gz\u001b[0m                                    \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:13</span> - Extracting <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta.gz</span>   <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">255</span></a>\n",
       "         to <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">rrna</span>                                                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:13\u001b[0m - Extracting \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta.gz\u001b[0m   \u001b]8;id=673944;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=359225;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#255\u001b\\\u001b[2m255\u001b[0m\u001b]8;;\u001b\\\n",
       "         to \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/\u001b[0m\u001b[95mrrna\u001b[0m                                                  \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:54:15</span> - Successfully extracted to                                                                                <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">288</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_lsu.fasta</span>                                       <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:54:15\u001b[0m - Successfully extracted to                                                                                \u001b]8;id=769819;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=785533;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#288\u001b\\\u001b[2m288\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtmp_lsu.fasta\u001b[0m                                       \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "silva_release = \"138.2\"\n",
    "\n",
    "# Download SILVA rRNA sequences (SSU and LSU)\n",
    "silva_ssu_path = os.path.join(\n",
    "    rrna_dir, f\"SILVA_{silva_release}_SSURef_NR99_tax_silva.fasta\"\n",
    ")\n",
    "silva_lsu_path = os.path.join(\n",
    "    rrna_dir, f\"SILVA_{silva_release}_LSURef_NR99_tax_silva.fasta\"\n",
    ")\n",
    "# skipping downloading (already done)\n",
    "fetch_and_extract(\n",
    "    f\"https://www.arb-silva.de/fileadmin/silva_databases/release_{silva_release.replace('.', '_')}/Exports/SILVA_{silva_release}_SSURef_NR99_tax_silva.fasta.gz\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"tmp_ssu.fasta.gz\"),\n",
    "    extract_to=rrna_dir,\n",
    "    rename_extracted=silva_ssu_path,\n",
    "    logger=logger,\n",
    ")\n",
    "fetch_and_extract(\n",
    "    f\"https://www.arb-silva.de/fileadmin/silva_databases/release_{silva_release.replace('.', '_')}/Exports/SILVA_{silva_release}_LSURef_NR99_tax_silva.fasta.gz\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"tmp_lsu.fasta.gz\"),\n",
    "    extract_to=rrna_dir,\n",
    "    rename_extracted=silva_lsu_path,\n",
    "    logger=logger,\n",
    ")\n",
    "\n",
    "# Download SILVA taxonomy mappings (maps accessions to NCBI taxids)\n",
    "silva_ssu_taxmap = pl.read_csv(\n",
    "    \"https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/taxonomy/ncbi/taxmap_embl-ebi_ena_ssu_ref_nr99_138.2.txt.gz\",\n",
    "    truncate_ragged_lines=True,\n",
    "    separator=\"\\t\",\n",
    "    infer_schema_length=123123,\n",
    ")\n",
    "silva_lsu_taxmap = pl.read_csv(\n",
    "    \"https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/taxonomy/ncbi/taxmap_embl-ebi_ena_lsu_ref_nr99_138.2.txt.gz\",\n",
    "    truncate_ragged_lines=True,\n",
    "    separator=\"\\t\",\n",
    "    infer_schema_length=123123,\n",
    ")\n",
    "silva_taxmap = pl.concat([silva_lsu_taxmap, silva_ssu_taxmap])\n",
    "silva_taxmap.write_parquet(\n",
    "    os.path.join(rrna_dir, \"silva_taxmap_embl-ebi_ena.parquet\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "af54d242",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total SILVA99 sequences 605774\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (4, 3)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>header</th><th>sequence</th><th>rRNA_type</th></tr><tr><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;AY846379.1.1791 Eukaryota;Arch…</td><td>&quot;AACCUGGUUGAUCCUGCCAGUAGUCAUAUG…</td><td>&quot;SSU&quot;</td></tr><tr><td>&quot;AY846382.1.1778 Eukaryota;Arch…</td><td>&quot;GUUGAUCCUGCCAGUAGUCAUAUGCUUGUC…</td><td>&quot;SSU&quot;</td></tr><tr><td>&quot;AB000393.1.1510 Bacteria;Pseud…</td><td>&quot;UGGCUCAGAUUGAACGCUGGCGGCAGGCCU…</td><td>&quot;SSU&quot;</td></tr><tr><td>&quot;AY909590.1.2352 Eukaryota;Arch…</td><td>&quot;GACUAAGCCAUGCAUGUCUAAGUAUAAACG…</td><td>&quot;SSU&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (4, 3)\n",
       "┌─────────────────────────────────┬─────────────────────────────────┬───────────┐\n",
       "│ header                          ┆ sequence                        ┆ rRNA_type │\n",
       "│ ---                             ┆ ---                             ┆ ---       │\n",
       "│ str                             ┆ str                             ┆ str       │\n",
       "╞═════════════════════════════════╪═════════════════════════════════╪═══════════╡\n",
       "│ AY846379.1.1791 Eukaryota;Arch… ┆ AACCUGGUUGAUCCUGCCAGUAGUCAUAUG… ┆ SSU       │\n",
       "│ AY846382.1.1778 Eukaryota;Arch… ┆ GUUGAUCCUGCCAGUAGUCAUAUGCUUGUC… ┆ SSU       │\n",
       "│ AB000393.1.1510 Bacteria;Pseud… ┆ UGGCUCAGAUUGAACGCUGGCGGCAGGCCU… ┆ SSU       │\n",
       "│ AY909590.1.2352 Eukaryota;Arch… ┆ GACUAAGCCAUGCAUGUCUAAGUAUAAACG… ┆ SSU       │\n",
       "└─────────────────────────────────┴─────────────────────────────────┴───────────┘"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse SILVA headers and extract accessions\n",
    "silva_fasta_df = pl.concat(\n",
    "    [\n",
    "        from_fastx_eager(silva_ssu_path).with_columns(\n",
    "            pl.lit(\"SSU\").alias(\"rRNA_type\")\n",
    "        ),\n",
    "        from_fastx_eager(silva_lsu_path).with_columns(\n",
    "            pl.lit(\"LSU\").alias(\"rRNA_type\")\n",
    "        ),\n",
    "    ]\n",
    ")\n",
    "silva_fasta_df.write_parquet(os.path.join(rrna_dir, \"silva99_fasta.parquet\"))\n",
    "print(f\"total SILVA99 sequences {silva_fasta_df.height}\")\n",
    "silva_fasta_df.head(4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0e52b38f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged taxonomy for 605774 SILVA sequences\n",
      "Total of 105645 unique NCBI taxids found in SILVA sequences\n"
     ]
    }
   ],
   "source": [
    "# Extract accession from header (format: >accession.version rest_of_header)\n",
    "silva_fasta_df = silva_fasta_df.with_columns(\n",
    "    primaryAccession=pl.col(\"header\").str.extract(\n",
    "        r\"^([A-Za-z0-9_]+)(?:\\.\\d+)*\", 1\n",
    "    ),  # DQ150555.1.2478 -> DQ150555\n",
    "    accession=pl.col(\"header\").str.extract(\n",
    "        r\"^([A-Za-z0-9_]+(?:\\.\\d+)?)\", 1\n",
    "    ),  # AY846379 or DQ150555.1\n",
    "    taxonomy_raw=pl.col(\"header\").str.replace(r\"^\\S+\\s+\", \"\"),\n",
    ")\n",
    "# silva_fasta_df = silva_fasta_df.with_columns(\n",
    "#     pl.col(\"sequence\").str.len_chars().alias(\"seq_length\")\n",
    "# )\n",
    "# silva_taxmap = silva_taxmap.with_columns(\n",
    "#     (pl.col(\"stop\") - pl.col(\"start\")).alias(\"seq_length\")\n",
    "# )\n",
    "\n",
    "silva_df = silva_fasta_df.join(\n",
    "    silva_taxmap.select(\n",
    "        [\"primaryAccession\", \"ncbi_taxonid\", \"submitted_path\"]\n",
    "    ).unique(),  # seq_length\n",
    "    on=[\"primaryAccession\"],\n",
    "    how=\"inner\",\n",
    ")\n",
    "# silva_df.height\n",
    "silva_df[\"ncbi_taxonid\"].null_count()\n",
    "\n",
    "# Load SILVA taxonomy mappings\n",
    "print(\n",
    "    f\"Merged taxonomy for {silva_df.filter(pl.col('ncbi_taxonid').is_not_null()).height} SILVA sequences\"\n",
    ")\n",
    "\n",
    "unique_taxids = (\n",
    "    silva_df.filter(pl.col(\"ncbi_taxonid\").is_not_null())\n",
    "    .select(\"ncbi_taxonid\")\n",
    "    .unique()[\"ncbi_taxonid\"]\n",
    "    .to_list()\n",
    ")\n",
    "print(\n",
    "    f\"Total of {len(unique_taxids)} unique NCBI taxids found in SILVA sequences\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a1eeb0a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (10, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>ncbi_taxonid</th><th>count</th></tr><tr><td>i64</td><td>u32</td></tr></thead><tbody><tr><td>77133</td><td>183357</td></tr><tr><td>155900</td><td>19293</td></tr><tr><td>115547</td><td>12243</td></tr><tr><td>256318</td><td>7959</td></tr><tr><td>562</td><td>7811</td></tr><tr><td>100272</td><td>6908</td></tr><tr><td>136703</td><td>5459</td></tr><tr><td>573</td><td>4190</td></tr><tr><td>1280</td><td>3213</td></tr><tr><td>171953</td><td>2670</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (10, 2)\n",
       "┌──────────────┬────────┐\n",
       "│ ncbi_taxonid ┆ count  │\n",
       "│ ---          ┆ ---    │\n",
       "│ i64          ┆ u32    │\n",
       "╞══════════════╪════════╡\n",
       "│ 77133        ┆ 183357 │\n",
       "│ 155900       ┆ 19293  │\n",
       "│ 115547       ┆ 12243  │\n",
       "│ 256318       ┆ 7959   │\n",
       "│ 562          ┆ 7811   │\n",
       "│ 100272       ┆ 6908   │\n",
       "│ 136703       ┆ 5459   │\n",
       "│ 573          ┆ 4190   │\n",
       "│ 1280         ┆ 3213   │\n",
       "│ 171953       ┆ 2670   │\n",
       "└──────────────┴────────┘"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "silva_df[\"ncbi_taxonid\"].value_counts(sort=True).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f490be8c",
   "metadata": {},
   "source": [
    "Removing 77133,155900,115547,256318,100272,136703,171953,86473 (uncultured) as they are not useful for genome fetching later on.\n",
    "Also removing things wiht \"uncultured\" in the submitted_path."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5c4f64d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "530930\n",
      "226832\n",
      "header\n",
      "['FN773278.1.1468 Bacteria;Campylobacterota;Campylobacteria;Campylobacterales;Sulfurospirillaceae;Sulfurospirillum;bacterium endosymbiont of Osedax mucofloris', 'AF163120.1.1458 Bacteria;Actinomycetota;Actinobacteria;Streptosporangiales;Thermomonosporaceae;Actinoallomurus;Actinoallomurus spadix', 'CP022663.4249352.4250905 Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Salmonella;Salmonella enterica', 'EU136629.1.1439 Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Pseudoalteromonadaceae;Pseudoalteromonas;Pseudoalteromonas sp. Z6', 'KC984863.1.1291 Bacteria;Pseudomonadota;Gammaproteobacteria;Pseudomonadales;Halomonadaceae;Halomonas;Halomonas sp. UADY-002']\n",
      "sequence\n",
      "['AACCUGGUUGAUCCUGCCAGUAGUCACUCGCUCGUCUCAAAGACUAAGCCAUGCAUGUCUAAGUAUAAAUCUUUUACUUUGAAACUGCGAACGGCUCAUUAUAUCAGUUAUAGUUUAUUUGAUAGUCCCUUACUACUUGGAUAACCGUAGUAAUUCUAGAGCUAAUACAUGCGUCAAUACCCUUCUGGGGUAGUAUUUAUUAGAUUGAAACCAACCCCUUCGGGGUGAUGUGGUGAUUCAUAAUAAGCUUGCGGAUCGCAUGGCUUUGCCGGCGAUGGAUCAUCCAAGUUUCUGCCCUAUCAGCUUUGGAUGGUAGGGUAUUGGCCUACCAUGGCUUUAACGGGUAACGGGAAAUNAGGGUUUGAUUCCGGAGAGGGAGCCUGAGAGACGGCUACCACAUCCAAGGAAGGCAGCAGGCGCGUAAAUUACCCAAUCCUGACACAGGGAGGUAGUGACAAUAAAUAACAAUGCCGGGCCUUUGUAGGUCUGGCAAUUGGAAUGAGAACAAUUUAAACCCCUUAUCGAGUAUCAAUUGGAGGGCAAGUCUGGUGCCAGCAGCCGCGGUAAUUCCAGCUCCAAUAGCGUAUAUUAAAGUUGUUGCAGUUAAAAAGCUCGUAGUUGGAUUUGUGGCAGUCGCAUGCGGCCCGGCAUUCGUGUCGGUGUUUGGUCGCGUCGCCAUCCUUGGGUGGAACUGGUGUGGCAUUAGGUUGUCGUGCAGGGGAUGCCCAUCAAUUACUGUGAAAAAAUUAGAGCGUUCAAAGCAGGCUUAUGCCGUUGAAUAUAUUAGCAUGGAAUAAUAAGAUAGGACUUUGGUACUAUUUUGUUGGUUUGCGCACCAAGGUAAUGAUUAAUAGGGACAGUUGGGGGUAUUCGUAUUCCAUUGUCAGAGGUGAAAUUCUUGGAUUUUUGGAAGACGAACUACUGCGAAAGCAUUUACCAAGGAUGUUUUCAUUAAUCAAGAACGAAAGUUAGGGGAUCGAAGAUGAUUAGAUACCAUCGUAGUCUUAACCAUAAACUAUGCCGACAAGGGAUUGGCAGAGUCUCGUUUGUCUCCGUCAGCACCUUAUGAGAAAUCACAAGUCUUUGGGUUCCGGGGGGAGUAUGGUCGCAAGGCUGAAACUUAAAGAAAUUGACGGAAGGGCACCACCAGGAGUGGAGCCUGCGGCUUAAUUUGACUCAACACGGGAAAACUUACCAGGUCCAGACAUAGUGAGGAUUCACAGAUUGAGAGCUCUUUCUUGAUCUUAUGGGUGGUGGUGCAUGGCCGUUCUUAGUUGGUGGAGUGAUUUGUCUGGUUAAUUCCGUUAACGAACGAGACCCCUGCUGCUGAAAUAGCUUGCCGAGUGAAUUUUCACUGGGUGAAGCUUCUUAGAGGGACGUGCAUUCUAUUAGAUGCAGGAAGAUAGGGGCAAUAACAGGUCUGUGAUGCCCUUAGAUGUCCUGGGCCGCACGCGCGCUACACUGAUGCAUGCAACGAGAGUACCUUGGCCGAGAGGCCUGGGAAAUCUGUGAACGUGCAUCGUGAUAGGGAUAGAUUAUUGCAAUUAUUAAUCUUGAACGAGGAAUUCCUAGUAAACGCAGAUCAUCAGUCUGCAUUGAUUAGGUCCCUGCCCUUUGUACACACCGCCCGUCGCACCUACCGAUUGAAUGGUCCGGUGAAGCCUCGGGAUUGUGAUUUGUCUCCUUUAUUGGGGAUUUGUCGCGAGAACUUGUCUAAACCUUAUCAUUUAG', 'UCAAGCUACAAAGGGCGCAUGGUGAAUGCCUUGGCAUCAGGAGCCGAUGAAGGACGUGAUAAGCUGCGAUAAGCUUCGGGUAGGCGCACAUAGCCUGAGAUCCGAAGAUCUCCGAAUGAGGAAACUCACAUGGGAAACCCCAUGUAUCAUUAAGUGAAUACAUAGCUUAAUGAAGGUAACCCAGGGAACUGAAACAUCUAAGUACCUGGAGGAAGAGAAAGAAAAAUCGAUUUUCUUAGUAGCGGCGAGCGAAAGGGAAAGAGCCCAAACCGGAAACUUGUUUCCGGGGUUGUGGACAUAACAUAAAAGUGGAGGCUAUUGUAACUGAAGAGAACUGGAAAGUUCUACCGUAGAAGGUAAUAGUCCUGUAAGUGAAACGAGAAGACCACGAGUUAUGAUCCAGAGUACCACGAGACACGUGAAACCUUGUGGGAAGCAGGGAGGACCACCUCCCAAGGCUAAAUACUACCUGAUGACCGAUAGUGAAGCAGUACCGUGAGGGAAAGGUGAAAAGAACCCCGGGAGGGGAGUGAAAUAGAACCUGAAACCGUGUGCCUACAACCGCUCAGAGCCCCUUAUGAGGGUGAUGAGGUGCUUUUUGUAGAACGAGCCAACGAGUUACGGUAUGUAGCAAGGUUAAGUACUUAAGGUACGGAGCCGAAGGGAAACCGAGUCUGAAUAGGGCGACUAGUUGCAUGCCGUAGACCCGAAACCGGGUGACCUAUCCAUGGCCAGGUUGAAGCGAGGGUAAAACCUCGUGGAGGACCGAACCACGUUGCUGUUGAAAAAGCAUGGGAUGAGCUGUGGAUAGCGGAGAAAUUCCAAUCGAACCCGGAGAUAGCUGGUUCUCUCCGAAAUAGCUUUAGGGCUAGCGUCGAGUAUGAGUAGUGGAGGUAGAGCACUGAAUGGGCUAGGGGGCAUAGCGCUUACCGAACCUUAUCAAACUCCGAAUGCCACAUACUUUGAGCUCGGCAGUCAGACUACGAAUGAUAAGAUCCGUGGUCAAAAGGGAAACAGCCCAGAUCGUCAGCUAAGGUCCCAAAGUGUAAGUUAAGUGGGAAAGGAUGUGAGAUUUCUAAGACAACUAGGAUGUUGGCUUAGAAGCAGCCAUUCAUUAAAAGAGUGCGUAAUAGCUCACUAGUCGAGAGAUCUUGCGCCGAAAAUGUAACGGGGCUAAACUUACCACCGAAGCUACGGGCUUGAACUUAGUUCAAGCGGUAGGAGAGCGUCGUAACCGGGCUGAAGUCGUACCGUAAGGAGCGGUGGACUGGUUACGAGUGAGAAUGUUGGCAUCAGUAGCGAGAUGUGGGUGAGAAUCCCACAGGCCGAAAACCUAAGGUUUCCUCAGGAAGGUUCGUCCGCUGAGGGUUAGUCGGGACCUAAGCCGAGGCCGAAAGGCGUAGGUGAUGGACAACUGGUUGAUAUUCCAGUACCACUACCAUCGUUAUUAUCGAUGGCGUGACGCAGGAGGAUAGUGUGUGCUAGCUAUUGGAUGCUAGUCUAAGCGUUUAGGCAGUGAGAGUAGGCAAAUCCGCUCUCAUAAUGCUGAGGCGUGAUGGGGAAGGUUCUACGGAACCGAAGUACAUGAUUCCACGCUGCCAAGAAAAGCGUCUAGAGAGAGAAGUAGUGCCCGUACCGCAAACCGACACAGGUAGGUGAGGAGAGAAUCCUAAGGCCGGCGGAAGAAUUACAGUUAAGGAACUCGGCAAAUUGACCCCGUAAGUUCGCGAGAAGGGGUGCCUACGAGAGUAGGCCGCAGAGAAUAGGCCCAAGCAACUGUUUAGCAAAAACACAGGUCUCUGCUAAAGCGUAAGCUGAUGUAUAGGGGCUGACGCCUGCCCGGUGCUGGAAGGUUAAGGGGAACACUUAGCGCAAGCGAAGGUGUGAACUUAAGCCCCAGUAAACGGCGGCCGUAACUAUAACGGUCCUAAGGUAGCGAAAUUCCUUGUCAGGUAAGUUCUGACCCGCACGAAUGGCGUAAUGACUUGGGCACUGUCUCAACUGUAAAUCCGGCGAAGUUGUAGUGCGAGUGAAGAUGCUCGCUACCCGCGAUUGGACGGAAAGACCCCGUAGAGCUUUACUGUAGAUUAGCAUUGAGUUUUGGUAUUGUCUGUACAGGAUAGGUGGGAGACUAGGAAACCAGGGCGUCAGCCUUGGUGGAGUCAACCUUGGGAUACCACCCUGACAGUACUGAAAUUCUAACUGGCGGCCAUGAAUCUGGUCACAGGACAUUGUUAGGUGGGCAGUUUGACUGGGGCGGUCGCCUCCUAAAAAGUAACGGAGGCGCCCAAAGGUUCCCUCAGAACGGUCGGAAAUCGUUCGAAGAGUGCAAAGGCAGAAGGGAGCCUGACUGCGACACCCACAAGUGGAGCAGGGACGAAAGUCGGGCUUAGUGAUCCGGUGGUACCUCGUGGGAGGGCCAUCGCUCAACGGAUAAAAGCUACCUCGGGGAUAACAGGCUGAUCUCCCCCAAGAGUCCACAUCGACGGGGAGGUUUGGCACCUCGAUGUCGGCUCGUCGCAUCCUGGGGCUGAAGUAGGUCCCAAGGGUUGGGCUGUUCGCCCAUUAAAGCGGCACGCGAGCUGGGUUCAGAACGUCGUGAGACAGUUCGGUCCCUAUCCGUCGCGGGCGUAGGAAAUUUGAGAGGAGCUGUCCUUAGUACGAGAGGACCGGGAUGGACUGACCUCUGGUGUACCAGUUGUUCCGCCAGGAGCACAGCUGGGUAGCUAAGUCGGGAAUGGAUAAACGCUGAAAGCAUCUAAGCGUGAAGCCAACCUCAAGAUGAGAUUUCCCAUAGCAUAAGCUAGUAAGAUCCCUUGAAGAACACAAGGUUGAUAGGUCGGGGGUGUAAGCAUGGUAACAUGUUCAGCUGACCGAUACUAAUAGAUCGAGGGCUUGACCAA', 'GGGCGAGCUACCUGCAGUCGAACGGCAGCACAGGAGAGCUUGCUCUCUGGGUGGCGAGUGGCGGACGGGUGAGGAAUACAUCGGAAUCUACUUUUUCGUGGGGGAUAACGUAGGGAAACUUACGCUAAUACCGCAUACGACCUACGGGUGAAAGCAGGGGAUCUUCGGACCUUGCGCGAUUGAAUGAGCCGAUGUCGGAUUAGCUAGUUGGCGGGGUAAAGGCCCACCAAGGCGACGAUCCGUAGCUGGUCUGAGAGGAUGAUCAGCCACACUGGAACUGAGACACGGUCCAGACUCCUACGGGAGGCAGCAGUGGGGAAUAUUGGACAAUGGGCGCAAGCCUGAUCCAGCCAUACCGCGUGGGUGAAGAAGGCCUUCGGGUUGUAAAGCCCUUUUGUUGGGAAAGAAAUCCAGCUGGCUAAUACCCGGUUGGGAUGACGGUACCCAAAGAAUAAGCACCGGCUAACUUCGUGCCAGCAGCCGCGGUAAUACGAAGGGUGCAAGCGUUACUCGGAAUUACUGGGCGUAAAGCGUGCGUAGGUGGUCGUUUAAGUCCGUUGUGAAAGCCCUGGGGCUCAACCUUGGGAACUGCAGUGGAUAACUGGGCGAACUAGGAGUGUGGUAGAGGGUAGCGGAAAUUCCUGGUGUAGCAGUGAAAUGCGUAGAGAUCAGGAGGAACAUCCAUGGCGAAGGCAGCUACCUGGACCAACACUGACACUGAGGCACGAAAGCGUGGGGAGCAAACAGGAUUAGAUACCCUGGUAGUCCACGCCCUAAACGAUGCGAACUGGAUGUUGGGUGCAAUUUGGCACGCAGUAUCGAAGCUAACGCGUUUAAGUUUCGCCGCCUGGGGGGAGUACGGUCGCAAGACUGAAACUCAAAGGAAUUGACGGGGGCCCGCACAAGCGGUGGAGUAUGUGGUUUAAUUCGAUGCAACGCGAAGAACCUUACCUGGCCUUGACAUGUCGAGAACUUUCCAGAGAUGGAUUGGUGCCUUCGGGAACUCGAACACAGGUGCUGCAUGGCUGUCGUCAGCUCGUGUCGUGAGAUGUUGGGUUAAGUCCCGCAACGAGCGCAACCCUUGUCCUUAGUUGCCAGCACAUAAUGGUGGGAACUCUAAGGAGACCGCCGGUGACAAACCGGAGGAAGGUGGGGAUGACGUCAAGUCAUCAUGGCCCUUACGGCCAGGGCUACACACGUACUACAAUGGUAGGGACAGAGGGCUGCAAGCCGGCGACGGUAAGCCAAUCCCAGAAACCCUAUCUCAGUCCGGAUUGGAGUCUGCAACUCGACUCCAUGAAGUCGGAAUCGCUAGUAAUCGCAGAUCAGCAUUGCUGCGGUGAAUACGUUCCCGGGCCUUGUACACACCGCCCGUCACACCAUGGGAGUUUGUUGCACCAGAAGCAGGUAGCUUAACCUUCGGGAGGGCGCUGCCACAUUAAC', 'CAUAUCACUAAGCGGAGAAAAAGAAAAUAACAAUGAUUUCUAAAGUAACGGCGAGUGAAAUAGAAAUAGCCCAGAGCCGAAUUCUGUUGUUCUUAGUGGCAACUAGAAGUUGUGGCGUUGAGAACUGUCUAACAUAACUGCUGUGUCGAGCUACGCGACUCCUCUUGAUUGUGGUUUGCCCAUAGCAGGUGCAAGGCCUAUAGUCGUAGUUCAGUUAUGGCACUCAGUGAUGGCAGUCCUUCGAGUCGGGUUGUUUGGGAUUGCAGCCCGAAGUAGGUGGUAAACUCCACCUAAGGCUAAAUACGAUAGCGAUUCCGAUAGCAAACAAGUACCAUGAGGGACAGUUGAAAAGAACUUUGAAGAGAGAGUUCAAAGUACGUGAAACCGUUCAGAGGUAAACAGGUAGAAAGUCGAAUUCAAAAUUGGCAAUCAAAUUGAAUUGUGGUAGAUUAACUUGCUUUCUAGCACGUCAUAGGCUAGUAUUGCUUGAAAUUUGCUAUAUCAAUUGAAUUUGUCAAUUUUAUAUCGCCGCAUCAGCAGUGUUGUAAUCUGUUUCAACUCGAUUUCCUCUGUUUAGCGUUUUGGCCUUUGGGUCUAUUAUGUUAAAAUAUGAAAUUGGGUUCUGAUUUUAAACCACAGCUGUACACAUUAUGUUUAUUCAUUCCAAUUUGCUUGUAUGUAGGCAAAGCUGACCUGUUGAUCGAUAAUUGUUCAGCAAUUGGCUUUUGUUUGCUGCAAGUUGGUUUCAUAAAGGAACAAAAACAUAGCGGCGUUGAUUUACGCAUUAUCUGCCUGACCCGUCUUGAAACACGGACCAAGGAGUCUAACUUGUGCGCAAGUCAUUGGAAUUUCUAUGAUAAAAUCCAAAGGCACAGUGAAAGCAAUAUUGGUGAGAUCUGACGUUUAUUAGUUGAUAAUAACGCCGGCGCAUCAUCGUCACGCCUCAAUUUGGCUUCGGCUAAAGGGGCGCAGACAGAGCGUACAAGUUGGGACCCGAAAGAUGGUGAACUAUGCGAGGCGAGGUCGAAGCCAGAGGAAACUCUGGUGGAGGACUGUAGCGGUUCUAACGUGCAAAUUGAUCGUCGCAGCUUCGUAUAGGGGCGAAAGACUAAUCGAACCAUCUAGUAGCUGGUUCCCUCCGAAGUUUCCCUCAGGAUAGCUGGCACUUGUUUUAAUAGCAGUUUUGUCCGGUAAAGCGAAUGACUAGAGGCCUUGGGGACAUAACGACCUCAACCUAUUCUCAAACUUUGAAUGGGCAAGAAAAAGGACUCGCUUAAAUGGAGUCACUUUAUACCGUUUAACGUGAAUGCUUGAGUGCCAAGUGGGCCAUUUUUGGUAAGCAGAACUGGCGCUGUGGGAUGAACCAAACGUCGAGCUAAGGCGCCAAAUGCUAACGCUCAUCCGAUCCCAUUAAAGGUGUUGGUUGAUAGAGACAGCAGGACGGUGGCCAUGGAAGUCGGAAUCCGCUAAGGAGUGUGUAACAACUCACCUGCCGAAUCAACCAGUCCUGAAAAUGGAUGGCGCUAAAGCGUUGAAAGCCUAUGCUUGACCGGCAUUACAGCUGUUUGUUAUGAUGUGAUGUCGAGUAGGAGGGUCGCUGUGGUAUGCAUUGAAGCUUUGGGCGUGGGCCCGAGUCGAGCUUCCACAGGUGCAGAUCUUGGUGGUAGUAGCAAAUAUUCAAACGAGAACUUUGAAGGCCGAAGUGGAGAAGGGUUCCAUGUGAACAGCAUUUGAACAUGGGUCAGUCGGUCCUAAGCGAUAGGCAAAUGCCGUUCUGAAAGCAGGCAAUGUUUUGCACUAAUGCCUGAUUUUGGUCGAAAGGGAAUCGGGUUAACAUUCCCGAACCUGGACGCGGAGAUCGAGUUCAGUGUUACAACUGAGCUCUGUGCGGUGACGCAAACGAUCUCGGAUUCGUUGGCGAAAGUCCUGGGAAGAGUUCUCUCUUCUUUGUAAGGGGCUGACUCCCUGGAAUCGGCUUAGCCGGAGAUAGGGAUGUUGCAGAUACCCGUAAAGCACCACACUUCUUGUGGUGUCCGGUGCACAAUCGUCGACACUUGAAAAUCCGAGCUAGACUUGAUACAUUUUCGCGCCAGGCCGUACCCAUAUCCGCAGCAGGUCUCCAAGGUGAACAGCCUCUGGUCUUAGACUGAUGUAGUCAAGGGAAGUCGGCAAAUUGGAUCCGUAACUUUGGGAAAAGGAUUGGCUCUAUGGGUUGGGCCGCGACGGGCUGUGUUAGAAAGCGAUGCUUGCUAUCAGUGCUUCGGGUCUGAUUGGCAACGAUUGGGCCUAGCGUUGAUUGGUGGCGCGUGGACCUGGCAUGGCUUCACCGGCUUCUCUUCGCGAGUUGUCGGCAUCGUUCGGCGAAAAGCAACCACCAUAGAACUGGUACGGACCAGGGGAA', 'AACGAACGCUGGCGGCAGGCUUAACACAUGCAAGUCGAGCGCCCCGCAAGGGGAGCGGCAGACGGGUGAGUAACGCGUGGGAACGUACCCUUUACUACGGAAUAACGCAGGGAAACUUGUGCUAAUACCGUAUGUGCCCUUCGGGGGAAAGAUUUAUCGGUAAGGGAUCGGCCCGCGUUGGAUUAGCUAGUUGGUGGGGUAAAGGCCUACCAAGGCGACGAUCCAUAGCUGGUCUGAGAGGAUGAUCAGCCACAUUGGGACUGAGACACGGCCCAAACUCCUACGGGAGGCAGCAGUGGGGAAUAUUGGACAAUGGGCGCAAGCCUGAUCCAGCCAUGCCGCGUGAGUGAUGAAGGCCCUAGGGUUGUAAAGCUCUUUCACCGGAGAAGAUAAUGACGGUAUCCGGAGAAGAAGCCCCGGCUAACUUCGUGCCAGCAGCCGCGGUAAUACGAAGGGGGCUAGCGUUGUUCGGAAUUACUGGGCGUAAAGCGCACGUAGGCGGACAUUUAAGUCAGGGGUGAAAUCCCAGAGCUCAACUCUGGAACUGCCUUUGAUACUGGGUGUCUGGAGUAUGGAAGAGGUGAGUGGAAUUCCGAGUGUAGAGGUGAAAUUCGUAGAUAUUCGGAGGAACACCAGUGGCGAAGGCGGCUCACUGGUCCAUUACUGACGCUGAGGUGCGAAAGCGUGGGGAGCAAACAGGAUUAGAUACCCUGGUAGAUACCCUGGUAGUCCACGCCGUAAACGAUGAAUGUUAGCCGUCGGCAAGUUUACUUGUCGGUGGCGCAGCUAACGCAUUAAACAUUCCGCCUGGGGAGUACGGUCGCAAGAUUAAAACUCAAAGGAAUUGACGGGGGCCCGCACAAGCGGUGGAGCAUGUGGUUUAAUUCGAAGCAACGCGCAGAACCUUACCAGCCCUUGACAUGCCCGGCCAUCUGCAGAGAUGCAGAGUUCCCUUCGGGGACCGGGACACAGGUGCUGCAUGGCUGUCGUCAGCUCGUGUCGUGAGAUGUUGGGGUAAGUCCCGCAACGAGCGCAACCCUCGCCCUUAGUUGCCAGCAUUCAGUUGGGCACUCUUAAGGGGACUGCCGGUGAUAAGCCGAGAGGAAGGUGGGGAUGACGUCAAGUCCUCAUGGCCCUUACGGGCUGGGCUACACACGUGCUACAAUGGUGGUGACAGUGGGCAGCGAGCACGCGAGUGUGAGCUAAUCUCCAAAAGCCAUCUCAGUUCGGAUUGCACUCUGCAACUCGAGUGCAUGAAGUUGGAAUCGCUAGUAAUCGCGGAUCAGCAUGCCGCGGUGAAUACGUUCCCGGGCCUUGUACACACCGCCCGUCACACCAUGGGAGUUGGUUUUACCCGAAGGUAGUGCGCUAACCGCAAGGAGGCAGCUAACCACGGUAGGGUCAGCGACUGGGGUGAAGUCGUAACAAGGUA']\n",
      "rRNA_type\n",
      "['SSU', 'LSU', 'SSU', 'SSU', 'LSU']\n",
      "primaryAccession\n",
      "['AY281078', 'AB915216', 'Y17190', 'EF423414', 'DQ631734']\n",
      "accession\n",
      "['CP018007.2993062', 'AJ295716.1', 'MKYQ01000782.2648941', 'CP023149.1613845', 'MG203880.1']\n",
      "taxonomy_raw\n",
      "['Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Escherichia-Shigella;Escherichia coli', 'Bacteria;Pseudomonadota;Gammaproteobacteria;Pseudomonadales;Pseudomonadaceae;Pseudomonas;Pseudomonas fluorescens A506', 'Eukaryota;Archaeplastida;Chloroplastida;Charophyta;Phragmoplastophyta;Gossypium hirsutum (upland cotton)', 'Bacteria;Bacillota;Bacilli;Bacillales;Bacillaceae;Priestia;Bacillus flexus', 'Bacteria;Pseudomonadota;Gammaproteobacteria;Enterobacterales;Enterobacteriaceae;Klebsiella;Klebsiella oxytoca']\n",
      "ncbi_taxonid\n",
      "[393060, 2077137, 231022, 435910, 216107]\n",
      "submitted_path\n",
      "['Bacteria;Firmicutes;Clostridia;Clostridiales;Clostridiaceae;Clostridium;', 'Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Yersiniaceae;Yersinia;', 'Bacteria;Proteobacteria;Gammaproteobacteria;Enterobacterales;Yersiniaceae;Serratia;', 'Eukaryota;Viridiplantae;Streptophyta;Embryophyta;Tracheophyta;Spermatophyta;Magnoliophyta;eudicotyledons;Gunneridae;Pentapetalae;rosids;malvids;Malvales;Malvaceae;Malvoideae;Gossypium;', 'Bacteria;Fusobacteria;Fusobacteriales;Fusobacteriaceae;Fusobacterium;']\n",
      "new_header\n",
      "['672@silva@LSU@2d164a3f1bb9ebd83210e6ec54eb6bbf', '294@silva@SSU@d73bdd1b0e1d43b44b1670f7a3eb74b7', '37333@silva@SSU@faf880eb4f8cd066471abbba70f09b3a', '1396@silva@SSU@8089e3257f22c07c272e0f4aa8cc01e8', '376815@silva@SSU@2dab3f95e66688bc564b2a0c92edcf0e']\n",
      "seq_hash\n",
      "['ce611f478fc73f8bbdae928d9f332c37', '51b7355bac448663b370a6ecb53da516', '3b14a2d467be501ede25a223cd80bd00', '07f19fb3fb68d9f2740ea2f185cf7dfc', '8e12e8db52a1696ebdab61152c4c7b7d']\n",
      "Finished writing SILVA rRNA sequences parquet file.\n",
      "total SILVA sequences after filtering: 226832, removed 304098 sequences\n"
     ]
    }
   ],
   "source": [
    "silva_df = silva_df.with_columns(\n",
    "    t1=pl.concat_str(\n",
    "        [pl.col(\"ncbi_taxonid\"), pl.lit(\"silva\"), pl.col(\"rRNA_type\")],\n",
    "        separator=\"@\",\n",
    "    )\n",
    ").with_columns(\n",
    "    new_header=pl.concat_str([pl.col(\"t1\"), pl.col(\"header\")], separator=\" \")\n",
    ")\n",
    "silva_df = silva_df.with_columns(\n",
    "    t1=pl.concat_str(\n",
    "        [pl.col(\"ncbi_taxonid\"), pl.lit(\"silva\"), pl.col(\"rRNA_type\")],\n",
    "        separator=\"@\",\n",
    "    )\n",
    ")\n",
    "\n",
    "silva_df = silva_df.with_columns(\n",
    "    pl.col(\"sequence\")\n",
    "    .map_elements(lambda x: hashlib.md5(x.encode()).hexdigest())\n",
    "    .alias(\"seq_hash\")\n",
    ").with_columns(\n",
    "    new_header=pl.concat_str([pl.col(\"t1\"), pl.col(\"seq_hash\")], separator=\"@\")\n",
    ")\n",
    "silva_df = silva_df.unique()\n",
    "silva_df = silva_df.unique(subset=[\"sequence\"])\n",
    "silva_df = silva_df.drop(\"t1\")\n",
    "\n",
    "silva_df[\"new_header\"].head(1).item()\n",
    "\n",
    "heigt_befrore = silva_df.height\n",
    "print(silva_df.height)\n",
    "silva_df = silva_df.filter(\n",
    "    ~pl.col(\"ncbi_taxonid\").is_in(\n",
    "        [77133, 155900, 115547, 256318, 100272, 136703, 171953, 86473, 56765]\n",
    "    )\n",
    ")\n",
    "silva_df = silva_df.filter(\n",
    "    ~pl.col(\"submitted_path\").str.contains_any(\n",
    "        [\"uncultured\", \"unclassified\", \"synthetic\", \"environmental sample\"]\n",
    "    )\n",
    ")\n",
    "\n",
    "print(silva_df.height)\n",
    "for coln in silva_df.columns:\n",
    "    print(coln)\n",
    "    print(silva_df[coln].sample(5).to_list())\n",
    "\n",
    "silva_df.write_parquet(os.path.join(rrna_dir, \"silva_rrna_sequences.parquet\"))\n",
    "print(\"Finished writing SILVA rRNA sequences parquet file.\")\n",
    "print(\n",
    "    f\"total SILVA sequences after filtering: {silva_df.height}, removed {heigt_befrore - silva_df.height} sequences\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42f22878",
   "metadata": {},
   "source": [
    "Deduplicate, mask kmers shared with RNA viruses, and apply entrophy masking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fd1a7c1c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:55:14</span> - Processed <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">226832</span> records: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">226832</span> unique, <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0</span> duplicates removed                                          <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/bio/sequences.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">sequences.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/bio/sequences.py#725\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">725</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:55:14\u001b[0m - Processed \u001b[1;36m226832\u001b[0m records: \u001b[1;36m226832\u001b[0m unique, \u001b[1;36m0\u001b[0m duplicates removed                                          \u001b]8;id=852034;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/bio/sequences.py\u001b\\\u001b[2msequences.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=718033;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/bio/sequences.py#725\u001b\\\u001b[2m725\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import rich_click as click\n",
    "\n",
    "# replace headers with md5 hash as the new_header breaks mappers. Will remap after this\n",
    "silva_df = silva_df.with_columns(\n",
    "    pl.col(\"sequence\")\n",
    "    .map_elements(lambda x: hashlib.md5(x.encode()).hexdigest())\n",
    "    .alias(\"tmp_header\")\n",
    ")\n",
    "\n",
    "write_fasta_file(\n",
    "    headers=silva_df[\"tmp_header\"].to_list(),\n",
    "    seqs=silva_df[\"sequence\"].to_list(),\n",
    "    output_file=os.path.join(rrna_dir, \"temp_silva_rrna_sequences.fasta\"),\n",
    ")\n",
    "# deduplicate sequences\n",
    "deduplicated_silva_fasta_path = os.path.join(\n",
    "    rrna_dir, \"silva_rrna_sequences_deduplicated.fasta\"\n",
    ")\n",
    "remove_duplicates(\n",
    "    input_file=os.path.join(rrna_dir, \"temp_silva_rrna_sequences.fasta\"),\n",
    "    output_file=deduplicated_silva_fasta_path,\n",
    "    logger=logger,\n",
    "    by=\"seq\",\n",
    "    # revcomp_as_distinct=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "2d04eb68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:55:34</span> - datadir used: <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">data</span>                             <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#71\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">71</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:55:34\u001b[0m - datadir used: \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/\u001b[0m\u001b[95mdata\u001b[0m                             \u001b]8;id=325631;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=88559;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#71\u001b\\\u001b[2m71\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: mmseqs easy-linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/masking/combined_entropy_masked.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna  --min-seq-id 0.7 --min-aln-len 80 --threads 14 --format-mode 1 --search-type 3 -v 3 --max-accept 1231 -a  \n",
      "easy-linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/masking/combined_entropy_masked.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna --min-seq-id 0.7 --min-aln-len 80 --threads 14 --format-mode 1 --search-type 3 -v 3 --max-accept 1231 -a \n",
      "\n",
      "MMseqs Version:                    \t18.8cc5c\n",
      "Substitution matrix                \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace                      \ttrue\n",
      "Alignment mode                     \t3\n",
      "Alignment mode                     \t0\n",
      "Allow wrapped scoring              \tfalse\n",
      "E-value threshold                  \t0.001\n",
      "Seq. id. threshold                 \t0.7\n",
      "Min alignment length               \t80\n",
      "Seq. id. mode                      \t0\n",
      "Alternative alignments             \t0\n",
      "Coverage threshold                 \t0\n",
      "Coverage mode                      \t0\n",
      "Max sequence length                \t65535\n",
      "Compositional bias                 \t1\n",
      "Compositional bias scale           \t1\n",
      "Max reject                         \t2147483647\n",
      "Max accept                         \t1231\n",
      "Include identical seq. id.         \tfalse\n",
      "Preload mode                       \t0\n",
      "Pseudo count a                     \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                     \tsubstitution:4.100,context:5.800\n",
      "Score bias                         \t0\n",
      "Realign hits                       \tfalse\n",
      "Realign score bias                 \t-0.2\n",
      "Realign max seqs                   \t2147483647\n",
      "Correlation score weight           \t0\n",
      "Gap open cost                      \taa:11,nucl:5\n",
      "Gap extension cost                 \taa:1,nucl:2\n",
      "Zdrop                              \t40\n",
      "Threads                            \t14\n",
      "Compressed                         \t0\n",
      "Verbosity                          \t3\n",
      "Seed substitution matrix           \taa:VTML80.out,nucl:nucleotide.out\n",
      "k-mers per sequence                \t21\n",
      "Scale k-mers per sequence          \taa:0.000,nucl:0.200\n",
      "Mask residues                      \t1\n",
      "Mask residues probability          \t0.9\n",
      "Mask lower case residues           \t0\n",
      "Mask lower letter repeating N times\t0\n",
      "Add N similar to search            \t1\n",
      "Result direction                   \t1\n",
      "Split memory limit                 \t0\n",
      "Min codons in orf                  \t30\n",
      "Max codons in length               \t32734\n",
      "Max orf gaps                       \t2147483647\n",
      "Contig start mode                  \t2\n",
      "Contig end mode                    \t2\n",
      "Orf start mode                     \t1\n",
      "Forward frames                     \t1,2,3\n",
      "Reverse frames                     \t1,2,3\n",
      "Translation table                  \t1\n",
      "Translate orf                      \t0\n",
      "Use all table starts               \tfalse\n",
      "Offset of numeric ids              \t0\n",
      "Create lookup                      \t0\n",
      "Chain overlapping alignments       \t0\n",
      "Merge query                        \t1\n",
      "Search type                        \t3\n",
      "MPI runner                         \t\n",
      "Force restart with latest tmp      \tfalse\n",
      "Remove temporary files             \ttrue\n",
      "Alignment format                   \t1\n",
      "Format alignment output            \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits\n",
      "Database output                    \tfalse\n",
      "\n",
      "Alignment backtraces will be computed, since they were requested by output format.\n",
      "createdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/masking/combined_entropy_masked.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query --shuffle 0 --threads 14 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[====================================================================\n",
      "Time for merging to query_h: 0h 0m 0s 6ms\n",
      "Time for merging to query: 0h 0m 0s 7ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 2s 24ms\n",
      "createdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target --shuffle 0 --write-lookup 0 --threads 14 --gpu 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[======================\n",
      "Time for merging to target_h: 0h 0m 0s 5ms\n",
      "Time for merging to target: 0h 0m 0s 6ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 1s 736ms\n",
      "Create directory /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp\n",
      "createlinindex /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 14 --min-length 30 --max-length 32734 --max-gaps 2147483647 --contig-start-mode 2 --contig-end-mode 2 --orf-start-mode 1 --forward-frames 1,2,3 --reverse-frames 1,2,3 --translation-table 1 --translate 0 --use-all-table-starts 0 --id-offset 0 --create-lookup 0 --compressed 0 --remove-tmp-files 1 \n",
      "\n",
      "createlinindex /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 14 --min-length 30 --max-length 32734 --max-gaps 2147483647 --contig-start-mode 2 --contig-end-mode 2 --orf-start-mode 1 --forward-frames 1,2,3 --reverse-frames 1,2,3 --translation-table 1 --translate 0 --use-all-table-starts 0 --id-offset 0 --create-lookup 0 --compressed 0 --remove-tmp-files 1 \n",
      "\n",
      "splitsequence /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp/16375924321180544713/nucl_split_seq --max-seq-len 65535 --sequence-overlap 0 --sequence-split-mode 1 --headers-split-mode 0 --create-lookup 0 --threads 14 --compressed 0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 41ms\n",
      "kmerindexdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp/16375924321180544713/nucl_split_seq /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 14 \n",
      "\n",
      "kmerindexdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp/16375924321180544713/nucl_split_seq /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 14 \n",
      "\n",
      "Database size: 226832 type: Nucleotide\n",
      "\n",
      "Process file into 1 parts\n",
      "Generate k-mers list 0\n",
      "[=================================================================] 226.83K 3s 289ms\n",
      "\n",
      "Adjusted k-mer length 17\n",
      "\n",
      "Time for fill: 0h 0m 3s 319ms\n",
      "Sort kmer ... Time for sort: 0h 0m 1s 701ms\n",
      "Write VERSION (0)\n",
      "Write META (1)\n",
      "Write ENTRIES (9)\n",
      "Write ENTRIESOFFSETS (10)\n",
      "Write ENTRIESGRIDSIZE (11)\n",
      "Write ENTRIESNUM (12)\n",
      "Write SEQCOUNT (13)\n",
      "Write SCOREMATRIXNAME (2)\n",
      "Write SPACEDPATTERN (23)\n",
      "Write DBR1INDEX (5)\n",
      "Write DBR1DATA (6)\n",
      "Write DBR2INDEX (7)\n",
      "Write DBR2DATA (8)\n",
      "Write HDR1INDEX (18)\n",
      "Write HDR1DATA (19)\n",
      "Write HDR2INDEX (20)\n",
      "Write HDR2DATA (21)\n",
      "Write GENERATOR (22)\n",
      "Time for fill: 0h 0m 1s 121ms\n",
      "Time for merging to target.linidx: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 6s 708ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/index_tmp/16375924321180544713/nucl_split_seq -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "Create directory /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp\n",
      "linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/result /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp -a 1 --alignment-mode 3 --min-seq-id 0.7 --min-aln-len 80 --max-accept 1231 --threads 14 -v 3 --search-type 3 --remove-tmp-files 1 \n",
      "\n",
      "linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/result /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp -a 1 --alignment-mode 3 --min-seq-id 0.7 --min-aln-len 80 --max-accept 1231 --threads 14 -v 3 --search-type 3 --remove-tmp-files 1 \n",
      "\n",
      "kmersearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref --seed-sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 0 -c 0 --max-seq-len 65535 --pick-n-sim-kmer 1 --result-direction 1 --split-memory-limit 0 --threads 14 --compressed 0 -v 3 \n",
      "\n",
      "kmersearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref --seed-sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 0 -c 0 --max-seq-len 65535 --pick-n-sim-kmer 1 --result-direction 1 --split-memory-limit 0 --threads 14 --compressed 0 -v 3 \n",
      "\n",
      "Process file into 1 parts\n",
      "[=================================================================] 680.81K 3s 146ms\n",
      "\n",
      "Adjusted k-mer length 17\n",
      "\n",
      "Time for fill: 0h 0m 3s 188ms\n",
      "Sort kmer ... Time for sort: 0h 0m 3s 80ms\n",
      "Time to find k-mers: 0h 0m 0s 426ms\n",
      "Time to sort: 0h 0m 0s 14ms\n",
      "Time for merging to pref: 0h 0m 0s 23ms\n",
      "Time for processing: 0h 0m 7s 303ms\n",
      "rescorediagonal /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/reverse_ungapaln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 2 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.9 -a 1 --cov-mode 1 --min-seq-id 0.7 --min-aln-len 80 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 14 --compressed 0 -v 3 \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "[=================================================================] 126.39K 2s 78ms\n",
      "Time for merging to reverse_ungapaln: 0h 0m 0s 100ms\n",
      "Time for processing: 0h 0m 2s 440ms\n",
      "filterdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref_filter --filter-file /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/reverse_ungapaln --positive-filter 0 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 126.39K 0s 11ms\n",
      "Time for merging to pref_filter: 0h 0m 0s 171ms\n",
      "Time for processing: 0h 0m 0s 427ms\n",
      "align /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref_filter /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/reverse_aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 1 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 100000 --min-seq-id 0.7 --min-aln-len 80 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 1231 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 14 --compressed 0 -v 3 \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 226832 type: Nucleotide\n",
      "Target database size: 680809 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 126.39K 2s 997ms\n",
      "Time for merging to reverse_aln: 0h 0m 0s 185ms\n",
      "1009369 alignments calculated\n",
      "703 sequence pairs passed the thresholds (0.000696 of overall calculated)\n",
      "0.005562 hits per query sequence\n",
      "Time for processing: 0h 0m 3s 440ms\n",
      "swapresults /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/reverse_aln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -e 0.001 --split-memory-limit 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --threads 14 --compressed 0 --db-load-mode 0 -v 3 \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Computing offsets.\n",
      "[=================================================================] 126.39K 0s 5ms\n",
      "\n",
      "Reading results.\n",
      "[=================================================================] 126.39K 0s 5ms\n",
      "\n",
      "Output database: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/aln\n",
      "[=================================================================] 680.81K 0s 143ms\n",
      "\n",
      "Time for merging to aln: 0h 0m 0s 249ms\n",
      "Time for processing: 0h 0m 0s 529ms\n",
      "swapresults /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/reverse_ungapaln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/ungap_aln \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Computing offsets.\n",
      "[=================================================================] 126.39K 0s 2ms\n",
      "\n",
      "Reading results.\n",
      "[=================================================================] 126.39K 0s 3ms\n",
      "\n",
      "Output database: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/ungap_aln\n",
      "[=================================================================] 680.81K 0s 154ms\n",
      "\n",
      "Time for merging to ungap_aln: 0h 0m 0s 318ms\n",
      "Time for processing: 0h 0m 0s 649ms\n",
      "concatdbs /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/ungap_aln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/aln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/aln_merged --preserve-keys --take-larger-entry \n",
      "\n",
      "[=================================================================] 680.81K 0s 101ms\n",
      "[=================================================================] 680.81K 0s 155ms\n",
      "Time for merging to aln_merged: 0h 0m 0s 105ms\n",
      "Time for processing: 0h 0m 0s 466ms\n",
      "offsetalignment /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/aln_merged /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/result \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Computing ORF lookup\n",
      "Computing contig offsets\n",
      "Computing contig lookup\n",
      "Time for contig lookup: 0h 0m 0s 10ms\n",
      "Writing results to: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/result\n",
      "[=================================================================] 680.81K 0s 72ms\n",
      "\n",
      "Time for merging to result: 0h 0m 0s 371ms\n",
      "Time for processing: 0h 0m 0s 849ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/search_tmp/8721202811700931978/reverse_aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "convertalis /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/result /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --format-mode 1 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits --translation-table 1 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --db-output 0 --db-load-mode 0 --search-type 3 --threads 14 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 680.81K 0s 80ms\n",
      "Time for merging to tmp_mapped.sam: 0h 0m 0s 270ms\n",
      "Time for processing: 0h 0m 0s 716ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 55ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 37ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 70ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/5693842807573795636/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 10ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">is </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">236303</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam'\u001b[0m\u001b[32m size \u001b[0m\n",
       "\u001b[32mis \u001b[0m\u001b[1;32m236303\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:56:03</span> - Finished running aligner mmseqs2                                                                        <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#220\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">220</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:56:03\u001b[0m - Finished running aligner mmseqs2                                                                        \u001b]8;id=816392;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=867661;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#220\u001b\\\u001b[2m220\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:56:03</span> - beginning bbmask <span style=\"font-weight: bold\">(</span>masking + entropy<span style=\"font-weight: bold\">)</span> step                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:56:03\u001b[0m - beginning bbmask \u001b[1m(\u001b[0mmasking + entropy\u001b[1m)\u001b[0m step                                                               \u001b]8;id=611612;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=125819;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:56:03</span> - equivalent to bbmask.sh                                                                                 <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#226\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">226</span></a>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">in</span>=<span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">silva_rrna_sequences_deduplicated.fasta</span>         <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">out</span>=<span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_masked.fasta</span>                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">sam</span>=<span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_mapped.sam</span> <span style=\"color: #808000; text-decoration-color: #808000\">overwrite</span>=<span style=\"color: #800080; text-decoration-color: #800080\">true</span>     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">threads</span>=<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">14</span> <span style=\"color: #808000; text-decoration-color: #808000\">Xmx</span>=<span style=\"color: #800080; text-decoration-color: #800080\">18g</span>                                                                                                            <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:56:03\u001b[0m - equivalent to bbmask.sh                                                                                 \u001b]8;id=327577;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=433601;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#226\u001b\\\u001b[2m226\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[33min\u001b[0m=\u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95msilva_rrna_sequences_deduplicated.fasta\u001b[0m         \u001b[2m               \u001b[0m\n",
       "         \u001b[33mout\u001b[0m=\u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[95mtmp_masked.fasta\u001b[0m                  \u001b[2m               \u001b[0m\n",
       "         \u001b[33msam\u001b[0m=\u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[95mtmp_mapped.sam\u001b[0m \u001b[33moverwrite\u001b[0m=\u001b[35mtrue\u001b[0m     \u001b[2m               \u001b[0m\n",
       "         \u001b[33mthreads\u001b[0m=\u001b[1;36m14\u001b[0m \u001b[33mXmx\u001b[0m=\u001b[35m18g\u001b[0m                                                                                                            \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">java -ea  --add-modules jdk.incubator.vector -Xmx18g -Xms18g -cp </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">bbmap/current/ jgi.BBMask </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated.fas</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">ta </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_masked.fasta </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sam</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">overwrite</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">true</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">threads</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">14</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> -Xmx18g</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mjava -ea  --add-modules jdk.incubator.vector -Xmx18g -Xms18g -cp \u001b[0m\n",
       "\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/\u001b[0m\n",
       "\u001b[1;31mbbmap/current/\u001b[0m\u001b[1;31m jgi.BBMask \u001b[0m\n",
       "\u001b[1;31min\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31msilva_rrna_sequences_deduplicated.fas\u001b[0m\n",
       "\u001b[1;31mta\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_masked.fasta\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31msam\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_mapped.sam\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31moverwrite\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31mtrue\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mthreads\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m14\u001b[0m\u001b[1;31m -Xmx18g\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">WARNING: Using incubator modules: jdk.incubator.vector</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWARNING: Using incubator modules: jdk.incubator.vector\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Executing jgi.BBMask </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">[</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated.fa</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sta, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_masked.fasta, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sam</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">overwrite</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">true</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">threads</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">14</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, -Xmx18g]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mExecuting jgi.BBMask \u001b[0m\n",
       "\u001b[1;31m[\u001b[0m\u001b[1;31min\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31msilva_rrna_sequences_deduplicated.fa\u001b[0m\n",
       "\u001b[1;31msta\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_masked.fasta\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31msam\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_mapped.sam\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31moverwrite\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31mtrue\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mthreads\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m14\u001b[0m\u001b[1;31m, -Xmx18g\u001b[0m\u001b[1;31m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Set threads to </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">14</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSet threads to \u001b[0m\u001b[1;31m14\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Loading input</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLoading input\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Loading Time:                   </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">1.552</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLoading Time:                   \u001b[0m\u001b[1;31m1.552\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Masking low-entropy (to disable, set </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">'mle=f'</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mMasking low-entropy \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31mto disable, set \u001b[0m\u001b[1;31m'\u001b[0m\u001b[1;31mmle\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31mf\u001b[0m\u001b[1;31m'\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Low Complexity Masking Time:    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.989</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLow Complexity Masking Time:    \u001b[0m\u001b[1;31m0.989\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Ref Bases:                 </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">431916037</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">436.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">73m bases/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mRef Bases:                 \u001b[0m\u001b[1;31m431916037\u001b[0m\u001b[1;31m    \u001b[0m\u001b[1;31m436.\u001b[0m\u001b[1;31m73m bases/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Low Complexity Bases:          </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">17198</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLow Complexity Bases:          \u001b[0m\u001b[1;31m17198\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Masking from sam</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mMasking from sam\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Masking Time:               </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.154</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Masking Time:               \u001b[0m\u001b[1;31m0.154\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Reads Processed:             </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">587</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">3.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">82k reads/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Reads Processed:             \u001b[0m\u001b[1;31m587\u001b[0m\u001b[1;31m    \u001b[0m\u001b[1;31m3.\u001b[0m\u001b[1;31m82k reads/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Bases Processed:          </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">116335</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">76m bases/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Bases Processed:          \u001b[0m\u001b[1;31m116335\u001b[0m\u001b[1;31m    \u001b[0m\u001b[1;31m0.\u001b[0m\u001b[1;31m76m bases/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Bases Masked:             </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">109930</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Bases Masked:             \u001b[0m\u001b[1;31m109930\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Converting masked bases to N</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mConverting masked bases to N\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Done Masking</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mDone Masking\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Conversion Time:                </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.546</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mConversion Time:                \u001b[0m\u001b[1;31m0.546\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Writing output</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWriting output\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Writing Time:                   </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.660</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWriting Time:                   \u001b[0m\u001b[1;31m0.660\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Total Bases Masked:           </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">127128</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">/</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">431916037</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.029</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTotal Bases Masked:           \u001b[0m\u001b[1;31m127128\u001b[0m\u001b[1;31m/\u001b[0m\u001b[1;31m431916037\u001b[0m\u001b[1;31m  \u001b[0m\u001b[1;31m0.029\u001b[0m\u001b[1;31m%\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Total Time:                     </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">4.004</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTotal Time:                     \u001b[0m\u001b[1;31m4.004\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:56:07</span> - Finished bbmask step                                                                                    <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#235\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">235</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:56:07\u001b[0m - Finished bbmask step                                                                                    \u001b]8;id=776703;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=18232;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#235\u001b\\\u001b[2m235\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:56:07</span> - Masking completed. Output saved to                                                                      <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#271\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">271</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">silva_rrna_sequences_deduplicated_masked.fasta</span>     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:56:07\u001b[0m - Masking completed. Output saved to                                                                      \u001b]8;id=219745;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=341603;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#271\u001b\\\u001b[2m271\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95msilva_rrna_sequences_deduplicated_masked.fasta\u001b[0m     \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">java -ea  --add-modules jdk.incubator.vector -Xmx8750m -Xms8750m -cp </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">bbmap/current/ jgi.BBDuk </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in1</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated_ma</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sked.fasta </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated_ma</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sked_entropy.fasta </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropy</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropyk</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropywindow</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">24</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> maskentropy </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">ziplevel</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">9</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mjava -ea  --add-modules jdk.incubator.vector -Xmx8750m -Xms8750m -cp \u001b[0m\n",
       "\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/\u001b[0m\n",
       "\u001b[1;31mbbmap/current/\u001b[0m\u001b[1;31m jgi.BBDuk \u001b[0m\n",
       "\u001b[1;31min1\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31msilva_rrna_sequences_deduplicated_ma\u001b[0m\n",
       "\u001b[1;31msked.fasta\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31msilva_rrna_sequences_deduplicated_ma\u001b[0m\n",
       "\u001b[1;31msked_entropy.fasta\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mentropy\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m.4\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mentropyk\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m4\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mentropywindow\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m24\u001b[0m\u001b[1;31m maskentropy \u001b[0m\u001b[1;31mziplevel\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m9\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">WARNING: Using incubator modules: jdk.incubator.vector</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWARNING: Using incubator modules: jdk.incubator.vector\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Executing jgi.BBDuk </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">[</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in1</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated_m</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">asked.fasta, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rrna_sequences_deduplicated_ma</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sked_entropy.fasta, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropy</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropyk</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropywindow</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">24</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, maskentropy, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">ziplevel</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">9</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mExecuting jgi.BBDuk \u001b[0m\n",
       "\u001b[1;31m[\u001b[0m\u001b[1;31min1\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31msilva_rrna_sequences_deduplicated_m\u001b[0m\n",
       "\u001b[1;31masked.fasta\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31msilva_rrna_sequences_deduplicated_ma\u001b[0m\n",
       "\u001b[1;31msked_entropy.fasta\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mentropy\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m.4\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mentropyk\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m4\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mentropywindow\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m24\u001b[0m\u001b[1;31m, maskentropy, \u001b[0m\u001b[1;31mziplevel\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m9\u001b[0m\u001b[1;31m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Version </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">39.55</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mVersion \u001b[0m\u001b[1;31m39.55\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Allocating kmer table:  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.035</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mAllocating kmer table:  \u001b[0m\u001b[1;31m0.035\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Initial:</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInitial:\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Memory: </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">max</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">9177m</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">total</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">9177m</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">free</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">9120m</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">used</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">57m</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mMemory: \u001b[0m\u001b[1;31mmax\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m9177m\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mtotal\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m9177m\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mfree\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m9120m\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mused\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m57m\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Input is being processed as unpaired</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInput is being processed as unpaired\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Started output streams: </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.051</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mStarted output streams: \u001b[0m\u001b[1;31m0.051\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Processing time:                </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">1.756</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mProcessing time:        \u001b[0m\u001b[1;31m        \u001b[0m\u001b[1;31m1.756\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Input:                          </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">226832</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads            </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">431916037</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInput:                          \u001b[0m\u001b[1;31m226832\u001b[0m\u001b[1;31m reads    \u001b[0m\u001b[1;31m        \u001b[0m\u001b[1;31m431916037\u001b[0m\u001b[1;31m bases.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Entropy-masked:                 </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">255</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.11</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)       </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">7917</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mEntropy-masked:                 \u001b[0m\u001b[1;31m255\u001b[0m\u001b[1;31m reads \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.11\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\u001b[1;31m       \u001b[0m\u001b[1;31m7917\u001b[0m\u001b[1;31m bases \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Total Removed:                  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)         </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTotal Removed:                  \u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m reads \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\u001b[1;31m         \u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m bases \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Result:                         </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">226832</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">100.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">431916037</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">100.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mResult:                         \u001b[0m\u001b[1;31m226832\u001b[0m\u001b[1;31m reads \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m100.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\u001b[1;31m  \u001b[0m\u001b[1;31m431916037\u001b[0m\u001b[1;31m bases \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m100.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Time:                           </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">1.808</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTime:                           \u001b[0m\u001b[1;31m1.808\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Reads Processed:        226k    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">125.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">46k reads/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mReads Processed:        226k    \u001b[0m\u001b[1;31m125.\u001b[0m\u001b[1;31m46k reads/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Bases Processed:        431m    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">238.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">89m bases/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mBases Processed:        431m    \u001b[0m\u001b[1;31m238.\u001b[0m\u001b[1;31m89m bases/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished writing SILVA rRNA sequences parquet file.\n",
      "finished writing masked SILVA rRNA fasta file: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/silva_rRNA_all_sequences_masked_entropy.fasta\n"
     ]
    }
   ],
   "source": [
    "# mask shared kmers with RNA viruses\n",
    "masked_silva_fasta_path = os.path.join(\n",
    "    rrna_dir, \"silva_rrna_sequences_deduplicated_masked.fasta\"\n",
    ")\n",
    "entrophy_masked_silva_fasta_path = os.path.join(\n",
    "    rrna_dir, \"silva_rrna_sequences_deduplicated_masked_entropy.fasta\"\n",
    ")\n",
    "\n",
    "mask_args = {\n",
    "    \"threads\": 14,\n",
    "    \"memory\": \"18gb\",\n",
    "    \"output\": str(masked_silva_fasta_path),\n",
    "    \"flatten\": False,\n",
    "    \"aligner\": \"mmseqs2\",\n",
    "    \"input\": str(deduplicated_silva_fasta_path),\n",
    "}\n",
    "\n",
    "# ensure we have the click Command object (the package import may have bound the module)\n",
    "from rolypoly.commands.reads.mask_dna import mask_dna as _mask_cmd\n",
    "\n",
    "# If it's a click.Command use a Context to invoke, otherwise call directly\n",
    "if isinstance(_mask_cmd, click.core.Command):\n",
    "    ctx = click.Context(_mask_cmd)\n",
    "    ctx.invoke(_mask_cmd, **mask_args)\n",
    "else:\n",
    "    # fallback: call the function directly\n",
    "    _mask_cmd(**mask_args)\n",
    "\n",
    "# # Apply entropy masking\n",
    "bbduk(\n",
    "    in1=masked_silva_fasta_path,\n",
    "    out=entrophy_masked_silva_fasta_path,\n",
    "    entropy=0.4,\n",
    "    entropyk=4,\n",
    "    entropywindow=24,\n",
    "    maskentropy=True,\n",
    "    ziplevel=9,\n",
    ")\n",
    "\n",
    "\n",
    "# read the masked fasta and remap to original headers\n",
    "silva_rrna_masked_df = from_fastx_eager(\n",
    "    entrophy_masked_silva_fasta_path\n",
    ").rename({\"header\": \"tmp_header\", \"sequence\": \"masked_sequence\"})\n",
    "silva_rrna_masked_df = silva_rrna_masked_df.join(\n",
    "    silva_df, on=[\"tmp_header\"], how=\"inner\"\n",
    ")\n",
    "silva_rrna_masked_df.write_parquet(\n",
    "    os.path.join(rrna_dir, \"silva_rrna_sequences_masked.parquet\")\n",
    ")\n",
    "print(\"Finished writing SILVA rRNA sequences parquet file.\")\n",
    "write_fasta_file(\n",
    "    headers=silva_rrna_masked_df[\"new_header\"].to_list(),\n",
    "    seqs=silva_rrna_masked_df[\"masked_sequence\"].to_list(),\n",
    "    output_file=os.path.join(\n",
    "        rrna_dir, \"silva_rRNA_all_sequences_masked_entropy.fasta\"\n",
    "    ),\n",
    ")\n",
    "print(\n",
    "    f\"finished writing masked SILVA rRNA fasta file: {os.path.join(rrna_dir, 'silva_rRNA_all_sequences_masked_entropy.fasta')}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff910e05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">21:12:54</span> - Downloading <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt</span> to                 <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">assembly_summary_genbank.txt.gz</span>                     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m21:12:54\u001b[0m - Downloading \u001b[4;94mhttps://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt\u001b[0m to                 \u001b]8;id=200560;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=689039;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95massembly_summary_genbank.txt.gz\u001b[0m                     \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">21:13:31</span> - Successfully downloaded to                                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">assembly_summary_genbank.txt.gz</span>                     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m21:13:31\u001b[0m - Successfully downloaded to                                                                               \u001b]8;id=244576;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=981640;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95massembly_summary_genbank.txt.gz\u001b[0m                     \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading NCBI GenBank assembly summary\n"
     ]
    }
   ],
   "source": [
    "fetch_and_extract(\n",
    "    url=\"https://ftp.ncbi.nlm.nih.gov/genomes/genbank/assembly_summary_genbank.txt\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"assembly_summary_genbank.txt.gz\"),\n",
    "    extract=False,\n",
    ")\n",
    "print(\"Loading NCBI GenBank assembly summary\")\n",
    "# genbank_summary = pl.read_csv(os.path.join(rrna_dir, \"assembly_summary_genbank.txt.gz\",),\n",
    "# infer_schema_length=100020, separator=\"\\t\", skip_rows=1,\n",
    "# null_values=[\"na\",\"NA\",\"-\"],ignore_errors=True,\n",
    "# has_header=True)\n",
    "# polars failed me, so using line by line iterator\n",
    "from gzip import open as gz_open\n",
    "\n",
    "with gz_open(\n",
    "    os.path.join(rrna_dir, \"assembly_summary_genbank.txt.gz\"), \"r\"\n",
    ") as f:\n",
    "    header = None\n",
    "    records = []\n",
    "    i = 0\n",
    "    for line in f:\n",
    "        if i == 0:\n",
    "            i += 1\n",
    "            continue\n",
    "        line = line.rstrip(b\"\\n\")\n",
    "        if i == 1:\n",
    "            header = line.decode()[1:].strip().split(\"\\t\")\n",
    "            i += 1\n",
    "            continue\n",
    "        fields = line.decode().strip().split(\"\\t\")\n",
    "        record = dict(zip(header, fields))\n",
    "        records.append(record)\n",
    "genbank_summary = pl.from_records(records).rename({\"taxid\": \"ncbi_taxonid\"})\n",
    "genbank_summary.collect_schema()\n",
    "# Schema([('assembly_accession', String),\n",
    "#         ('bioproject', String),\n",
    "#         ('biosample', String),\n",
    "#         ('wgs_master', String),\n",
    "#         ('refseq_category', String),\n",
    "#         ('ncbi_taxonid', String),\n",
    "#         ('species_taxid', String),\n",
    "#         ('organism_name', String),\n",
    "#         ('infraspecific_name', String),\n",
    "#         ('isolate', String),\n",
    "#         ('version_status', String),\n",
    "#         ('assembly_level', String),\n",
    "#         ('release_type', String),\n",
    "#         ('genome_rep', String),\n",
    "#         ('seq_rel_date', String),\n",
    "#         ('asm_name', String),\n",
    "#         ('asm_submitter', String),\n",
    "#         ('gbrs_paired_asm', String),\n",
    "#         ('paired_asm_comp', String),\n",
    "#         ('ftp_path', String),\n",
    "#         ('excluded_from_refseq', String),\n",
    "#         ('relation_to_type_material', String),\n",
    "#         ('asm_not_live_date', String),\n",
    "#         ('assembly_type', String),\n",
    "#         ('group', String),\n",
    "#         ('genome_size', String),\n",
    "#         ('genome_size_ungapped', String),\n",
    "#         ('gc_percent', String),\n",
    "#         ('replicon_count', String),\n",
    "#         ('scaffold_count', String),\n",
    "#         ('contig_count', String),\n",
    "#         ('annotation_provider', String),\n",
    "#         ('annotation_name', String),\n",
    "#         ('annotation_date', String),\n",
    "#         ('total_gene_count', String),\n",
    "#         ('protein_coding_gene_count', String),\n",
    "#         ('non_coding_gene_count', String),\n",
    "#         ('pubmed_id', String)])\n",
    "\n",
    "genbank_summary.write_csv(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.tsv\"), separator=\"\\t\"\n",
    ")\n",
    "genbank_summary = pl.read_csv(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.tsv\"),\n",
    "    infer_schema_length=100020,\n",
    "    separator=\"\\t\",\n",
    "    null_values=[\"na\", \"NA\", \"-\"],\n",
    "    ignore_errors=True,\n",
    "    has_header=True,\n",
    ")\n",
    "\n",
    "# In [91]: genbank_summary.collect_schema()\n",
    "# Out[91]:\n",
    "# Schema([('assembly_accession', String),\n",
    "#         ('bioproject', String),\n",
    "#         ('biosample', String),\n",
    "#         ('wgs_master', String),\n",
    "#         ('refseq_category', String),\n",
    "#         ('ncbi_taxonid', Int64),\n",
    "#         ('species_taxid', Int64),\n",
    "#         ('organism_name', String),\n",
    "#         ('infraspecific_name', String),\n",
    "#         ('isolate', String),\n",
    "#         ('version_status', String),\n",
    "#         ('assembly_level', String),\n",
    "#         ('release_type', String),\n",
    "#         ('genome_rep', String),\n",
    "#         ('seq_rel_date', String),\n",
    "#         ('asm_name', String),\n",
    "#         ('asm_submitter', String),\n",
    "#         ('gbrs_paired_asm', String),\n",
    "#         ('paired_asm_comp', String),\n",
    "#         ('ftp_path', String),\n",
    "#         ('excluded_from_refseq', String),\n",
    "#         ('relation_to_type_material', String),\n",
    "#         ('asm_not_live_date', String),\n",
    "#         ('assembly_type', String),\n",
    "#         ('group', String),\n",
    "#         ('genome_size', Int64),\n",
    "#         ('genome_size_ungapped', Int64),\n",
    "#         ('gc_percent', Float64),\n",
    "#         ('replicon_count', Int64),\n",
    "#         ('scaffold_count', Int64),\n",
    "#         ('contig_count', Int64),\n",
    "#         ('annotation_provider', String),\n",
    "#         ('annotation_name', String),\n",
    "#         ('annotation_date', String),\n",
    "#         ('total_gene_count', Int64),\n",
    "#         ('protein_coding_gene_count', Int64),\n",
    "#         ('non_coding_gene_count', Int64),\n",
    "#         ('pubmed_id', String)])\n",
    "\n",
    "genbank_summary.write_parquet(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.parquet\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c74b962",
   "metadata": {},
   "source": [
    "Now the same for refseq as I'm not sure genbank contain all of the stuff in refseq?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67fd5300",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">21:17:12</span> - Downloading <span style=\"color: #0000ff; text-decoration-color: #0000ff; text-decoration: underline\">https://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt</span> to                   <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">215</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">assembly_summary_refseq.txt.gz</span>                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m21:17:12\u001b[0m - Downloading \u001b[4;94mhttps://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt\u001b[0m to                   \u001b]8;id=867597;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=118293;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#215\u001b\\\u001b[2m215\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95massembly_summary_refseq.txt.gz\u001b[0m                      \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">21:17:20</span> - Successfully downloaded to                                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">various.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">assembly_summary_refseq.txt.gz</span>                      <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">              </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m21:17:20\u001b[0m - Successfully downloaded to                                                                               \u001b]8;id=777524;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py\u001b\\\u001b[2mvarious.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=5085;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/utils/various.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95massembly_summary_refseq.txt.gz\u001b[0m                      \u001b[2m              \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading NCBI RefSeq assembly summary\n"
     ]
    }
   ],
   "source": [
    "fetch_and_extract(\n",
    "    url=\"https://ftp.ncbi.nlm.nih.gov/genomes/refseq/assembly_summary_refseq.txt\",\n",
    "    fetched_to=os.path.join(rrna_dir, \"assembly_summary_refseq.txt.gz\"),\n",
    "    extract=False,\n",
    ")\n",
    "print(\"Loading NCBI RefSeq assembly summary\")\n",
    "# refseq_summary = pl.read_csv(os.path.join(rrna_dir, \"assembly_summary_refseq.txt.gz\",),\n",
    "# infer_schema_length=100020, separator=\"\\t\", skip_rows=1,\n",
    "# null_values=[\"na\",\"NA\",\"-\"],ignore_errors=True,\n",
    "# has_header=True)\n",
    "# polars failed me, so using line by line iterator\n",
    "from gzip import open as gz_open\n",
    "\n",
    "with gz_open(\n",
    "    os.path.join(rrna_dir, \"assembly_summary_refseq.txt.gz\"), \"r\"\n",
    ") as f:\n",
    "    header = None\n",
    "    records = []\n",
    "    i = 0\n",
    "    for line in f:\n",
    "        if i == 0:\n",
    "            i += 1\n",
    "            continue\n",
    "        line = line.rstrip(b\"\\n\")\n",
    "        if i == 1:\n",
    "            header = line.decode()[1:].strip().split(\"\\t\")\n",
    "            i += 1\n",
    "            continue\n",
    "        fields = line.decode().strip().split(\"\\t\")\n",
    "        record = dict(zip(header, fields))\n",
    "        records.append(record)\n",
    "refseq_summary = pl.from_records(records).rename({\"taxid\": \"ncbi_taxonid\"})\n",
    "refseq_summary.collect_schema()\n",
    "# Schema([('assembly_accession', String),\n",
    "#         ('bioproject', String),\n",
    "#         ('biosample', String),\n",
    "#         ('wgs_master', String),\n",
    "#         ('refseq_category', String),\n",
    "#         ('ncbi_taxonid', String),\n",
    "#         ('species_taxid', String),\n",
    "#         ('organism_name', String),\n",
    "#         ('infraspecific_name', String),\n",
    "#         ('isolate', String),\n",
    "#         ('version_status', String),\n",
    "#         ('assembly_level', String),\n",
    "#         ('release_type', String),\n",
    "#         ('genome_rep', String),\n",
    "#         ('seq_rel_date', String),\n",
    "#         ('asm_name', String),\n",
    "#         ('asm_submitter', String),\n",
    "#         ('gbrs_paired_asm', String),\n",
    "#         ('paired_asm_comp', String),\n",
    "#         ('ftp_path', String),\n",
    "#         ('excluded_from_refseq', String),\n",
    "#         ('relation_to_type_material', String),\n",
    "#         ('asm_not_live_date', String),\n",
    "#         ('assembly_type', String),\n",
    "#         ('group', String),\n",
    "#         ('genome_size', String),\n",
    "#         ('genome_size_ungapped', String),\n",
    "#         ('gc_percent', String),\n",
    "#         ('replicon_count', String),\n",
    "#         ('scaffold_count', String),\n",
    "#         ('contig_count', String),\n",
    "#         ('annotation_provider', String),\n",
    "#         ('annotation_name', String),\n",
    "#         ('annotation_date', String),\n",
    "#         ('total_gene_count', String),\n",
    "#         ('protein_coding_gene_count', String),\n",
    "#         ('non_coding_gene_count', String),\n",
    "#         ('pubmed_id', String)])\n",
    "\n",
    "refseq_summary.write_csv(\n",
    "    os.path.join(rrna_dir, \"refseq_assembly_summary.tsv\"), separator=\"\\t\"\n",
    ")\n",
    "refseq_summary = pl.read_csv(\n",
    "    os.path.join(rrna_dir, \"refseq_assembly_summary.tsv\"),\n",
    "    infer_schema_length=100020,\n",
    "    separator=\"\\t\",\n",
    "    null_values=[\"na\", \"NA\", \"-\"],\n",
    "    ignore_errors=True,\n",
    "    has_header=True,\n",
    ")\n",
    "\n",
    "# In [91]: refseq_summary.collect_schema()\n",
    "# # Out[91]:\n",
    "# Schema([('assembly_accession', String),\n",
    "#         ('bioproject', String),\n",
    "#         ('biosample', String),\n",
    "#         ('wgs_master', String),\n",
    "#         ('refseq_category', String),\n",
    "#         ('ncbi_taxonid', Int64),\n",
    "#         ('species_taxid', Int64),\n",
    "#         ('organism_name', String),\n",
    "#         ('infraspecific_name', String),\n",
    "#         ('isolate', String),\n",
    "#         ('version_status', String),\n",
    "#         ('assembly_level', String),\n",
    "#         ('release_type', String),\n",
    "#         ('genome_rep', String),\n",
    "#         ('seq_rel_date', String),\n",
    "#         ('asm_name', String),\n",
    "#         ('asm_submitter', String),\n",
    "#         ('gbrs_paired_asm', String),\n",
    "#         ('paired_asm_comp', String),\n",
    "#         ('ftp_path', String),\n",
    "#         ('excluded_from_refseq', String),\n",
    "#         ('relation_to_type_material', String),\n",
    "#         ('asm_not_live_date', String),\n",
    "#         ('assembly_type', String),\n",
    "#         ('group', String),\n",
    "#         ('genome_size', Int64),\n",
    "#         ('genome_size_ungapped', Int64),\n",
    "#         ('gc_percent', Float64),\n",
    "#         ('replicon_count', Int64),\n",
    "#         ('scaffold_count', Int64),\n",
    "#         ('contig_count', Int64),\n",
    "#         ('annotation_provider', String),\n",
    "#         ('annotation_name', String),\n",
    "#         ('annotation_date', String),\n",
    "#         ('total_gene_count', Int64),\n",
    "#         ('protein_coding_gene_count', Int64),\n",
    "#         ('non_coding_gene_count', Int64),\n",
    "#         ('pubmed_id', String)])\n",
    "refseq_summary.write_parquet(\n",
    "    os.path.join(rrna_dir, \"refseq_assembly_summary.parquet\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4571252",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'total_gene_count', 'version_status', 'annotation_date', 'non_coding_gene_count', 'biosample', 'pubmed_id', 'bioproject', 'species_taxid', 'assembly_type', 'scaffold_count', 'protein_coding_gene_count', 'assembly_accession', 'ncbi_taxonid', 'seq_rel_date', 'genome_size', 'asm_submitter', 'genome_size_ungapped', 'annotation_name', 'genome_rep', 'isolate', 'asm_name', 'wgs_master', 'replicon_count', 'refseq_category', 'infraspecific_name', 'organism_name', 'gbrs_paired_asm', 'group', 'assembly_level', 'excluded_from_refseq', 'paired_asm_comp', 'ftp_path', 'gc_percent', 'release_type', 'contig_count', 'asm_not_live_date', 'relation_to_type_material', 'annotation_provider'}\n",
      "38\n",
      "Total of 0 RefSeq assemblies found in GenBank assemblies\n",
      "Total of 3718859 unique assemblies found in NCBI GenBank and RefSeq\n"
     ]
    }
   ],
   "source": [
    "genbank_summary = pl.read_parquet(\n",
    "    os.path.join(rrna_dir, \"genbank_assembly_summary.parquet\")\n",
    ")\n",
    "refseq_summary = pl.read_parquet(\n",
    "    os.path.join(rrna_dir, \"refseq_assembly_summary.parquet\")\n",
    ")\n",
    "test = set(genbank_summary.columns).intersection(set(refseq_summary.columns))\n",
    "print(test)\n",
    "print(len(test))\n",
    "# is refseqs in genbank?\n",
    "refseq_in_genbank = refseq_summary.join(\n",
    "    genbank_summary.select([\"assembly_accession\"]).unique(),\n",
    "    on=[\"assembly_accession\"],\n",
    "    how=\"inner\",\n",
    ")\n",
    "print(\n",
    "    f\"Total of {refseq_in_genbank.height} RefSeq assemblies found in GenBank assemblies\"\n",
    ")\n",
    "ncbi_summary = pl.concat(\n",
    "    [genbank_summary, refseq_summary], how=\"diagonal_relaxed\"\n",
    ").unique(subset=[\"assembly_accession\"])\n",
    "ncbi_summary.write_parquet(\n",
    "    os.path.join(rrna_dir, \"ncbi_assembly_summary.parquet\")\n",
    ")\n",
    "ncbi_summary.write_csv(\n",
    "    os.path.join(rrna_dir, \"ncbi_assembly_summary.tsv\"), separator=\"\\t\"\n",
    ")\n",
    "print(\n",
    "    f\"Total of {ncbi_summary.height} unique assemblies found in NCBI GenBank and RefSeq\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71bcc534",
   "metadata": {},
   "source": [
    "Looks like they are not in genbank, so stacking them"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7de5b83",
   "metadata": {},
   "outputs": [],
   "source": [
    "ncbi_summary = pl.read_parquet(\n",
    "    os.path.join(rrna_dir, \"ncbi_assembly_summary.parquet\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18135e03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered GenBank summary to 244185 unique taxid entries for SILVA sequences\n"
     ]
    }
   ],
   "source": [
    "# next, for every unique ncbi_taxonid, we select the one that has the most protein_coding_gene_count, then refseq_category, then tie breaking with non_coding_gene_count, tie breaking by latest assembly (by seq_rel_date).\n",
    "mini_ncbi = ncbi_summary.sort(\n",
    "    by=[\n",
    "        pl.col(\"refseq_category\").reverse(),\n",
    "        pl.col(\"protein_coding_gene_count\").cast(pl.Int64).reverse(),\n",
    "        pl.col(\"non_coding_gene_count\").cast(pl.Int64).reverse(),\n",
    "        pl.col(\"seq_rel_date\").reverse(),\n",
    "        pl.col(\"genome_size\").reverse(),\n",
    "    ]\n",
    ").unique(subset=[\"ncbi_taxonid\"], keep=\"first\")\n",
    "print(\n",
    "    f\"Filtered GenBank summary to {mini_ncbi.height} unique taxid entries for SILVA sequences\"\n",
    ")\n",
    "\n",
    "# # in_silva = mini_ncbi.filter(pl.col(\"ncbi_taxonid\").is_in(silva_taxmap[\"ncbi_taxonid\"])).unique()\n",
    "# print(in_silva.height)\n",
    "# print(temp_genbank.height)\n",
    "# only 30503 out of ~240k?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4008b431",
   "metadata": {},
   "source": [
    "## rRNAs, tRNA, mtDNA, and plastid-DNA from NCBI BLAST DBs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9bf454",
   "metadata": {},
   "source": [
    "This is a little hacky, but works?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ed29977",
   "metadata": {},
   "source": [
    "rRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6991ebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, getting the contents of https://ftp.ncbi.nlm.nih.gov/blast/db/ and filtering to files containing \"16S\", \"18S\", \"28S\", \"ITS\"\n",
    "# LSU, SSU\n",
    "\n",
    "terms = [\"16S\", \"18S\", \"28S\", \"ITS\", \"LSU\", \"SSU\", \"ribosomal\"]\n",
    "files_2_download = []\n",
    "FTP_HOST = \"ftp.ncbi.nlm.nih.gov\"\n",
    "FTP_DIR = \"/blast/db/\"\n",
    "\n",
    "# List to store the file and directory names\n",
    "file_list = []\n",
    "\n",
    "# Connect to the FTP host\n",
    "with FTP(FTP_HOST) as ftp:\n",
    "    #  Log in anonymously (default behavior for .login())\n",
    "    ftp.login()\n",
    "    print(f\"Connected to {FTP_HOST}\")\n",
    "    ftp.cwd(FTP_DIR)\n",
    "    logger.debug(f\"Changed directory to {FTP_DIR}\")\n",
    "\n",
    "    # nlst() returns only names, not detailed information\n",
    "    file_list = ftp.nlst()\n",
    "\n",
    "    print(\"\\n--- Directory Contents ---\")\n",
    "    for item in file_list:\n",
    "        if any(term in item for term in terms):\n",
    "            files_2_download.append(item)\n",
    "            logger.debug(item)\n",
    "\n",
    "            # Download file directly using FTP\n",
    "            local_path = os.path.join(rrna_dir, item)\n",
    "            print(f\"Downloading {item}\")\n",
    "            with open(local_path, \"wb\") as local_file:\n",
    "                ftp.retrbinary(f\"RETR {item}\", local_file.write)\n",
    "            logger.debug(f\"Successfully downloaded {item}\")\n",
    "\n",
    "print(f\"Downloaded {len(files_2_download)} files matching terms: {terms}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca9f924a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each of these, will run blastdbcmd to extract sequences into fasta\n",
    "for item in files_2_download:\n",
    "    if not item.endswith(\".tar.gz\"):\n",
    "        continue\n",
    "    print(f\"Processing {item}...\")\n",
    "    extract(archive_path=os.path.join(rrna_dir, item), extract_to=rrna_dir)\n",
    "    # blastdbcmd -entry all -db LSU_prokaryote_rRNA  -out reference.fasta -outfmt \"%T;%t;%s\" # taxid;header;sequence\n",
    "    run_command_comp(\n",
    "        base_cmd=\"blastdbcmd\",\n",
    "        positional_args=[\n",
    "            \"-entry\",\n",
    "            \"all\",\n",
    "            \"-db\",\n",
    "            os.path.join(rrna_dir, item.replace(\".tar.gz\", \"\")),\n",
    "            \"-out\",\n",
    "            os.path.join(\n",
    "                rrna_dir,\n",
    "                item.replace(\".tar.gz\", \"\").replace(\".tar\", \"\") + \".tab\",\n",
    "            ),\n",
    "            \"-outfmt\",\n",
    "            str('\"%T@%t@%K@%s\"'),\n",
    "        ],\n",
    "        positional_args_location=\"end\",\n",
    "        params={},\n",
    "        logger=logger,\n",
    "        output_file=os.path.join(\n",
    "            rrna_dir, item.replace(\".tar.gz\", \"\").replace(\".tar\", \"\") + \".tab\"\n",
    "        ),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78a669bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 1 files to remove\n",
      "Keeping 54 files (.tab, .fasta, .parquet, .tsv, .gz)\n",
      "Removed: 18S_fungal_sequences.tar.gz.md5\n",
      "Cleanup complete\n"
     ]
    }
   ],
   "source": [
    "# get all files to remove (everything but the .tab files)\n",
    "all_files_in_rrna = glob.glob(os.path.join(rrna_dir, \"*\"))\n",
    "all_tab_files = glob.glob(os.path.join(rrna_dir, \"*.tab\"))\n",
    "all_fasta_files = glob.glob(os.path.join(rrna_dir, \"*.fasta\"))\n",
    "\n",
    "# Files to keep (tab files, fasta files, and parquet/tsv files)\n",
    "files_to_keep = set(\n",
    "    all_tab_files\n",
    "    + all_fasta_files\n",
    "    + glob.glob(os.path.join(rrna_dir, \"*.parquet\"))\n",
    "    + glob.glob(os.path.join(rrna_dir, \"*.tsv\"))\n",
    "    + glob.glob(os.path.join(rrna_dir, \"*.gz\"))\n",
    ")\n",
    "\n",
    "# Get all files to remove (exclude directories and files to keep)\n",
    "files_2_remove = [\n",
    "    f for f in all_files_in_rrna if os.path.isfile(f) and f not in files_to_keep\n",
    "]\n",
    "\n",
    "print(f\"Found {len(files_2_remove)} files to remove\")\n",
    "print(f\"Keeping {len(files_to_keep)} files (.tab, .fasta, .parquet, .tsv, .gz)\")\n",
    "\n",
    "# Remove the files\n",
    "for file_path in files_2_remove:\n",
    "    try:\n",
    "        os.remove(file_path)\n",
    "        print(f\"Removed: {os.path.basename(file_path)}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to remove {os.path.basename(file_path)}: {e}\")\n",
    "\n",
    "print(\"Cleanup complete\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9b854f6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read 159996 rRNA sequences from\n"
     ]
    }
   ],
   "source": [
    "rrna_df = pl.scan_csv(\n",
    "    os.path.join(rrna_dir, \"*.tab\"),\n",
    "    separator=\"@\",\n",
    "    has_header=False,\n",
    "    null_values=[\"N/A\"],\n",
    "    new_columns=[\"taxid\", \"header\", \"name\", \"sequence\"],\n",
    "    include_file_paths=\"type\",\n",
    ").collect()\n",
    "rrna_df = rrna_df.drop(\"name\")\n",
    "rrna_df = rrna_df.with_columns(\n",
    "    rRNA_type=pl.col(\"type\").str.extract(r\"([^/]+)\\.tab$\", 1)\n",
    ")\n",
    "rrna_df = rrna_df.drop(\"type\")\n",
    "print(f\"Read {rrna_df.height} rRNA sequences from\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "143e16a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (8, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>rRNA_type</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;ITS_eukaryote_sequences&quot;</td><td>77582</td></tr><tr><td>&quot;16S_ribosomal_RNA&quot;</td><td>27708</td></tr><tr><td>&quot;ITS_RefSeq_Fungi&quot;</td><td>19541</td></tr><tr><td>&quot;28S_fungal_sequences&quot;</td><td>11754</td></tr><tr><td>&quot;SSU_eukaryote_rRNA&quot;</td><td>8784</td></tr><tr><td>&quot;LSU_eukaryote_rRNA&quot;</td><td>6575</td></tr><tr><td>&quot;LSU_prokaryote_rRNA&quot;</td><td>4047</td></tr><tr><td>&quot;18S_fungal_sequences&quot;</td><td>4005</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (8, 2)\n",
       "┌─────────────────────────┬───────┐\n",
       "│ rRNA_type               ┆ count │\n",
       "│ ---                     ┆ ---   │\n",
       "│ str                     ┆ u32   │\n",
       "╞═════════════════════════╪═══════╡\n",
       "│ ITS_eukaryote_sequences ┆ 77582 │\n",
       "│ 16S_ribosomal_RNA       ┆ 27708 │\n",
       "│ ITS_RefSeq_Fungi        ┆ 19541 │\n",
       "│ 28S_fungal_sequences    ┆ 11754 │\n",
       "│ SSU_eukaryote_rRNA      ┆ 8784  │\n",
       "│ LSU_eukaryote_rRNA      ┆ 6575  │\n",
       "│ LSU_prokaryote_rRNA     ┆ 4047  │\n",
       "│ 18S_fungal_sequences    ┆ 4005  │\n",
       "└─────────────────────────┴───────┘"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrna_df[\"rRNA_type\"].value_counts(sort=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af59267e",
   "metadata": {},
   "source": [
    "### rename fasta seqs to contam notation\n",
    "for files to be used in filtering reads later on, the headers should bew parseable by other rolypoly functions. So renaming them here.\n",
    "`taxid@Source@type@seq_md5_hash`\n",
    "Using @ as separator as it's not common in fasta headers, and just going to assume the original header doesn't contain @, and that mappers who drop everything after the first white space will just ignore the really long headers\n",
    "```fasta\n",
    ">2888294@NCBI@ITS_RefSeq_Fungi@55c18f2cfc8490f5d5cb9b34162a44f5\n",
    "atgtgatgaga...\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb8e54bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2888294@NCBI@ITS_RefSeq_Fungi@55c18f2cfc8490f5d5cb9b34162a44f5'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrna_df = rrna_df.with_columns(\n",
    "    t1=pl.concat_str(\n",
    "        [pl.col(\"taxid\"), pl.lit(\"NCBI\"), pl.col(\"rRNA_type\")], separator=\"@\"\n",
    "    )\n",
    ")\n",
    "\n",
    "rrna_df = rrna_df.with_columns(\n",
    "    pl.col(\"sequence\")\n",
    "    .map_elements(lambda x: hashlib.md5(x.encode()).hexdigest())\n",
    "    .alias(\"seq_hash\")\n",
    ").with_columns(\n",
    "    new_header=pl.concat_str([pl.col(\"t1\"), pl.col(\"seq_hash\")], separator=\"@\")\n",
    ")\n",
    "rrna_df = rrna_df.unique()\n",
    "rrna_df = rrna_df.unique(subset=[\"sequence\"])\n",
    "\n",
    "\n",
    "rrna_df[\"new_header\"].head(1).item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4c31f12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrna_df.height == rrna_df.with_columns(\n",
    "    header_hash=pl.col(\"header\").map_elements(\n",
    "        lambda x: hashlib.md5(x.encode()).hexdigest()\n",
    "    )\n",
    ").select(\"header_hash\").unique().height"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a811456c",
   "metadata": {},
   "source": [
    "Maybe some have more than 1 unique SSU/LSU?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d6f629e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "104"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rrna_df.with_columns(length=pl.col(\"sequence\").str.len_chars()).select(pl.col(\"length\").min()).item()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfe943ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote combined ncbi rRNA fasta to /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ncbi_rRNA_all_sequences.fasta\n"
     ]
    }
   ],
   "source": [
    "rrna_fasta_path = os.path.join(rrna_dir, \"ncbi_rRNA_all_sequences.fasta\")\n",
    "\n",
    "rrna_df = rrna_df.filter(\n",
    "    ~pl.col(\"taxid\").is_in(\n",
    "        [77133, 155900, 115547, 256318, 100272, 136703, 171953, 86473, 56765]\n",
    "    )\n",
    ")\n",
    "write_fasta_file(\n",
    "    seqs=rrna_df[\"sequence\"].to_list(),\n",
    "    headers=rrna_df[\"new_header\"].to_list(),\n",
    "    output_file=rrna_fasta_path,\n",
    ")\n",
    "print(f\"Wrote combined ncbi rRNA fasta to {rrna_fasta_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72fc161e",
   "metadata": {},
   "source": [
    "Next is the esame entropy and RNA virus masking as above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "938958ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:47:10</span> - datadir used: <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">data</span>                             <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#71\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">71</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:47:10\u001b[0m - datadir used: \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/\u001b[0m\u001b[95mdata\u001b[0m                             \u001b]8;id=933960;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=343840;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#71\u001b\\\u001b[2m71\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running command: mmseqs easy-linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/masking/combined_entropy_masked.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna  --min-seq-id 0.7 --min-aln-len 80 --threads 15 --format-mode 1 --search-type 3 -v 3 --max-accept 1231 -a  \n",
      "easy-linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/masking/combined_entropy_masked.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna --min-seq-id 0.7 --min-aln-len 80 --threads 15 --format-mode 1 --search-type 3 -v 3 --max-accept 1231 -a \n",
      "\n",
      "MMseqs Version:                    \t18.8cc5c\n",
      "Substitution matrix                \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace                      \ttrue\n",
      "Alignment mode                     \t3\n",
      "Alignment mode                     \t0\n",
      "Allow wrapped scoring              \tfalse\n",
      "E-value threshold                  \t0.001\n",
      "Seq. id. threshold                 \t0.7\n",
      "Min alignment length               \t80\n",
      "Seq. id. mode                      \t0\n",
      "Alternative alignments             \t0\n",
      "Coverage threshold                 \t0\n",
      "Coverage mode                      \t0\n",
      "Max sequence length                \t65535\n",
      "Compositional bias                 \t1\n",
      "Compositional bias scale           \t1\n",
      "Max reject                         \t2147483647\n",
      "Max accept                         \t1231\n",
      "Include identical seq. id.         \tfalse\n",
      "Preload mode                       \t0\n",
      "Pseudo count a                     \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                     \tsubstitution:4.100,context:5.800\n",
      "Score bias                         \t0\n",
      "Realign hits                       \tfalse\n",
      "Realign score bias                 \t-0.2\n",
      "Realign max seqs                   \t2147483647\n",
      "Correlation score weight           \t0\n",
      "Gap open cost                      \taa:11,nucl:5\n",
      "Gap extension cost                 \taa:1,nucl:2\n",
      "Zdrop                              \t40\n",
      "Threads                            \t15\n",
      "Compressed                         \t0\n",
      "Verbosity                          \t3\n",
      "Seed substitution matrix           \taa:VTML80.out,nucl:nucleotide.out\n",
      "k-mers per sequence                \t21\n",
      "Scale k-mers per sequence          \taa:0.000,nucl:0.200\n",
      "Mask residues                      \t1\n",
      "Mask residues probability          \t0.9\n",
      "Mask lower case residues           \t0\n",
      "Mask lower letter repeating N times\t0\n",
      "Add N similar to search            \t1\n",
      "Result direction                   \t1\n",
      "Split memory limit                 \t0\n",
      "Min codons in orf                  \t30\n",
      "Max codons in length               \t32734\n",
      "Max orf gaps                       \t2147483647\n",
      "Contig start mode                  \t2\n",
      "Contig end mode                    \t2\n",
      "Orf start mode                     \t1\n",
      "Forward frames                     \t1,2,3\n",
      "Reverse frames                     \t1,2,3\n",
      "Translation table                  \t1\n",
      "Translate orf                      \t0\n",
      "Use all table starts               \tfalse\n",
      "Offset of numeric ids              \t0\n",
      "Create lookup                      \t0\n",
      "Chain overlapping alignments       \t0\n",
      "Merge query                        \t1\n",
      "Search type                        \t3\n",
      "MPI runner                         \t\n",
      "Force restart with latest tmp      \tfalse\n",
      "Remove temporary files             \ttrue\n",
      "Alignment format                   \t1\n",
      "Format alignment output            \tquery,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits\n",
      "Database output                    \tfalse\n",
      "\n",
      "Alignment backtraces will be computed, since they were requested by output format.\n",
      "createdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/masking/combined_entropy_masked.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query --shuffle 0 --threads 15 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[====================================================================\n",
      "Time for merging to query_h: 0h 0m 0s 5ms\n",
      "Time for merging to query: 0h 0m 0s 5ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 1s 977ms\n",
      "createdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences.fasta /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target --shuffle 0 --write-lookup 0 --threads 15 --gpu 0 -v 3 \n",
      "\n",
      "Converting sequences\n",
      "[==============\n",
      "Time for merging to target_h: 0h 0m 0s 5ms\n",
      "Time for merging to target: 0h 0m 0s 8ms\n",
      "Database type: Nucleotide\n",
      "Time for processing: 0h 0m 0s 746ms\n",
      "Create directory /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp\n",
      "createlinindex /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 15 --min-length 30 --max-length 32734 --max-gaps 2147483647 --contig-start-mode 2 --contig-end-mode 2 --orf-start-mode 1 --forward-frames 1,2,3 --reverse-frames 1,2,3 --translation-table 1 --translate 0 --use-all-table-starts 0 --id-offset 0 --create-lookup 0 --compressed 0 --remove-tmp-files 1 \n",
      "\n",
      "createlinindex /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 15 --min-length 30 --max-length 32734 --max-gaps 2147483647 --contig-start-mode 2 --contig-end-mode 2 --orf-start-mode 1 --forward-frames 1,2,3 --reverse-frames 1,2,3 --translation-table 1 --translate 0 --use-all-table-starts 0 --id-offset 0 --create-lookup 0 --compressed 0 --remove-tmp-files 1 \n",
      "\n",
      "splitsequence /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp/15181030682690789959/nucl_split_seq --max-seq-len 65535 --sequence-overlap 0 --sequence-split-mode 1 --headers-split-mode 0 --create-lookup 0 --threads 15 --compressed 0 -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 32ms\n",
      "kmerindexdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp/15181030682690789959/nucl_split_seq /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 15 \n",
      "\n",
      "kmerindexdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp/15181030682690789959/nucl_split_seq /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --hash-shift 67 --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --min-seq-id 0.7 --adjust-kmer-len 0 --split-memory-limit 0 --ignore-multi-kmer 0 --alph-size aa:21,nucl:5 --max-seq-len 65535 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --check-compatible 0 --search-type 3 --spaced-kmer-mode 1 -v 3 --threads 15 \n",
      "\n",
      "Database size: 145146 type: Nucleotide\n",
      "\n",
      "Process file into 1 parts\n",
      "Generate k-mers list 0\n",
      "[=================================================================] 145.15K 1s 186ms\n",
      "\n",
      "Adjusted k-mer length 17\n",
      "\n",
      "Time for fill: 0h 0m 1s 217ms\n",
      "Sort kmer ... Time for sort: 0h 0m 0s 568ms\n",
      "Write VERSION (0)\n",
      "Write META (1)\n",
      "Write ENTRIES (9)\n",
      "Write ENTRIESOFFSETS (10)\n",
      "Write ENTRIESGRIDSIZE (11)\n",
      "Write ENTRIESNUM (12)\n",
      "Write SEQCOUNT (13)\n",
      "Write SCOREMATRIXNAME (2)\n",
      "Write SPACEDPATTERN (23)\n",
      "Write DBR1INDEX (5)\n",
      "Write DBR1DATA (6)\n",
      "Write DBR2INDEX (7)\n",
      "Write DBR2DATA (8)\n",
      "Write HDR1INDEX (18)\n",
      "Write HDR1DATA (19)\n",
      "Write HDR2INDEX (20)\n",
      "Write HDR2DATA (21)\n",
      "Write GENERATOR (22)\n",
      "Time for fill: 0h 0m 0s 570ms\n",
      "Time for merging to target.linidx: 0h 0m 0s 7ms\n",
      "Time for processing: 0h 0m 2s 568ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/index_tmp/15181030682690789959/nucl_split_seq -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 8ms\n",
      "Create directory /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp\n",
      "linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/result /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp -a 1 --alignment-mode 3 --min-seq-id 0.7 --min-aln-len 80 --max-accept 1231 --threads 15 -v 3 --search-type 3 --remove-tmp-files 1 \n",
      "\n",
      "linsearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/result /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp -a 1 --alignment-mode 3 --min-seq-id 0.7 --min-aln-len 80 --max-accept 1231 --threads 15 -v 3 --search-type 3 --remove-tmp-files 1 \n",
      "\n",
      "kmersearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref --seed-sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 0 -c 0 --max-seq-len 65535 --pick-n-sim-kmer 1 --result-direction 1 --split-memory-limit 0 --threads 15 --compressed 0 -v 3 \n",
      "\n",
      "kmersearch /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref --seed-sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --kmer-per-seq 21 --kmer-per-seq-scale aa:0.000,nucl:0.200 --mask 0 --mask-prob 0.9 --mask-lower-case 0 --mask-n-repeat 0 --cov-mode 0 -c 0 --max-seq-len 65535 --pick-n-sim-kmer 1 --result-direction 1 --split-memory-limit 0 --threads 15 --compressed 0 -v 3 \n",
      "\n",
      "Process file into 1 parts\n",
      "[=================================================================] 680.81K 6s 494ms\n",
      "\n",
      "Adjusted k-mer length 17\n",
      "\n",
      "Time for fill: 0h 0m 6s 535ms\n",
      "Sort kmer ... Time for sort: 0h 0m 2s 883ms\n",
      "Time to find k-mers: 0h 0m 0s 376ms\n",
      "Time to sort: 0h 0m 0s 12ms\n",
      "Time for merging to pref: 0h 0m 0s 26ms\n",
      "Time for processing: 0h 0m 10s 478ms\n",
      "rescorediagonal /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/reverse_ungapaln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --rescore-mode 2 --wrapped-scoring 0 --filter-hits 0 -e 0.001 -c 0.9 -a 1 --cov-mode 1 --min-seq-id 0.7 --min-aln-len 80 --seq-id-mode 0 --add-self-matches 0 --sort-results 0 --db-load-mode 0 --threads 15 --compressed 0 -v 3 \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "[=================================================================] 114.22K 0s 488ms\n",
      "Time for merging to reverse_ungapaln: 0h 0m 0s 100ms\n",
      "Time for processing: 0h 0m 0s 843ms\n",
      "filterdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref_filter --filter-file /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/reverse_ungapaln --positive-filter 0 \n",
      "\n",
      "Filtering using file(s)\n",
      "[=================================================================] 114.22K 0s 14ms\n",
      "Time for merging to pref_filter: 0h 0m 0s 209ms\n",
      "Time for processing: 0h 0m 0s 480ms\n",
      "align /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref_filter /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/reverse_aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -a 1 --alignment-mode 3 --alignment-output-mode 0 --wrapped-scoring 0 -e 100000 --min-seq-id 0.7 --min-aln-len 80 --seq-id-mode 0 --alt-ali 0 -c 0 --cov-mode 0 --max-seq-len 65535 --comp-bias-corr 1 --comp-bias-corr-scale 1 --max-rejected 2147483647 --max-accept 1231 --add-self-matches 0 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --score-bias 0 --realign 0 --realign-score-bias -0.2 --realign-max-seqs 2147483647 --corr-score-weight 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --zdrop 40 --threads 15 --compressed 0 -v 3 \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Compute score, coverage and sequence identity\n",
      "Query database size: 145146 type: Nucleotide\n",
      "Target database size: 680809 type: Nucleotide\n",
      "Calculation of alignments\n",
      "[=================================================================] 114.22K 2s 442ms\n",
      "Time for merging to reverse_aln: 0h 0m 0s 113ms\n",
      "935331 alignments calculated\n",
      "523 sequence pairs passed the thresholds (0.000559 of overall calculated)\n",
      "0.004579 hits per query sequence\n",
      "Time for processing: 0h 0m 2s 799ms\n",
      "swapresults /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/reverse_aln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/aln --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' -e 0.001 --split-memory-limit 0 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --threads 15 --compressed 0 --db-load-mode 0 -v 3 \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Computing offsets.\n",
      "[=================================================================] 114.22K 0s 8ms\n",
      "\n",
      "Reading results.\n",
      "[=================================================================] 114.22K 0s 4ms\n",
      "\n",
      "Output database: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/aln\n",
      "[=================================================================] 680.81K 0s 130ms\n",
      "\n",
      "Time for merging to aln: 0h 0m 0s 221ms\n",
      "Time for processing: 0h 0m 0s 493ms\n",
      "swapresults /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/reverse_ungapaln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/ungap_aln \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Computing offsets.\n",
      "[=================================================================] 114.22K 0s 5ms\n",
      "\n",
      "Reading results.\n",
      "[=================================================================] 114.22K 0s 3ms\n",
      "\n",
      "Output database: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/ungap_aln\n",
      "[=================================================================] 680.81K 0s 169ms\n",
      "\n",
      "Time for merging to ungap_aln: 0h 0m 0s 328ms\n",
      "Time for processing: 0h 0m 0s 678ms\n",
      "concatdbs /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/ungap_aln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/aln /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/aln_merged --preserve-keys --take-larger-entry \n",
      "\n",
      "[=================================================================] 680.81K 0s 100ms\n",
      "[=================================================================] 680.81K 0s 154ms\n",
      "Time for merging to aln_merged: 0h 0m 0s 102ms\n",
      "Time for processing: 0h 0m 0s 437ms\n",
      "offsetalignment /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target.linidx /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/aln_merged /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/result \n",
      "\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Index version: 16\n",
      "Generated by:  18.8cc5c\n",
      "ScoreMatrix:  nucleotide.out\n",
      "Computing ORF lookup\n",
      "Computing contig offsets\n",
      "Computing contig lookup\n",
      "Time for contig lookup: 0h 0m 0s 8ms\n",
      "Writing results to: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/result\n",
      "[=================================================================] 680.81K 0s 59ms\n",
      "\n",
      "Time for merging to result: 0h 0m 0s 578ms\n",
      "Time for processing: 0h 0m 1s 47ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/pref -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 10ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/search_tmp/7525828659598130507/reverse_aln -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 31ms\n",
      "convertalis /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/result /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --format-mode 1 --format-output query,target,fident,alnlen,mismatch,gapopen,qstart,qend,tstart,tend,evalue,bits --translation-table 1 --gap-open aa:11,nucl:5 --gap-extend aa:1,nucl:2 --db-output 0 --db-load-mode 0 --search-type 3 --threads 15 --compressed 0 -v 3 \n",
      "\n",
      "[=================================================================] 680.81K 0s 80ms\n",
      "Time for merging to tmp_mapped.sam: 0h 0m 0s 392ms\n",
      "Time for processing: 0h 0m 0s 828ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/result -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 65ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 24ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/target_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 6ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 74ms\n",
      "rmdb /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/2407584767492699454/query_h -v 3 \n",
      "\n",
      "Time for processing: 0h 0m 0s 9ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008000; text-decoration-color: #008000\">File </span><span style=\"color: #008000; text-decoration-color: #008000\">'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam'</span><span style=\"color: #008000; text-decoration-color: #008000\"> size </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">is </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">191029</span><span style=\"color: #008000; text-decoration-color: #008000\"> bytes </span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">(</span><span style=\"color: #008000; text-decoration-color: #008000\">not empty</span><span style=\"color: #008000; text-decoration-color: #008000; font-weight: bold\">)</span><span style=\"color: #008000; text-decoration-color: #008000\">. </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[32mFile \u001b[0m\u001b[32m'/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam'\u001b[0m\u001b[32m size \u001b[0m\n",
       "\u001b[32mis \u001b[0m\u001b[1;32m191029\u001b[0m\u001b[32m bytes \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mnot empty\u001b[0m\u001b[1;32m)\u001b[0m\u001b[32m. \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:47:35</span> - Finished running aligner mmseqs2                                                                        <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#220\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">220</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:47:35\u001b[0m - Finished running aligner mmseqs2                                                                        \u001b]8;id=988782;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=121695;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#220\u001b\\\u001b[2m220\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:47:35</span> - beginning bbmask <span style=\"font-weight: bold\">(</span>masking + entropy<span style=\"font-weight: bold\">)</span> step                                                               <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#221\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">221</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:47:35\u001b[0m - beginning bbmask \u001b[1m(\u001b[0mmasking + entropy\u001b[1m)\u001b[0m step                                                               \u001b]8;id=867646;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=322806;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#221\u001b\\\u001b[2m221\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:47:35</span> - equivalent to bbmask.sh                                                                                 <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#226\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">226</span></a>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">in</span>=<span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">temp_ncbi_rRNA_all_sequences.fasta</span>              <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">out</span>=<span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_masked.fasta</span>                  <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">sam</span>=<span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">tmp_mapped.sam</span> <span style=\"color: #808000; text-decoration-color: #808000\">overwrite</span>=<span style=\"color: #800080; text-decoration-color: #800080\">true</span>     <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "         <span style=\"color: #808000; text-decoration-color: #808000\">threads</span>=<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">15</span> <span style=\"color: #808000; text-decoration-color: #808000\">Xmx</span>=<span style=\"color: #800080; text-decoration-color: #800080\">68g</span>                                                                                                            <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:47:35\u001b[0m - equivalent to bbmask.sh                                                                                 \u001b]8;id=16356;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=601432;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#226\u001b\\\u001b[2m226\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[33min\u001b[0m=\u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtemp_ncbi_rRNA_all_sequences.fasta\u001b[0m              \u001b[2m               \u001b[0m\n",
       "         \u001b[33mout\u001b[0m=\u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[95mtmp_masked.fasta\u001b[0m                  \u001b[2m               \u001b[0m\n",
       "         \u001b[33msam\u001b[0m=\u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[95mtmp_mapped.sam\u001b[0m \u001b[33moverwrite\u001b[0m=\u001b[35mtrue\u001b[0m     \u001b[2m               \u001b[0m\n",
       "         \u001b[33mthreads\u001b[0m=\u001b[1;36m15\u001b[0m \u001b[33mXmx\u001b[0m=\u001b[35m68g\u001b[0m                                                                                                            \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">java -ea  --add-modules jdk.incubator.vector -Xmx68g -Xms68g -cp </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">bbmap/current/ jgi.BBMask </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences.fasta </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_masked.fasta </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sam</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">overwrite</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">true</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">threads</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> -Xmx68g</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mjava -ea  --add-modules jdk.incubator.vector -Xmx68g -Xms68g -cp \u001b[0m\n",
       "\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/\u001b[0m\n",
       "\u001b[1;31mbbmap/current/\u001b[0m\u001b[1;31m jgi.BBMask \u001b[0m\n",
       "\u001b[1;31min\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31mtemp_ncbi_rRNA_all_sequences.fasta\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_masked.fasta\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31msam\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_mapped.sam\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31moverwrite\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31mtrue\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mthreads\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15\u001b[0m\u001b[1;31m -Xmx68g\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">WARNING: Using incubator modules: jdk.incubator.vector</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWARNING: Using incubator modules: jdk.incubator.vector\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Executing jgi.BBMask </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">[</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences.fasta, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_masked.fasta, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">sam</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/tmp_mapped.sam, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">overwrite</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">true</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">threads</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, -Xmx68g]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mExecuting jgi.BBMask \u001b[0m\n",
       "\u001b[1;31m[\u001b[0m\u001b[1;31min\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31mtemp_ncbi_rRNA_all_sequences.fasta\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_masked.fasta\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31msam\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/tmp_mask_dna/\u001b[0m\u001b[1;31mtmp_mapped.sam\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31moverwrite\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31mtrue\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mthreads\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15\u001b[0m\u001b[1;31m, -Xmx68g\u001b[0m\u001b[1;31m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Set threads to </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSet threads to \u001b[0m\u001b[1;31m15\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Loading input</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLoading input\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Loading Time:                   </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.499</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLoading Time:                   \u001b[0m\u001b[1;31m0.499\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Masking low-entropy (to disable, set </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">'mle=f'</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mMasking low-entropy \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31mto disable, set \u001b[0m\u001b[1;31m'\u001b[0m\u001b[1;31mmle\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31mf\u001b[0m\u001b[1;31m'\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Low Complexity Masking Time:    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.373</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLow Complexity Masking Time:    \u001b[0m\u001b[1;31m0.373\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Ref Bases:                 </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">160837643</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">430.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">65m bases/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mRef Bases:                 \u001b[0m\u001b[1;31m160837643\u001b[0m\u001b[1;31m    \u001b[0m\u001b[1;31m430.\u001b[0m\u001b[1;31m65m bases/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Low Complexity Bases:          </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">17758</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mLow Complexity Bases:          \u001b[0m\u001b[1;31m17758\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Masking from sam</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mMasking from sam\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Masking Time:               </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.085</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Masking Time:               \u001b[0m\u001b[1;31m0.085\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Reads Processed:             </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">428</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">5.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">04k reads/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Reads Processed:             \u001b[0m\u001b[1;31m428\u001b[0m\u001b[1;31m    \u001b[0m\u001b[1;31m5.\u001b[0m\u001b[1;31m04k reads/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Bases Processed:           </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">77637</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">91m bases/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Bases Processed:           \u001b[0m\u001b[1;31m77637\u001b[0m\u001b[1;31m    \u001b[0m\u001b[1;31m0.\u001b[0m\u001b[1;31m91m bases/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Sam Bases Masked:              </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">75030</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSam Bases Masked:              \u001b[0m\u001b[1;31m75030\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Converting masked bases to N</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mConverting masked bases to N\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Done Masking</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mDone Masking\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Conversion Time:                </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.211</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mConversion Time:                \u001b[0m\u001b[1;31m0.211\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Writing output</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWriting output\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Writing Time:                   </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.409</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWriting Time:                   \u001b[0m\u001b[1;31m0.409\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Total Bases Masked:            </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">92788</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">/</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">160837643</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.058</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTotal Bases Masked:            \u001b[0m\u001b[1;31m92788\u001b[0m\u001b[1;31m/\u001b[0m\u001b[1;31m160837643\u001b[0m\u001b[1;31m  \u001b[0m\u001b[1;31m0.058\u001b[0m\u001b[1;31m%\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Total Time:                     </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">1.683</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTotal Time:                     \u001b[0m\u001b[1;31m1.683\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:47:37</span> - Finished bbmask step                                                                                    <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#235\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">235</span></a>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:47:37\u001b[0m - Finished bbmask step                                                                                    \u001b]8;id=213621;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=537910;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#235\u001b\\\u001b[2m235\u001b[0m\u001b]8;;\u001b\\\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #000080; text-decoration-color: #000080\">INFO    </span> <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2025</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">12</span>-<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">28</span> <span style=\"color: #00ff00; text-decoration-color: #00ff00; font-weight: bold\">22:47:37</span> - Masking completed. Output saved to                                                                      <a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">mask_dna.py</span></a><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">:</span><a href=\"file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#271\" target=\"_blank\"><span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">271</span></a>\n",
       "         <span style=\"color: #800080; text-decoration-color: #800080\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/</span><span style=\"color: #ff00ff; text-decoration-color: #ff00ff\">temp_ncbi_rRNA_all_sequences_masked.fasta</span>          <span style=\"color: #7f7f7f; text-decoration-color: #7f7f7f\">               </span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[34mINFO    \u001b[0m \u001b[1;36m2025\u001b[0m-\u001b[1;36m12\u001b[0m-\u001b[1;36m28\u001b[0m \u001b[1;92m22:47:37\u001b[0m - Masking completed. Output saved to                                                                      \u001b]8;id=971148;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py\u001b\\\u001b[2mmask_dna.py\u001b[0m\u001b]8;;\u001b\\\u001b[2m:\u001b[0m\u001b]8;id=272644;file:///clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/src/rolypoly/commands/reads/mask_dna.py#271\u001b\\\u001b[2m271\u001b[0m\u001b]8;;\u001b\\\n",
       "         \u001b[35m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[95mtemp_ncbi_rRNA_all_sequences_masked.fasta\u001b[0m          \u001b[2m               \u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">java -ea  --add-modules jdk.incubator.vector -Xmx14541m -Xms14541m -cp </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">bbmap/current/ jgi.BBDuk </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in1</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences_masked.</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">fasta </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences_masked_</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropy.fasta </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropy</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropyk</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropywindow</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">24</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> maskentropy </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">threads</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mjava -ea  --add-modules jdk.incubator.vector -Xmx14541m -Xms14541m -cp \u001b[0m\n",
       "\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/.pixi/envs/dev/lib/python3.9/site-packages/bbmapy/vendor/\u001b[0m\n",
       "\u001b[1;31mbbmap/current/\u001b[0m\u001b[1;31m jgi.BBDuk \u001b[0m\n",
       "\u001b[1;31min1\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31mtemp_ncbi_rRNA_all_sequences_masked.\u001b[0m\n",
       "\u001b[1;31mfasta\u001b[0m\u001b[1;31m \u001b[0m\n",
       "\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31mtemp_ncbi_rRNA_all_sequences_masked_\u001b[0m\n",
       "\u001b[1;31mentropy.fasta\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mentropy\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m.4\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mentropyk\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m4\u001b[0m\u001b[1;31m \u001b[0m\u001b[1;31mentropywindow\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m24\u001b[0m\u001b[1;31m maskentropy \u001b[0m\u001b[1;31mthreads\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">WARNING: Using incubator modules: jdk.incubator.vector</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mWARNING: Using incubator modules: jdk.incubator.vector\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Executing jgi.BBDuk </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">[</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">in1</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences_masked</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">.fasta, </span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">out</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/temp_ncbi_rRNA_all_sequences_masked_</span>\n",
       "<span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropy.fasta, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropy</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropyk</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">4</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">entropywindow</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">24</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, maskentropy, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">threads</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mExecuting jgi.BBDuk \u001b[0m\n",
       "\u001b[1;31m[\u001b[0m\u001b[1;31min1\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31mtemp_ncbi_rRNA_all_sequences_masked\u001b[0m\n",
       "\u001b[1;31m.fasta\u001b[0m\u001b[1;31m, \u001b[0m\n",
       "\u001b[1;31mout\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m/clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/\u001b[0m\u001b[1;31mtemp_ncbi_rRNA_all_sequences_masked_\u001b[0m\n",
       "\u001b[1;31mentropy.fasta\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mentropy\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m.4\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mentropyk\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m4\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mentropywindow\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m24\u001b[0m\u001b[1;31m, maskentropy, \u001b[0m\u001b[1;31mthreads\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15\u001b[0m\u001b[1;31m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Version </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">39.55</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mVersion \u001b[0m\u001b[1;31m39.55\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Set threads to </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mSet threads to \u001b[0m\u001b[1;31m15\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Allocating kmer table:  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.036</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mAllocating kmer table:  \u001b[0m\u001b[1;31m0.036\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Initial:</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInitial:\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Memory: </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">max</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15250m</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">total</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15250m</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">free</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">15180m</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">, </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">used</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">=</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">70m</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mMemory: \u001b[0m\u001b[1;31mmax\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15250m\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mtotal\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15250m\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mfree\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m15180m\u001b[0m\u001b[1;31m, \u001b[0m\u001b[1;31mused\u001b[0m\u001b[1;31m=\u001b[0m\u001b[1;31m70m\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Input is being processed as unpaired</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInput is being processed as unpaired\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Started output streams: </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.031</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mStarted output streams: \u001b[0m\u001b[1;31m0.031\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Processing time:                </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.785</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mProcessing time:        \u001b[0m\u001b[1;31m        \u001b[0m\u001b[1;31m0.785\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Input:                          </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">145146</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads            </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">160837643</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mInput:                          \u001b[0m\u001b[1;31m145146\u001b[0m\u001b[1;31m reads    \u001b[0m\u001b[1;31m        \u001b[0m\u001b[1;31m160837643\u001b[0m\u001b[1;31m bases.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Entropy-masked:                 </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">394</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.27</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)       </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">12514</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.01</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mEntropy-masked:                 \u001b[0m\u001b[1;31m394\u001b[0m\u001b[1;31m reads \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.27\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\u001b[1;31m       \u001b[0m\u001b[1;31m12514\u001b[0m\u001b[1;31m bases \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.01\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Total Removed:                  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)         </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTotal Removed:                  \u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m reads \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\u001b[1;31m         \u001b[0m\u001b[1;31m0\u001b[0m\u001b[1;31m bases \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m0.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Result:                         </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">145146</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> reads (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">100.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)  </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">160837643</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> bases (</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">100.00</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">%)</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mResult:                         \u001b[0m\u001b[1;31m145146\u001b[0m\u001b[1;31m reads \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m100.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\u001b[1;31m  \u001b[0m\u001b[1;31m160837643\u001b[0m\u001b[1;31m bases \u001b[0m\u001b[1;31m(\u001b[0m\u001b[1;31m100.00\u001b[0m\u001b[1;31m%\u001b[0m\u001b[1;31m)\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Time:                           </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">0.816</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\"> seconds.</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mTime:                           \u001b[0m\u001b[1;31m0.816\u001b[0m\u001b[1;31m seconds.\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Reads Processed:        145k    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">177.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">86k reads/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mReads Processed:        145k    \u001b[0m\u001b[1;31m177.\u001b[0m\u001b[1;31m86k reads/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">Bases Processed:        160m    </span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">197.</span><span style=\"color: #800000; text-decoration-color: #800000; font-weight: bold\">09m bases/sec</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;31mBases Processed:        160m    \u001b[0m\u001b[1;31m197.\u001b[0m\u001b[1;31m09m bases/sec\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished writing NCBI rRNA sequences parquet file.\n",
      "finished writing masked NCBI rRNA fasta file: /clusterfs/jgi/scratch/science/metagen/neri/code/rolypoly/data/contam/rrna/ncbi_rRNA_all_sequences_masked_entropy.fasta\n"
     ]
    }
   ],
   "source": [
    "import rich_click as click\n",
    "\n",
    "# first replace headers with md5 hash as the new_header breaks mappers. Will remap after this\n",
    "write_fasta_file(\n",
    "    headers=rrna_df[\"new_header\"].to_list(),\n",
    "    seqs=rrna_df[\"sequence\"].to_list(),\n",
    "    output_file=os.path.join(rrna_dir, \"temp_ncbi_rRNA_all_sequences.fasta\"),\n",
    ")\n",
    "\n",
    "# mask shared kmers with RNA viruses\n",
    "masked_ncbi_fasta_path = os.path.join(\n",
    "    rrna_dir, \"temp_ncbi_rRNA_all_sequences_masked.fasta\"\n",
    ")\n",
    "\n",
    "entrophy_masked_ncbi_fasta_path = os.path.join(\n",
    "    rrna_dir, \"temp_ncbi_rRNA_all_sequences_masked_entropy.fasta\"\n",
    ")\n",
    "\n",
    "mask_args = {\n",
    "    \"threads\": 15,\n",
    "    \"memory\": \"68gb\",\n",
    "    \"output\": str(masked_ncbi_fasta_path),\n",
    "    \"flatten\": False,\n",
    "    \"aligner\": \"mmseqs2\",\n",
    "    \"input\": os.path.join(rrna_dir, \"temp_ncbi_rRNA_all_sequences.fasta\"),\n",
    "}\n",
    "\n",
    "# ensure we have the click Command object (the package import may have bound the module)\n",
    "from rolypoly.commands.reads.mask_dna import mask_dna as _mask_cmd\n",
    "\n",
    "# If it's a click.Command use a Context to invoke, otherwise call directly\n",
    "if isinstance(_mask_cmd, click.core.Command):\n",
    "    ctx = click.Context(_mask_cmd)\n",
    "    ctx.invoke(_mask_cmd, **mask_args)\n",
    "else:\n",
    "    # fallback: call the function directly\n",
    "    _mask_cmd(**mask_args)\n",
    "\n",
    "# # Apply entropy masking\n",
    "bbduk(\n",
    "    in1=masked_ncbi_fasta_path,\n",
    "    out=entrophy_masked_ncbi_fasta_path,\n",
    "    entropy=0.4,\n",
    "    entropyk=4,\n",
    "    entropywindow=24,\n",
    "    maskentropy=True,\n",
    "    threads=15,\n",
    ")\n",
    "\n",
    "# read the masked fasta and remap to original headers\n",
    "ncbi_rrna_masked_df = from_fastx_eager(entrophy_masked_ncbi_fasta_path).rename(\n",
    "    {\"header\": \"new_header\", \"sequence\": \"masked_sequence\"}\n",
    ")\n",
    "ncbi_rrna_masked_df = ncbi_rrna_masked_df.join(\n",
    "    rrna_df, on=[\"new_header\"], how=\"inner\"\n",
    ")\n",
    "ncbi_rrna_masked_df.write_parquet(\n",
    "    os.path.join(rrna_dir, \"ncbi_rrna_sequences_masked.parquet\")\n",
    ")\n",
    "print(\"Finished writing NCBI rRNA sequences parquet file.\")\n",
    "write_fasta_file(\n",
    "    headers=ncbi_rrna_masked_df[\"new_header\"].to_list(),\n",
    "    seqs=ncbi_rrna_masked_df[\"masked_sequence\"].to_list(),\n",
    "    output_file=os.path.join(\n",
    "        rrna_dir, \"ncbi_rRNA_all_sequences_masked_entropy.fasta\"\n",
    "    ),\n",
    ")\n",
    "print(\n",
    "    f\"finished writing masked NCBI rRNA fasta file: {os.path.join(rrna_dir, 'ncbi_rRNA_all_sequences_masked_entropy.fasta')}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3cdd589",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (25_057, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>organism_name</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>null</td><td>105619</td></tr><tr><td>&quot;Globisporangium heterothallicu…</td><td>93</td></tr><tr><td>&quot;Babesia bovis&quot;</td><td>65</td></tr><tr><td>&quot;Globisporangium sylvaticum&quot;</td><td>61</td></tr><tr><td>&quot;Globisporangium irregulare&quot;</td><td>56</td></tr><tr><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;Sphingobacterium deserti&quot;</td><td>1</td></tr><tr><td>&quot;Pararcticibacter amylolyticus&quot;</td><td>1</td></tr><tr><td>&quot;Leontodon rigens&quot;</td><td>1</td></tr><tr><td>&quot;Bradyrhizobium diversitatis&quot;</td><td>1</td></tr><tr><td>&quot;Botrimarina colliarenosi&quot;</td><td>1</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (25_057, 2)\n",
       "┌─────────────────────────────────┬────────┐\n",
       "│ organism_name                   ┆ count  │\n",
       "│ ---                             ┆ ---    │\n",
       "│ str                             ┆ u32    │\n",
       "╞═════════════════════════════════╪════════╡\n",
       "│ null                            ┆ 105619 │\n",
       "│ Globisporangium heterothallicu… ┆ 93     │\n",
       "│ Babesia bovis                   ┆ 65     │\n",
       "│ Globisporangium sylvaticum      ┆ 61     │\n",
       "│ Globisporangium irregulare      ┆ 56     │\n",
       "│ …                               ┆ …      │\n",
       "│ Sphingobacterium deserti        ┆ 1      │\n",
       "│ Pararcticibacter amylolyticus   ┆ 1      │\n",
       "│ Leontodon rigens                ┆ 1      │\n",
       "│ Bradyrhizobium diversitatis     ┆ 1      │\n",
       "│ Botrimarina colliarenosi        ┆ 1      │\n",
       "└─────────────────────────────────┴────────┘"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df = ncbi_rrna_masked_df.join(\n",
    "    mini_ncbi.select([\"ncbi_taxonid\", \"organism_name\"]).unique(),  # seq_length\n",
    "    left_on=[\"taxid\"],\n",
    "    right_on=[\"ncbi_taxonid\"],\n",
    "    how=\"left\",\n",
    ")\n",
    "temp_df[\"organism_name\"].value_counts(sort=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12737bc5",
   "metadata": {},
   "source": [
    "some 105k nulls, these are the ones that will need to patch to link to cloest related organism that does have genome/transcriptome data on genbank.  \n",
    "That is done in [rrna_genome_mapping_taxonomy.ipynb](rrna_genome_mapping_taxonomy.ipynb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7a6e65e",
   "metadata": {},
   "source": [
    "## Now tRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6d2cb41",
   "metadata": {},
   "outputs": [],
   "source": [
    "file_url = (\n",
    "    \"https://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/fasta_files/RF00005.fa.gz\"\n",
    ")\n",
    "trna_seqs = os.path.join(trna_dir, \"tRNA_sequences.fasta\")\n",
    "gz_filename = \"RF00005.fa.gz\"\n",
    "fetch_and_extract(\n",
    "    url=file_url,\n",
    "    fetched_to=os.path.join(trna_dir, gz_filename),\n",
    "    extract_to=trna_dir,\n",
    "    expected_file=trna_seqs,\n",
    ")\n",
    "print(f\"Downloaded tRNA sequences to {trna_seqs}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "515356d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove duplicates\n",
    "deduplicated_fasta = os.path.join(trna_dir, \"tRNA_sequences_deduplicated.fasta\")\n",
    "remove_duplicates(\n",
    "    input_file=trna_seqs,\n",
    "    output_file=deduplicated_fasta,\n",
    "    return_stats=True,\n",
    "    by=\"seq\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d02b442",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rolypoly.utils.bio.polars_fastx import compute_aggregate_stats, fasta_stats\n",
    "\n",
    "info_table = fasta_stats(deduplicated_fasta)\n",
    "# print(info_table)\n",
    "\n",
    "some_stats = compute_aggregate_stats(\n",
    "    info_table, fields=[\"length\", \"gc_content\", \"n_content\"]\n",
    ")\n",
    "some_stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d21a27f",
   "metadata": {},
   "source": [
    "I don't trust this (min len 30, min gc 8% ), so will do some light filteration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f927d279",
   "metadata": {},
   "outputs": [],
   "source": [
    "info_table = info_table.filter(\n",
    "    pl.col(\"length\").is_between(60, 250), pl.col(\"gc_content\") >= 0.01\n",
    ")\n",
    "some_stats_filtered = compute_aggregate_stats(\n",
    "    info_table, fields=[\"length\", \"gc_content\", \"n_content\"]\n",
    ")\n",
    "some_stats_filtered\n",
    "# info_table.write_parquet(os.path.join(trna_dir, \"tRNA_sequences_deduplicated_stats.parquet\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d1bb6c0",
   "metadata": {},
   "source": [
    "looks a little better, will just accept that very low or very high GC content tRNAs exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbd4ccc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_fasta_file(\n",
    "    seqs=info_table[\"sequence\"].to_list(),\n",
    "    headers=info_table[\"header\"].to_list(),\n",
    "    output_file=os.path.join(\n",
    "        trna_dir, \"tRNA_sequences_deduplicated_filtered.fasta\"\n",
    "    ),\n",
    ")\n",
    "print(\n",
    "    f\"Wrote filtered tRNA sequences to {os.path.join(trna_dir, 'tRNA_sequences_deduplicated_filtered.fasta')}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b21a1429",
   "metadata": {},
   "source": [
    "## plastid and mitochondria genomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "545d92dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from rolypoly.commands.misc.build_data import (\n",
    "    prepare_plastid_data,\n",
    "    prepare_mito_data,\n",
    ")\n",
    "\n",
    "prepare_plastid_data(data_dir, logger)\n",
    "prepare_mito_data(data_dir, logger)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
